

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Analysis of binned data &mdash; HiC pipeline for iterative correction v0.4 documentation</title>
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.4',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="HiC pipeline for iterative correction v0.4 documentation" href="index.html" />
    <link rel="prev" title="Fragment-level HiC correction" href="fragment.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="np-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="fragment.html" title="Fragment-level HiC correction"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">HiC pipeline for iterative correction v0.4 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="module-binnedData">
<span id="analysis-of-binned-data"></span><h1>Analysis of binned data<a class="headerlink" href="#module-binnedData" title="Permalink to this headline">¶</a></h1>
<p>Binned data - analysis of HiC, binned to resolution.</p>
<dl class="class">
<dt id="binnedData.binnedData">
<em class="property">class </em><tt class="descclassname">binnedData.</tt><tt class="descname">binnedData</tt><big>(</big><em>resolution</em>, <em>genome</em><big>)</big><a class="headerlink" href="#binnedData.binnedData" title="Permalink to this definition">¶</a></dt>
<dd><p>base class to work with binned data</p>
<p class="rubric">Methods</p>
<p>Sets up the Genome object and resolution.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>resolution</strong> : int</p>
<blockquote>
<p>Resolution of all datasets</p>
</blockquote>
<p class="last"><strong>genome</strong> : genome Folder or Genome object</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<dl class="method">
<dt id="binnedData.binnedData.initChromosomes">
<tt class="descname">initChromosomes</tt><big>(</big><big>)</big><a class="headerlink" href="#binnedData.binnedData.initChromosomes" title="Permalink to this definition">¶</a></dt>
<dd><p>loads mappings from the genome class based on resolution</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.simpleLoad">
<tt class="descname">simpleLoad</tt><big>(</big><em>filename</em>, <em>name</em>, <em>checkGenomes=True</em><big>)</big><a class="headerlink" href="#binnedData.binnedData.simpleLoad" title="Permalink to this definition">¶</a></dt>
<dd><p>load standard type data</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.loadGC">
<tt class="descname">loadGC</tt><big>(</big><big>)</big><a class="headerlink" href="#binnedData.binnedData.loadGC" title="Permalink to this definition">¶</a></dt>
<dd><p>loads GC content at given resolution</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.removeDiagonal">
<tt class="descname">removeDiagonal</tt><big>(</big><em>m</em><big>)</big><a class="headerlink" href="#binnedData.binnedData.removeDiagonal" title="Permalink to this definition">¶</a></dt>
<dd><p>does what it says from all datasets</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.removeStandalone">
<tt class="descname">removeStandalone</tt><big>(</big><em>offset=3</em><big>)</big><a class="headerlink" href="#binnedData.binnedData.removeStandalone" title="Permalink to this definition">¶</a></dt>
<dd><p>removes standalone bins (groups of less-than-offset-long bins)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>offset</strong> : int</p>
<blockquote class="last">
<p>Maximum length of group of bins to be removed</p>
</blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.removePoorRegions">
<tt class="descname">removePoorRegions</tt><big>(</big><em>names=None</em>, <em>cutoff=2</em><big>)</big><a class="headerlink" href="#binnedData.binnedData.removePoorRegions" title="Permalink to this definition">¶</a></dt>
<dd><p>removes cutoff persent of bins with least counts</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.trunkTrans">
<tt class="descname">trunkTrans</tt><big>(</big><em>high=0.0005</em><big>)</big><a class="headerlink" href="#binnedData.binnedData.trunkTrans" title="Permalink to this definition">¶</a></dt>
<dd><p>trunkates trans contacts to remove blowouts</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.removeCis">
<tt class="descname">removeCis</tt><big>(</big><big>)</big><a class="headerlink" href="#binnedData.binnedData.removeCis" title="Permalink to this definition">¶</a></dt>
<dd><p>sets to zero all cis contacts</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.fakeCisOnce">
<tt class="descname">fakeCisOnce</tt><big>(</big><em>extraMask=None</em><big>)</big><a class="headerlink" href="#binnedData.binnedData.fakeCisOnce" title="Permalink to this definition">¶</a></dt>
<dd><p>used to fake cis counts.
If extra mask is supplied, it also fakes stuff in the extra mask</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.fakeCis">
<tt class="descname">fakeCis</tt><big>(</big><big>)</big><a class="headerlink" href="#binnedData.binnedData.fakeCis" title="Permalink to this definition">¶</a></dt>
<dd><p>fakes cis contacts in an interative way</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.emulateCis">
<tt class="descname">emulateCis</tt><big>(</big><big>)</big><a class="headerlink" href="#binnedData.binnedData.emulateCis" title="Permalink to this definition">¶</a></dt>
<dd><p>if you want to have fun creating syntetic data, this emulates cis contacts. 
adjust cis/trans ratio in the code</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.fakeMissing">
<tt class="descname">fakeMissing</tt><big>(</big><em>stay=False</em><big>)</big><a class="headerlink" href="#binnedData.binnedData.fakeMissing" title="Permalink to this definition">¶</a></dt>
<dd><p>fakes missing megabases</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.correct">
<tt class="descname">correct</tt><big>(</big><em>names=None</em><big>)</big><a class="headerlink" href="#binnedData.binnedData.correct" title="Permalink to this definition">¶</a></dt>
<dd><p>performs single correction without SS</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.iterativeCorrectWithoutSS">
<tt class="descname">iterativeCorrectWithoutSS</tt><big>(</big><em>names=None</em>, <em>M=50</em><big>)</big><a class="headerlink" href="#binnedData.binnedData.iterativeCorrectWithoutSS" title="Permalink to this definition">¶</a></dt>
<dd><p>performs iterative correction wihtout SS</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.iterativeCorrectWithSS">
<tt class="descname">iterativeCorrectWithSS</tt><big>(</big><em>names=None</em>, <em>M=55</em><big>)</big><a class="headerlink" href="#binnedData.binnedData.iterativeCorrectWithSS" title="Permalink to this definition">¶</a></dt>
<dd><p>performs iterative correction with SS</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.iterativeCorrectByTrans">
<tt class="descname">iterativeCorrectByTrans</tt><big>(</big><em>names=None</em><big>)</big><a class="headerlink" href="#binnedData.binnedData.iterativeCorrectByTrans" title="Permalink to this definition">¶</a></dt>
<dd><p>performs iterative correction by trans data only, corrects cis also</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.removeZeros">
<tt class="descname">removeZeros</tt><big>(</big><big>)</big><a class="headerlink" href="#binnedData.binnedData.removeZeros" title="Permalink to this definition">¶</a></dt>
<dd><p>removes bins with zero counts
keeps chromosome starts, ends, etc. consistent</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.restoreZeros">
<tt class="descname">restoreZeros</tt><big>(</big><em>value=nan</em><big>)</big><a class="headerlink" href="#binnedData.binnedData.restoreZeros" title="Permalink to this definition">¶</a></dt>
<dd><p>Restores zeros that were removed by removeZeros command.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">You can restore zeros only if you used removeZeros once.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.doPCA">
<tt class="descname">doPCA</tt><big>(</big><big>)</big><a class="headerlink" href="#binnedData.binnedData.doPCA" title="Permalink to this definition">¶</a></dt>
<dd><p>performs PCA on the data
creates dictionary self.PCA with results</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.doEig">
<tt class="descname">doEig</tt><big>(</big><big>)</big><a class="headerlink" href="#binnedData.binnedData.doEig" title="Permalink to this definition">¶</a></dt>
<dd><p>performs eigenvector expansion on the data
creates dictionary self.EIG with results</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedData.cisToTrans">
<tt class="descname">cisToTrans</tt><big>(</big><em>mode='All'</em>, <em>filename='GM-all'</em><big>)</big><a class="headerlink" href="#binnedData.binnedData.cisToTrans" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates cis-to-trans ratio.
&#8220;All&#8221; - treating SS as trans reads
&#8220;Dummy&#8221; - fake SS reads proportional to cis reads with the same total sum
&#8220;Matrix&#8221; - use heatmap only</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="binnedData.binnedDataAnalysis">
<em class="property">class </em><tt class="descclassname">binnedData.</tt><tt class="descname">binnedDataAnalysis</tt><big>(</big><em>resolution</em>, <em>genome</em><big>)</big><a class="headerlink" href="#binnedData.binnedDataAnalysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Class containing experimental features and data analysis scripts</p>
<p class="rubric">Methods</p>
<p>Sets up the Genome object and resolution.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>resolution</strong> : int</p>
<blockquote>
<p>Resolution of all datasets</p>
</blockquote>
<p class="last"><strong>genome</strong> : genome Folder or Genome object</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<dl class="method">
<dt id="binnedData.binnedDataAnalysis.plotScaling">
<tt class="descname">plotScaling</tt><big>(</big><em>name</em>, <em>label='BLA'</em>, <em>color=None</em><big>)</big><a class="headerlink" href="#binnedData.binnedDataAnalysis.plotScaling" title="Permalink to this definition">¶</a></dt>
<dd><p>plots scaling of a heatmap,treating arms separately</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedDataAnalysis.averageTransMap">
<tt class="descname">averageTransMap</tt><big>(</big><em>name='GM-all'</em>, <em>mycmap='hot_r'</em>, <em>chromMax=22</em>, <em>vmin=None</em>, <em>vmax=None</em><big>)</big><a class="headerlink" href="#binnedData.binnedDataAnalysis.averageTransMap" title="Permalink to this definition">¶</a></dt>
<dd><p>plots and returns average inter-chromosomal inter-arm map</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedDataAnalysis.perArmCorrelation">
<tt class="descname">perArmCorrelation</tt><big>(</big><em>data1</em>, <em>data2</em>, <em>doByArms=</em><span class="optional">[</span><span class="optional">]</span><big>)</big><a class="headerlink" href="#binnedData.binnedDataAnalysis.perArmCorrelation" title="Permalink to this definition">¶</a></dt>
<dd><p>does inter-chromosomal spearman correlation 
of two vectors for each chromosomes separately.
Averages over chromosomes with weight of chromosomal length
For chromosomes in &#8220;doByArms&#8221; treats arms as separatre chromosomes
returns average Spearman r correlation</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedDataAnalysis.divideOutAveragesPerChromosome">
<tt class="descname">divideOutAveragesPerChromosome</tt><big>(</big><big>)</big><a class="headerlink" href="#binnedData.binnedDataAnalysis.divideOutAveragesPerChromosome" title="Permalink to this definition">¶</a></dt>
<dd><p>divides each interchromosomal map by it&#8217;s mean value</p>
</dd></dl>

<dl class="method">
<dt id="binnedData.binnedDataAnalysis.interchromosomalValues">
<tt class="descname">interchromosomalValues</tt><big>(</big><em>filename='GM-all'</em>, <em>returnAll=False</em><big>)</big><a class="headerlink" href="#binnedData.binnedDataAnalysis.interchromosomalValues" title="Permalink to this definition">¶</a></dt>
<dd><p>returns average inter-chromosome-interaction values, ordered always the same way</p>
</dd></dl>

</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="fragment.html"
                        title="previous chapter">Fragment-level HiC correction</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/binneddata.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" size="18" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="np-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="fragment.html" title="Fragment-level HiC correction"
             >previous</a> |</li>
        <li><a href="index.html">HiC pipeline for iterative correction v0.4 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2012, Maxim Imakaev, Anton Goloborodko, Geoffrey Fudenberg, Leonid Mirny.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.1.
    </div>
  </body>
</html>