

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Analysis of binned data &mdash; mirnylab v0.1 documentation</title>
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="mirnylab v0.1 documentation" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="np-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="index.html">mirnylab v0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="module-hiclib.binnedData">
<span id="analysis-of-binned-data"></span><h1>Analysis of binned data<a class="headerlink" href="#module-hiclib.binnedData" title="Permalink to this headline">¶</a></h1>
<p>Binned data - analysis of HiC, binned to resolution.</p>
<div class="section" id="concepts">
<h2>Concepts<a class="headerlink" href="#concepts" title="Permalink to this headline">¶</a></h2>
<p>class Binned Data allows low-level manipulation of  multiple HiC datasets, 
binned to the same resolution from the same genome.</p>
<p>When working with multiple datasets, all the filters will be synchronized,
so only bins present in all datasets will be considered for the analysis. 
Removal of bins from one dataset will remove them from the others. 
E.g. removing 1% of bins with lowest # of count might remove more than 1% of total bins, 
when working with 2 or more datasets.</p>
<p>Class has significant knowledge about filters that have been applied. 
If an essential filter was not applied, it will throw an exception; 
if adviced filter is not applied, it will throw a warning. 
Most of the methods have an optional &#8220;force&#8221; argument that will ignore dependensies.</p>
<p>We provide example scripts that show ideal protocols for certain types of the analysis, 
but they don&#8217;t cover the realm of all possible manipulations that can be performed with this class.</p>
</div>
<div class="section" id="input-data">
<h2>Input data<a class="headerlink" href="#input-data" title="Permalink to this headline">¶</a></h2>
<p>method <a class="reference internal" href="#hiclib.binnedData.binnedData.simpleLoad" title="hiclib.binnedData.binnedData.simpleLoad"><tt class="xref py py-func docutils literal"><span class="pre">SimpleLoad</span></tt></a> may be used to load the data. 
It automatically checks for possible genome length mismatch.</p>
<p>This method works best with h5dict files, created by fragmentHiC.
In this case you just need to supply the filename.</p>
<p>It can also accept any dictionary-like object with the following keys, 
where all but &#8220;heatmap&#8221; is optional.</p>
<ul class="simple">
<li>[&#8220;heatmap&#8221;] : all-by-all heatmap</li>
<li>[&#8220;singles&#8221;] : vector of SS reads, optional</li>
<li>[&#8220;frags&#8221;] : number of rsites per bin, optional</li>
<li>[&#8220;resolution&#8221;] : resolution</li>
</ul>
<p>All information about the genome, including GC content and restriction sites, can be obtained from the Genome class.</p>
<p>Genomic tracks can be loaded using an automated parser that accepts bigWig files and fixed step wiggle files. 
See documentation for <a class="reference internal" href="#hiclib.binnedData.experimentalBinnedData.loadWigFile" title="hiclib.binnedData.experimentalBinnedData.loadWigFile"><tt class="xref py py-func docutils literal"><span class="pre">experimentalBinnedData.loadWigFile()</span></tt></a> that describes exactly how the data is averaged and parsed.</p>
</div>
<div class="section" id="variables">
<h2>Variables<a class="headerlink" href="#variables" title="Permalink to this headline">¶</a></h2>
<p>self.dataDict - dictionary with heatmaps; keys are provided when loading the data.</p>
<p>self.singlesDict - dictionary with SS read vectors. Keys are the same.</p>
<p>self.fragsDict - dictionary with fragment density data</p>
<p>self.trackDict - dictionary with genomic tracks, such as GC content. Custom tracks should be added here.</p>
<p>self.biasDict - dictionary with biases as calculated by iterative correction (incomplete)</p>
<p>self.PCDict - dictionary with principal components of each datasets. Keys as in dataDict</p>
<p>self.EigEict - dictionary with eigenvectors for each dataset. Keys as in datadict.</p>
</div>
<div class="section" id="hierarchy-of-filters">
<h2>Hierarchy of filters<a class="headerlink" href="#hierarchy-of-filters" title="Permalink to this headline">¶</a></h2>
<p>hierarchy</p>
<hr class="docutils" />
<dl class="class">
<dt id="hiclib.binnedData.binnedData">
<em class="property">class </em><tt class="descclassname">hiclib.binnedData.</tt><tt class="descname">binnedData</tt><big>(</big><em>resolution</em>, <em>genome</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData" title="Permalink to this definition">¶</a></dt>
<dd><p>Base class to work with binned data, the most documented and robust part of the code. 
Further classes for other analysis are inherited from this class.</p>
<p class="rubric">Methods</p>
<p>self.__init__ - initializes an empty dataset.</p>
<p>This method sets up a Genome object and resolution.</p>
<p>Genome object specifies genome version and inclusion/exclusion of sex chromosomes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>resolution</strong> : int</p>
<blockquote>
<p>Resolution of all datasets</p>
</blockquote>
<p class="last"><strong>genome</strong> : genome Folder or Genome object</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<dl class="method">
<dt id="hiclib.binnedData.binnedData.simpleLoad">
<tt class="descname">simpleLoad</tt><big>(</big><em>in_data</em>, <em>name</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.simpleLoad"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.simpleLoad" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads data from h5dict file or dict-like object</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>in_data</strong> : str or dict-like</p>
<blockquote>
<p>h5dict filename or dictionary-like object with input data</p>
</blockquote>
<p><strong>name</strong> : str</p>
<blockquote class="last">
<p>Key of the dataset in self.dataDict</p>
</blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.removeDiagonal">
<tt class="descname">removeDiagonal</tt><big>(</big><em>m=1</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.removeDiagonal"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.removeDiagonal" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes all bins on a diagonal, and bins that are up to m away from the diagonal, including m. 
By default, removes all bins touching the diagonal.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>m</strong> : int, optional</p>
<blockquote class="last">
<p>Number of bins to remove</p>
</blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.removeStandalone">
<tt class="descname">removeStandalone</tt><big>(</big><em>offset=3</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.removeStandalone"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.removeStandalone" title="Permalink to this definition">¶</a></dt>
<dd><p>removes standalone bins (groups of less-than-offset bins)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>offset</strong> : int</p>
<blockquote class="last">
<p>Maximum length of group of bins to be removed</p>
</blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.removePoorRegions">
<tt class="descname">removePoorRegions</tt><big>(</big><em>names=None</em>, <em>cutoff=2</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.removePoorRegions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.removePoorRegions" title="Permalink to this definition">¶</a></dt>
<dd><p>removes cutoff persent of bins with least counts</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>names</strong> : list of str</p>
<blockquote>
<p>List of datasets to perform the filter. All by default.</p>
</blockquote>
<p><strong>cutoff</strong> : int, 0&lt;cutoff&lt;100</p>
<blockquote class="last">
<p>Percent of lowest-counts bins to be removed</p>
</blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.truncTrans">
<tt class="descname">truncTrans</tt><big>(</big><em>high=0.0005</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.truncTrans"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.truncTrans" title="Permalink to this definition">¶</a></dt>
<dd><p>Truncates trans contacts to remove blowouts</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>high</strong> : float, 0&lt;high&lt;1, optional</p>
<blockquote class="last">
<p>Fraction of top trans interactions to be removed</p>
</blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.removeCis">
<tt class="descname">removeCis</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.removeCis"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.removeCis" title="Permalink to this definition">¶</a></dt>
<dd><p>sets to zero all cis contacts</p>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.fakeCisOnce">
<tt class="descname">fakeCisOnce</tt><big>(</big><em>mask='CisCounts'</em>, <em>silent=False</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.fakeCisOnce"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.fakeCisOnce" title="Permalink to this definition">¶</a></dt>
<dd><p>Used to fake cis counts or any other region with random trans counts. 
If extra mask is supplied, it is used instead of cis counts.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>mask</strong> : NxN boolean array or &#8220;CisCounts&#8221;</p>
<blockquote>
<p>Mask of elements to be faked. 
If set to &#8220;CisCounts&#8221;, cis counts will be faked</p>
</blockquote>
<p><strong>silent</strong> : bool</p>
<blockquote class="last">
<p>Do not print anything</p>
</blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.fakeCis">
<tt class="descname">fakeCis</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.fakeCis"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.fakeCis" title="Permalink to this definition">¶</a></dt>
<dd><p>This method fakes cis contacts in an interative way
It is done to achieve faking cis contacts that is independent of normalization of the data.</p>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.fakeTranslocations">
<tt class="descname">fakeTranslocations</tt><big>(</big><em>translocationRegions</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.fakeTranslocations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.fakeTranslocations" title="Permalink to this definition">¶</a></dt>
<dd><p>This method fakes reads corresponding to a translocation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>translocationRegions: list of tuples</strong> :</p>
<blockquote class="last">
<p>List of tuples (chr1,start1,end1,chr2,start2,end2), masking a high-count region around visible translocation.
If chromosome end is None, it is treated as length of chromosome. So, use (chr1,0,None,chr2,0,None) to remove inter-chromosomal interaction entirely.</p>
</blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.correct">
<tt class="descname">correct</tt><big>(</big><em>names=None</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.correct"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.correct" title="Permalink to this definition">¶</a></dt>
<dd><p>performs single correction without SS</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>names</strong> : list of str or None</p>
<blockquote class="last">
<p>Keys of datasets to be corrected. If none, all are corrected.</p>
</blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.iterativeCorrectWithoutSS">
<tt class="descname">iterativeCorrectWithoutSS</tt><big>(</big><em>names=None</em>, <em>M=50</em>, <em>force=False</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.iterativeCorrectWithoutSS"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.iterativeCorrectWithoutSS" title="Permalink to this definition">¶</a></dt>
<dd><p>performs iterative correction without SS</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>names</strong> : list of str or None, optional</p>
<blockquote>
<p>Keys of datasets to be corrected. By default, all are corrected.</p>
</blockquote>
<p><strong>M</strong> : int, optional</p>
<blockquote class="last">
<p>Number of iterations to perform.</p>
</blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.iterativeCorrectWithSS">
<tt class="descname">iterativeCorrectWithSS</tt><big>(</big><em>names=None</em>, <em>M=55</em>, <em>force=False</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.iterativeCorrectWithSS"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.iterativeCorrectWithSS" title="Permalink to this definition">¶</a></dt>
<dd><p>performs iterative correction with SS</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>names</strong> : list of str or None, optional</p>
<blockquote>
<p>Keys of datasets to be corrected. By default, all are corrected.</p>
</blockquote>
<p><strong>M</strong> : int, optional</p>
<blockquote>
<p>Number of iterations to perform.</p>
</blockquote>
<p><strong>force</strong> : bool, optional</p>
<blockquote class="last">
<p>Force current operation</p>
</blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.removeChromosome">
<tt class="descname">removeChromosome</tt><big>(</big><em>chromNum</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.removeChromosome"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.removeChromosome" title="Permalink to this definition">¶</a></dt>
<dd><p>removes certain chromosome from all tracks and heatmaps, setting all values to zero</p>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.removeZeros">
<tt class="descname">removeZeros</tt><big>(</big><em>zerosMask=None</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.removeZeros"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.removeZeros" title="Permalink to this definition">¶</a></dt>
<dd><p>removes bins with zero counts
keeps chromosome starts, ends, etc. consistent</p>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.restoreZeros">
<tt class="descname">restoreZeros</tt><big>(</big><em>value=nan</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.restoreZeros"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.restoreZeros" title="Permalink to this definition">¶</a></dt>
<dd><p>Restores zeros that were removed by removeZeros command.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">You can restore zeros only if you used removeZeros once.</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>value</strong> : number-like, optional.</p>
<blockquote class="last">
<p>Value to fill in missing regions. By default, NAN.</p>
</blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.doPCA">
<tt class="descname">doPCA</tt><big>(</big><em>force=False</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.doPCA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.doPCA" title="Permalink to this definition">¶</a></dt>
<dd><p>performs PCA on the data
creates dictionary self.PCADict with results
Last column of PC matrix is first PC, second to last - second, etc.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Returns :</th><td class="field-body"><strong>Dictionary of principal component matrices for different datasets</strong> :</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.doEig">
<tt class="descname">doEig</tt><big>(</big><em>numPCs=3</em>, <em>force=False</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.doEig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.doEig" title="Permalink to this definition">¶</a></dt>
<dd><p>performs eigenvector expansion on the data
creates dictionary self.EigDict with results
Last row of the eigenvector matrix is the largest eigenvector, etc.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Returns :</th><td class="field-body"><strong>Dictionary of eigenvector matrices for different datasets</strong> :</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedData.cisToTrans">
<tt class="descname">cisToTrans</tt><big>(</big><em>mode='All'</em>, <em>filename='GM-all'</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedData.cisToTrans"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedData.cisToTrans" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates cis-to-trans ratio.
&#8220;All&#8221; - treating SS as trans reads
&#8220;Dummy&#8221; - fake SS reads proportional to cis reads with the same total sum
&#8220;Matrix&#8221; - use heatmap only</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="hiclib.binnedData.binnedDataAnalysis">
<em class="property">class </em><tt class="descclassname">hiclib.binnedData.</tt><tt class="descname">binnedDataAnalysis</tt><big>(</big><em>resolution</em>, <em>genome</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedDataAnalysis"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedDataAnalysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Class containing experimental features and data analysis scripts</p>
<p class="rubric">Methods</p>
<p>self.__init__ - initializes an empty dataset.</p>
<p>This method sets up a Genome object and resolution.</p>
<p>Genome object specifies genome version and inclusion/exclusion of sex chromosomes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>resolution</strong> : int</p>
<blockquote>
<p>Resolution of all datasets</p>
</blockquote>
<p class="last"><strong>genome</strong> : genome Folder or Genome object</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<dl class="method">
<dt id="hiclib.binnedData.binnedDataAnalysis.plotScaling">
<tt class="descname">plotScaling</tt><big>(</big><em>name</em>, <em>label='BLA'</em>, <em>color=None</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedDataAnalysis.plotScaling"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedDataAnalysis.plotScaling" title="Permalink to this definition">¶</a></dt>
<dd><p>plots scaling of a heatmap,treating arms separately</p>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedDataAnalysis.averageTransMap">
<tt class="descname">averageTransMap</tt><big>(</big><em>name</em>, <em>mycmap='hot_r'</em>, <em>vmin=None</em>, <em>vmax=None</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedDataAnalysis.averageTransMap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedDataAnalysis.averageTransMap" title="Permalink to this definition">¶</a></dt>
<dd><p>plots and returns average inter-chromosomal inter-arm map</p>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedDataAnalysis.perArmCorrelation">
<tt class="descname">perArmCorrelation</tt><big>(</big><em>data1</em>, <em>data2</em>, <em>doByArms=</em><span class="optional">[</span><span class="optional">]</span><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedDataAnalysis.perArmCorrelation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedDataAnalysis.perArmCorrelation" title="Permalink to this definition">¶</a></dt>
<dd><p>does inter-chromosomal spearman correlation 
of two vectors for each chromosomes separately.
Averages over chromosomes with weight of chromosomal length
For chromosomes in &#8220;doByArms&#8221; treats arms as separatre chromosomes
returns average Spearman r correlation</p>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedDataAnalysis.divideOutAveragesPerChromosome">
<tt class="descname">divideOutAveragesPerChromosome</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedDataAnalysis.divideOutAveragesPerChromosome"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedDataAnalysis.divideOutAveragesPerChromosome" title="Permalink to this definition">¶</a></dt>
<dd><p>divides each interchromosomal map by it&#8217;s mean value</p>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.binnedDataAnalysis.interchromosomalValues">
<tt class="descname">interchromosomalValues</tt><big>(</big><em>filename='GM-all'</em>, <em>returnAll=False</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#binnedDataAnalysis.interchromosomalValues"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.binnedDataAnalysis.interchromosomalValues" title="Permalink to this definition">¶</a></dt>
<dd><p>returns average inter-chromosome-interaction values, ordered always the same way</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="hiclib.binnedData.experimentalBinnedData">
<em class="property">class </em><tt class="descclassname">hiclib.binnedData.</tt><tt class="descname">experimentalBinnedData</tt><big>(</big><em>resolution</em>, <em>genome</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#experimentalBinnedData"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.experimentalBinnedData" title="Permalink to this definition">¶</a></dt>
<dd><p>Contains some poorly-implemented new features</p>
<p class="rubric">Methods</p>
<p>self.__init__ - initializes an empty dataset.</p>
<p>This method sets up a Genome object and resolution.</p>
<p>Genome object specifies genome version and inclusion/exclusion of sex chromosomes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>resolution</strong> : int</p>
<blockquote>
<p>Resolution of all datasets</p>
</blockquote>
<p class="last"><strong>genome</strong> : genome Folder or Genome object</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<dl class="method">
<dt id="hiclib.binnedData.experimentalBinnedData.emulateCis">
<tt class="descname">emulateCis</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#experimentalBinnedData.emulateCis"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.experimentalBinnedData.emulateCis" title="Permalink to this definition">¶</a></dt>
<dd><p>if you want to have fun creating syntetic data, this emulates cis contacts. 
adjust cis/trans ratio in the C code</p>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.experimentalBinnedData.fakeMissing">
<tt class="descname">fakeMissing</tt><big>(</big><em>iterative=True</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#experimentalBinnedData.fakeMissing"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.experimentalBinnedData.fakeMissing" title="Permalink to this definition">¶</a></dt>
<dd><p>fakes megabases that have no reads. For cis reads fakes with cis reads at the same distance. For trans fakes with random trans read at the same diagonal.</p>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.experimentalBinnedData.iterativeCorrectByTrans">
<tt class="descname">iterativeCorrectByTrans</tt><big>(</big><em>names=None</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#experimentalBinnedData.iterativeCorrectByTrans"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.experimentalBinnedData.iterativeCorrectByTrans" title="Permalink to this definition">¶</a></dt>
<dd><p>performs iterative correction by trans data only, corrects cis also</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>names</strong> : list of str or None, optional</p>
<blockquote class="last">
<p>Keys of datasets to be corrected. By default, all are corrected.</p>
</blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.experimentalBinnedData.loadWigFile">
<tt class="descname">loadWigFile</tt><big>(</big><em>filename</em>, <em>label</em>, <em>control=None</em>, <em>wigFileType='Auto'</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#experimentalBinnedData.loadWigFile"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.experimentalBinnedData.loadWigFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Currently only fixed-step wig files and bigWig files are supported!!!
Import from fixedStep wig files is very fast, however is not the most reliable.</p>
<p>TODO: rewrite parser to support any resolution.</p>
<p>for VariableStep files use wigToBigWig utility to convert them to bigWig format first. 
To use it you will also need to have fetchChromSizes script.</p>
<p>Then you just run 
$bash fetchChromSizes hg18 &gt; hg18.chrom.sizes
$./wigToBigWig myWig.wig hg18.chrom.sizes myWig.bigWig</p>
<p>And you enjoy your favourite bigWig.</p>
<p>BigWig import is implemented using bx-python module.
It is normally very fast; however, it has a bug at low resolutions. 
I have an ugly workaround for it (chopping the quiery into many smaller pieces), but I hope 
that they actually fix this bug.</p>
<p>Anyway, check their repo on BitBucket, maybe they&#8217;ve fixed my issue # 39 :) 
<a class="reference external" href="https://bitbucket.org/james_taylor/bx-python/overview">https://bitbucket.org/james_taylor/bx-python/overview</a></p>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.experimentalBinnedData.loadErezEigenvector1MB">
<tt class="descname">loadErezEigenvector1MB</tt><big>(</big><em>erezFolder</em><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#experimentalBinnedData.loadErezEigenvector1MB"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.experimentalBinnedData.loadErezEigenvector1MB" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads Erez chromatin domain eigenvector for HindIII</p>
</dd></dl>

<dl class="method">
<dt id="hiclib.binnedData.experimentalBinnedData.loadTanayDomains">
<tt class="descname">loadTanayDomains</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/hiclib/binnedData.html#experimentalBinnedData.loadTanayDomains"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hiclib.binnedData.experimentalBinnedData.loadTanayDomains" title="Permalink to this definition">¶</a></dt>
<dd><p>domains, extracted from Tanay paper image</p>
</dd></dl>

</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Analysis of binned data</a><ul>
<li><a class="reference internal" href="#concepts">Concepts</a></li>
<li><a class="reference internal" href="#input-data">Input data</a></li>
<li><a class="reference internal" href="#variables">Variables</a></li>
<li><a class="reference internal" href="#hierarchy-of-filters">Hierarchy of filters</a></li>
</ul>
</li>
</ul>

  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/binneddata.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" size="18" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="np-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="index.html">mirnylab v0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2012, the Leonids.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.1.
    </div>
  </body>
</html>