

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>hiclib.binnedData &mdash; mirnylab v0.1 documentation</title>
    <link rel="stylesheet" href="../../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <link rel="top" title="mirnylab v0.1 documentation" href="../../index.html" />
    <link rel="up" title="Module code" href="../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../../np-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../../index.html">mirnylab v0.1 documentation</a> &raquo;</li>
          <li><a href="../index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <h1>Source code for hiclib.binnedData</h1><div class="highlight"><pre>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Binned data - analysis of HiC, binned to resolution.</span>

<span class="sd">Concepts</span>
<span class="sd">--------</span>

<span class="sd">class Binned Data allows low-level manipulation of  multiple HiC datasets, </span>
<span class="sd">binned to the same resolution from the same genome. </span>

<span class="sd">When working with multiple datasets, all the filters will be synchronized,</span>
<span class="sd">so only bins present in all datasets will be considered for the analysis. </span>
<span class="sd">Removal of bins from one dataset will remove them from the others. </span>
<span class="sd">E.g. removing 1% of bins with lowest # of count might remove more than 1% of total bins, </span>
<span class="sd">when working with 2 or more datasets.  </span>

<span class="sd">Class has significant knowledge about filters that have been applied. </span>
<span class="sd">If an essential filter was not applied, it will throw an exception; </span>
<span class="sd">if adviced filter is not applied, it will throw a warning. </span>
<span class="sd">Most of the methods have an optional &quot;force&quot; argument that will ignore dependensies.</span>

<span class="sd">We provide example scripts that show ideal protocols for certain types of the analysis, </span>
<span class="sd">but they don&#39;t cover the realm of all possible manipulations that can be performed with this class.</span>

<span class="sd">Input data</span>
<span class="sd">----------</span>

<span class="sd">method :py:func:`SimpleLoad &lt;binnedData.simpleLoad&gt;` may be used to load the data. </span>
<span class="sd">It automatically checks for possible genome length mismatch. </span>

<span class="sd">This method works best with h5dict files, created by fragmentHiC.</span>
<span class="sd">In this case you just need to supply the filename.  </span>

<span class="sd">It can also accept any dictionary-like object with the following keys, </span>
<span class="sd">where all but &quot;heatmap&quot; is optional.  </span>

<span class="sd">* [&quot;heatmap&quot;] : all-by-all heatmap</span>

<span class="sd">* [&quot;singles&quot;] : vector of SS reads, optional </span>

<span class="sd">* [&quot;frags&quot;] : number of rsites per bin, optional</span>

<span class="sd">* [&quot;resolution&quot;] : resolution </span>

<span class="sd">All information about the genome, including GC content and restriction sites, can be obtained from the Genome class. </span>

<span class="sd">Genomic tracks can be loaded using an automated parser that accepts bigWig files and fixed step wiggle files. </span>
<span class="sd">See documentation for :py:func:`experimentalBinnedData.loadWigFile` that describes exactly how the data is averaged and parsed.    </span>


<span class="sd">Variables</span>
<span class="sd">---------</span>

<span class="sd">self.dataDict - dictionary with heatmaps; keys are provided when loading the data.</span>

<span class="sd">self.singlesDict - dictionary with SS read vectors. Keys are the same.</span>

<span class="sd">self.fragsDict - dictionary with fragment density data  </span>

<span class="sd">self.trackDict - dictionary with genomic tracks, such as GC content. Custom tracks should be added here. </span>

<span class="sd">self.biasDict - dictionary with biases as calculated by iterative correction (incomplete)  </span>

<span class="sd">self.PCDict - dictionary with principal components of each datasets. Keys as in dataDict</span>

<span class="sd">self.EigEict - dictionary with eigenvectors for each dataset. Keys as in datadict. </span>


<span class="sd">Hierarchy of filters</span>
<span class="sd">--------------------</span>

<span class="sd">hierarchy</span>

<span class="sd">--------------------------------------------------------------- </span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">os</span> 
<span class="kn">from</span> <span class="nn">mirnylab</span> <span class="kn">import</span> <span class="n">numutils</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">from</span> <span class="nn">mirnylab.plotting</span> <span class="kn">import</span>  <span class="n">removeBorder</span> <span class="p">,</span> <span class="n">mat_img</span>
<span class="kn">from</span> <span class="nn">mirnylab.numutils</span> <span class="kn">import</span> <span class="n">PCA</span><span class="p">,</span> <span class="n">EIG</span><span class="p">,</span><span class="n">correct</span><span class="p">,</span> <span class="n">ultracorrectSymmetricWithVector</span><span class="p">,</span>\
    <span class="n">isInteger</span>
<span class="kn">from</span> <span class="nn">mirnylab.genome</span> <span class="kn">import</span> <span class="n">Genome</span> 
<span class="kn">import</span>  <span class="nn">numpy</span>
<span class="kn">from</span> <span class="nn">math</span> <span class="kn">import</span> <span class="n">exp</span>
<span class="kn">from</span> <span class="nn">mirnylab.h5dict</span> <span class="kn">import</span> <span class="n">h5dict</span>  
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">weave</span> 
<span class="kn">from</span> <span class="nn">scipy.stats.stats</span> <span class="kn">import</span> <span class="n">spearmanr</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">plt</span> 
<span class="kn">from</span> <span class="nn">mirnylab.systemutils</span> <span class="kn">import</span> <span class="n">setExceptionHook</span>


    
<div class="viewcode-block" id="binnedData"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData">[docs]</a><span class="k">class</span> <span class="nc">binnedData</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Base class to work with binned data, the most documented and robust part of the code. </span>
<span class="sd">    Further classes for other analysis are inherited from this class. </span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">resolution</span><span class="p">,</span><span class="n">genome</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        </span>
<span class="sd">        self.__init__ - initializes an empty dataset. </span>
<span class="sd">        </span>
<span class="sd">        This method sets up a Genome object and resolution. </span>
<span class="sd">        </span>
<span class="sd">        Genome object specifies genome version and inclusion/exclusion of sex chromosomes.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        resolution : int </span>
<span class="sd">            Resolution of all datasets</span>
<span class="sd">        genome : genome Folder or Genome object</span>
<span class="sd">        </span>
<span class="sd">        &quot;&quot;&quot;</span>        
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span> 
            <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">Genome</span><span class="p">(</span><span class="n">genomePath</span> <span class="o">=</span> <span class="n">genome</span><span class="p">,</span> <span class="n">readChrms</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;#&quot;</span><span class="p">,</span><span class="s">&quot;X&quot;</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">genome</span> 
            
            
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="n">Genome</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="n">resolution</span> <span class="o">!=</span> <span class="bp">None</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">resolution</span> <span class="o">=</span> <span class="n">resolution</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmLens</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">resolution</span> <span class="o">=</span> <span class="n">resolution</span>                    
        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">setResolution</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">resolution</span><span class="p">)</span>                        
        <span class="bp">self</span><span class="o">.</span><span class="n">_initChromosomes</span><span class="p">()</span>        
        <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">biasDict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">trackDict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">singlesDict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fragsDict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">PCDict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">EigDict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">eigEigenvalueDict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">PCAEigenvalueDict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">dicts</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">trackDict</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">biasDict</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">singlesDict</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragsDict</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">eigDicts</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">PCDict</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">EigDict</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_loadGC</span><span class="p">()</span> 
        <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span> <span class="o">=</span> <span class="p">{}</span>
        
    <span class="k">def</span> <span class="nf">_initChromosomes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="s">&quot;internal: loads mappings from the genome class based on resolution&quot;</span>        
        <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeStarts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmStartsBinCont</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">centromerePositions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">cntrMidsBinCont</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeEnds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmEndsBinCont</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">trackLength</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span>
                
        <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmCount</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmIdxBinCont</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">positionIndex</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">posBinCont</span>        
        <span class="bp">self</span><span class="o">.</span><span class="n">armIndex</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span> <span class="o">*</span> <span class="mi">2</span> <span class="o">+</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">positionIndex</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">cntrMids</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">],</span><span class="nb">int</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_giveMask</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="s">&quot;Returns index of all bins with non-zero read counts&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">values</span><span class="p">()[</span><span class="mi">0</span><span class="p">]),</span><span class="n">numpy</span><span class="o">.</span><span class="n">bool</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">data</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">datasum</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
            <span class="n">datamask</span> <span class="o">=</span> <span class="n">datasum</span> <span class="o">&gt;</span> <span class="mi">0</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mask</span> <span class="o">*=</span> <span class="n">datamask</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span>
    
    <span class="k">def</span> <span class="nf">_giveMask2D</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="s">&quot;Returns outer product of _giveMask with itself, i.e. bins with possibly non-zero counts&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_giveMask</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mask2D</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span><span class="p">[:,</span><span class="bp">None</span><span class="p">]</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask</span><span class="p">[</span><span class="bp">None</span><span class="p">,:]</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask2D</span>   

    <span class="k">def</span> <span class="nf">_loadGC</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>        
        <span class="s">&quot;loads GC content at given resolution&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">trackDict</span><span class="p">[</span><span class="s">&quot;GC&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">GCBin</span><span class="p">)</span>
        
    <span class="k">def</span> <span class="nf">_checkItertiveCorrectionError</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="s">&quot;internal method for checking if iterative correction might be bad to apply&quot;</span>
        <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>           
             
            <span class="k">if</span> <span class="n">isInteger</span><span class="p">(</span><span class="n">value</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">()</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>
                <span class="n">s</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">value</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
                <span class="n">sums</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">s</span><span class="p">[</span><span class="n">s</span><span class="o">!=</span><span class="mi">0</span><span class="p">])</span>
                <span class="k">if</span> <span class="n">sums</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">100</span><span class="p">:</span>
                    <span class="n">error</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="mf">100.</span> <span class="o">/</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">sums</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
                    <span class="n">message1</span> <span class="o">=</span>  <span class="s">&quot;Lowest 5 sums of an array rows are: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">sums</span><span class="p">[:</span><span class="mi">5</span><span class="p">])</span>                    
                    <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="si">%s</span><span class="se">\n</span><span class="s">Iterative correction will lead to about </span><span class="si">%d</span><span class="s"> </span><span class="si">%%</span><span class="s"> relative error for certain columns&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">message1</span><span class="p">,</span><span class="n">error</span><span class="p">))</span>
                    
                    <span class="k">if</span> <span class="n">sums</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">5</span><span class="p">:</span> 
                        <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Iterative correction is very dangerous. Use force=true to override.&quot;</span><span class="p">)</span> 
                          
            <span class="k">else</span><span class="p">:</span>
                <span class="n">s</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">value</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
                <span class="n">sums</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">s</span><span class="p">[</span><span class="n">s</span><span class="o">!=</span><span class="mi">0</span><span class="p">])</span>
                <span class="k">if</span> <span class="n">sums</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="nb">min</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">value</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span><span class="p">):</span>
                    <span class="n">error</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="mf">100.</span> <span class="o">/</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">sums</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
                    <span class="k">print</span> <span class="s">&quot;Got floating-point array for correction. Rows with 5 least entrees are:&quot;</span><span class="p">,</span><span class="n">sums</span><span class="p">[:</span><span class="mi">5</span><span class="p">]</span>
                    <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">Iterative correction might lead to about </span><span class="si">%d</span><span class="s"> </span><span class="si">%%</span><span class="s"> relative error for certain columns&quot;</span> <span class="o">%</span> <span class="n">error</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">sums</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">4</span><span class="p">:</span> 
                        <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Iterative correction is very dangerous. Use force=true to override.&quot;</span><span class="p">)</span>
                    
    <span class="k">def</span> <span class="nf">_checkAppliedOperations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">neededKeys</span><span class="o">=</span><span class="p">[],</span><span class="n">advicedKeys</span><span class="o">=</span><span class="p">[],</span><span class="n">excludedKeys</span><span class="o">=</span><span class="p">[]):</span>
        <span class="s">&quot;Internal method to check if all needed operations were applied&quot;</span>
        
        <span class="k">if</span> <span class="p">(</span><span class="bp">True</span> <span class="ow">in</span> <span class="p">[</span><span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">excludedKeys</span><span class="p">]):</span>
            <span class="k">print</span> <span class="s">&quot;Operations that are not allowed:&quot;</span><span class="p">,</span> <span class="n">excludedKeys</span>
            <span class="k">print</span> <span class="s">&quot;applied operations: &quot;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span>
            <span class="k">print</span> <span class="s">&quot;use &#39;force = True&#39; to override this message&quot;</span> 
            <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Prohibited filter was applied&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="p">(</span><span class="bp">False</span> <span class="ow">in</span> <span class="p">[</span><span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">neededKeys</span><span class="p">]):</span> 
            <span class="k">print</span> <span class="s">&quot;needed operations:&quot;</span><span class="p">,</span><span class="n">neededKeys</span>
            <span class="k">print</span> <span class="s">&quot;applied operations:&quot;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span>
            <span class="k">print</span> <span class="s">&quot;use &#39;force = True&#39; to override this message&quot;</span> 
            <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Critical filter not applied&quot;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="p">(</span><span class="bp">False</span> <span class="ow">in</span> <span class="p">[</span><span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">advicedKeys</span><span class="p">]):</span>
            <span class="k">print</span> <span class="s">&quot;Adviced operations:&quot;</span><span class="p">,</span><span class="n">advicedKeys</span>
            <span class="k">print</span> <span class="s">&quot;Applied operations:&quot;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">Not all adviced filters applied&quot;</span><span class="p">)</span>                        
        
    
<div class="viewcode-block" id="binnedData.simpleLoad"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.simpleLoad">[docs]</a>    <span class="k">def</span> <span class="nf">simpleLoad</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">in_data</span><span class="p">,</span><span class="n">name</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Loads data from h5dict file or dict-like object</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        </span>
<span class="sd">        in_data : str or dict-like</span>
<span class="sd">            h5dict filename or dictionary-like object with input data            </span>
<span class="sd">        name : str</span>
<span class="sd">            Key of the dataset in self.dataDict</span>
<span class="sd">        </span>
<span class="sd">        &quot;&quot;&quot;</span>        
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">in_data</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">in_data</span><span class="p">)</span> <span class="o">==</span> <span class="bp">False</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s">&quot;HDF5 dict do not exist, </span><span class="si">%s</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">in_data</span><span class="p">)</span> 
            <span class="n">alldata</span> <span class="o">=</span> <span class="n">h5dict</span><span class="p">(</span><span class="n">in_data</span><span class="p">,</span><span class="n">mode</span> <span class="o">=</span> <span class="s">&quot;r&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">alldata</span> <span class="o">=</span> <span class="n">in_data</span>
                             
        <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">alldata</span><span class="p">[</span><span class="s">&quot;heatmap&quot;</span><span class="p">]</span>        
        <span class="k">try</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">singlesDict</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">alldata</span><span class="p">[</span><span class="s">&quot;singles&quot;</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;No SS reads found&quot;</span>
        <span class="k">try</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragsDict</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">alldata</span><span class="p">[</span><span class="s">&quot;frags&quot;</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span><span class="k">pass</span> 
        
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">resolution</span> <span class="o">!=</span> <span class="n">alldata</span><span class="p">[</span><span class="s">&quot;resolution&quot;</span><span class="p">]:</span>
            <span class="k">print</span> <span class="s">&quot;resolution mismatch!!!&quot;</span>
            <span class="k">print</span> <span class="s">&quot;--------------&gt; Bye &lt;-------------&quot;</span>
            <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Resolution mismatch! &quot;</span><span class="p">)</span> 
        
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">alldata</span><span class="p">[</span><span class="s">&quot;heatmap&quot;</span><span class="p">]):</span>              
            <span class="k">print</span> <span class="s">&quot;Genome length mismatch!!!&quot;</span>
            <span class="k">print</span> <span class="s">&quot;source genome&quot;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">alldata</span><span class="p">[</span><span class="s">&quot;heatmap&quot;</span><span class="p">])</span>
            <span class="k">print</span> <span class="s">&quot;our genome&quot;</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span>
            <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Genome size mismatch! &quot;</span><span class="p">)</span>

              
        
        
  
    </div>
<div class="viewcode-block" id="binnedData.removeDiagonal"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.removeDiagonal">[docs]</a>    <span class="k">def</span> <span class="nf">removeDiagonal</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">m</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Removes all bins on a diagonal, and bins that are up to m away from the diagonal, including m. </span>
<span class="sd">        By default, removes all bins touching the diagonal.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        m : int, optional </span>
<span class="sd">            Number of bins to remove</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="mf">1.</span>             
            <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
            <span class="n">N</span>   <span class="c">#Eclipse warning remover </span>
            <span class="n">code</span> <span class="o">=</span> <span class="s">r&quot;&quot;&quot;</span>
<span class="s">            #line 841 &quot;binary_search.py&quot;</span>
<span class="s">            using namespace std; </span>
<span class="s">            for (int i = 0; i &lt; N; i++)    </span>
<span class="s">            {    </span>
<span class="s">                for (int j = max(i-m,0); j&lt;min(i+m+1,N); j++)</span>
<span class="s">                {</span>
<span class="s">                    data[i*N + j] = 0;</span>
<span class="s">                }</span>
<span class="s">            } </span>
<span class="s">            &quot;&quot;&quot;</span>
            <span class="n">support</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">            #include &lt;math.h&gt;</span>
<span class="s">            &quot;&quot;&quot;</span>
            <span class="n">weave</span><span class="o">.</span><span class="n">inline</span><span class="p">(</span><span class="n">code</span><span class="p">,</span> <span class="p">[</span><span class="s">&#39;m&#39;</span><span class="p">,</span><span class="s">&#39;data&#39;</span><span class="p">,</span><span class="s">&quot;N&quot;</span><span class="p">],</span> <span class="n">extra_compile_args</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;-march=native -malign-double -O3&#39;</span><span class="p">],</span><span class="n">support_code</span> <span class="o">=</span><span class="n">support</span> <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span><span class="p">[</span><span class="s">&quot;RemovedDiagonal&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">True</span>        
        
        
    </div>
<div class="viewcode-block" id="binnedData.removeStandalone"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.removeStandalone">[docs]</a>    <span class="k">def</span> <span class="nf">removeStandalone</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">offset</span> <span class="o">=</span> <span class="mi">3</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;removes standalone bins (groups of less-than-offset bins)</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        offset : int </span>
<span class="sd">            Maximum length of group of bins to be removed</span>
<span class="sd">        &quot;&quot;&quot;</span>                
        <span class="n">diffs</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">r_</span><span class="p">[</span><span class="bp">False</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_giveMask</span><span class="p">(),</span> <span class="bp">False</span><span class="p">],</span><span class="nb">int</span><span class="p">))</span>
        <span class="n">begins</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">diffs</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> 
        <span class="n">ends</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">diffs</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">beginsmask</span> <span class="o">=</span> <span class="p">(</span><span class="n">ends</span> <span class="o">-</span> <span class="n">begins</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">offset</span>
        <span class="n">newbegins</span> <span class="o">=</span> <span class="n">begins</span><span class="p">[</span><span class="n">beginsmask</span><span class="p">]</span>
        <span class="n">newends</span> <span class="o">=</span> <span class="n">ends</span><span class="p">[</span><span class="n">beginsmask</span><span class="p">]</span>        
        <span class="k">print</span> <span class="s">&quot;removing </span><span class="si">%d</span><span class="s"> standalone bins&quot;</span><span class="o">%</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">newends</span> <span class="o">-</span> <span class="n">newbegins</span><span class="p">)</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_giveMask</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">newbegins</span><span class="p">)):</span> <span class="n">mask</span><span class="p">[</span><span class="n">newbegins</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span><span class="n">newends</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span> <span class="o">=</span> <span class="bp">False</span>
        <span class="n">mask2D</span> <span class="o">=</span> <span class="n">mask</span><span class="p">[:,</span><span class="bp">None</span><span class="p">]</span> <span class="o">*</span> <span class="n">mask</span><span class="p">[</span><span class="bp">None</span><span class="p">,:]</span>        
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">values</span><span class="p">():</span> <span class="n">i</span><span class="p">[</span><span class="n">mask2D</span> <span class="o">==</span> <span class="bp">False</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span><span class="p">[</span><span class="s">&quot;RemovedStandalone&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">True</span>
        </div>
<div class="viewcode-block" id="binnedData.removePoorRegions"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.removePoorRegions">[docs]</a>    <span class="k">def</span> <span class="nf">removePoorRegions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">names</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span> <span class="n">cutoff</span> <span class="o">=</span> <span class="mi">2</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;removes cutoff persent of bins with least counts</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        names : list of str</span>
<span class="sd">            List of datasets to perform the filter. All by default. </span>
<span class="sd">        cutoff : int, 0&lt;cutoff&lt;100</span>
<span class="sd">            Percent of lowest-counts bins to be removed         </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">statmask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">values</span><span class="p">()[</span><span class="mi">0</span><span class="p">]),</span><span class="n">numpy</span><span class="o">.</span><span class="n">bool</span><span class="p">)</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">values</span><span class="p">()[</span><span class="mi">0</span><span class="p">]),</span><span class="n">numpy</span><span class="o">.</span><span class="n">bool</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">names</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span> <span class="n">names</span> <span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>  
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">names</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">datasum</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>            
            <span class="n">datamask</span> <span class="o">=</span> <span class="n">datasum</span> <span class="o">&gt;</span> <span class="mi">0</span>
            <span class="n">mask</span> <span class="o">*=</span> <span class="n">datamask</span>  
            <span class="n">countsum</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>  
            <span class="n">newmask</span> <span class="o">=</span> <span class="n">countsum</span> <span class="o">&gt;=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">countsum</span><span class="p">[</span><span class="n">datamask</span><span class="p">],</span><span class="n">cutoff</span><span class="p">)</span>
            <span class="n">mask</span> <span class="o">*=</span> <span class="n">newmask</span>  
            <span class="n">statmask</span> <span class="p">[(</span><span class="n">newmask</span> <span class="o">==</span> <span class="bp">False</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">datamask</span> <span class="o">==</span> <span class="bp">True</span><span class="p">)]</span> <span class="o">=</span> <span class="bp">True</span>
        <span class="k">print</span> <span class="s">&quot;removed </span><span class="si">%d</span><span class="s"> poor bins&quot;</span><span class="p">,</span> <span class="n">statmask</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> 
        <span class="n">mask2D</span> <span class="o">=</span> <span class="n">mask</span><span class="p">[:,</span><span class="bp">None</span><span class="p">]</span> <span class="o">*</span> <span class="n">mask</span><span class="p">[</span><span class="bp">None</span><span class="p">,:]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">values</span><span class="p">():</span> <span class="n">i</span><span class="p">[</span><span class="n">mask2D</span> <span class="o">==</span> <span class="bp">False</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span><span class="p">[</span><span class="s">&quot;RemovedPoor&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">True</span>
              </div>
<div class="viewcode-block" id="binnedData.truncTrans"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.truncTrans">[docs]</a>    <span class="k">def</span> <span class="nf">truncTrans</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">high</span> <span class="o">=</span> <span class="mf">0.0005</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Truncates trans contacts to remove blowouts</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        high : float, 0&lt;high&lt;1, optional </span>
<span class="sd">            Fraction of top trans interactions to be removed</span>
<span class="sd">        &quot;&quot;&quot;</span>                
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">transmask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[:,</span><span class="bp">None</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[</span><span class="bp">None</span><span class="p">,:]</span>
            <span class="n">lim</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">transmask</span><span class="p">],</span><span class="mf">100.</span><span class="o">*</span><span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">high</span><span class="p">))</span>
            <span class="k">print</span> <span class="s">&quot;dataset </span><span class="si">%s</span><span class="s"> truncated at </span><span class="si">%lf</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="n">lim</span><span class="p">)</span> 
            <span class="n">tdata</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">transmask</span><span class="p">]</span>
            <span class="n">tdata</span><span class="p">[</span><span class="n">tdata</span> <span class="o">&gt;</span> <span class="n">lim</span><span class="p">]</span> <span class="o">=</span> <span class="n">lim</span>            
            <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">transmask</span><span class="p">]</span> <span class="o">=</span> <span class="n">tdata</span>            
        <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span><span class="p">[</span><span class="s">&quot;TruncedTrans&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">True</span>                        
        </div>
<div class="viewcode-block" id="binnedData.removeCis"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.removeCis">[docs]</a>    <span class="k">def</span> <span class="nf">removeCis</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="s">&quot;sets to zero all cis contacts&quot;</span>
        
        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[:,</span><span class="bp">None</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[</span><span class="bp">None</span><span class="p">,:]</span>         
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>                
            <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">mask</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>   
        <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span><span class="p">[</span><span class="s">&quot;RemovedCis&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">True</span>           
        <span class="k">print</span><span class="p">(</span><span class="s">&quot;All cis counts set to zero&quot;</span><span class="p">)</span>
            </div>
<div class="viewcode-block" id="binnedData.fakeCisOnce"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.fakeCisOnce">[docs]</a>    <span class="k">def</span> <span class="nf">fakeCisOnce</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">mask</span> <span class="o">=</span> <span class="s">&quot;CisCounts&quot;</span><span class="p">,</span> <span class="n">silent</span> <span class="o">=</span> <span class="bp">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Used to fake cis counts or any other region with random trans counts. </span>
<span class="sd">        If extra mask is supplied, it is used instead of cis counts. </span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        mask : NxN boolean array or &quot;CisCounts&quot; </span>
<span class="sd">            Mask of elements to be faked. </span>
<span class="sd">            If set to &quot;CisCounts&quot;, cis counts will be faked</span>
<span class="sd">        silent : bool</span>
<span class="sd">            Do not print anything</span>
<span class="sd">             </span>
<span class="sd">        </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">silent</span> <span class="o">==</span> <span class="bp">False</span><span class="p">:</span> <span class="k">print</span><span class="p">(</span><span class="s">&quot;All cis counts are substituted with matching trans count&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>                         
            <span class="n">data</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">key</span><span class="p">],</span><span class="n">order</span> <span class="o">=</span> <span class="s">&quot;C&quot;</span><span class="p">,</span><span class="n">dtype</span> <span class="o">=</span> <span class="nb">float</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">mask</span> <span class="o">==</span> <span class="s">&quot;CisCounts&quot;</span><span class="p">:</span> 
                <span class="n">_mask</span> <span class="o">=</span>  <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[:,</span><span class="bp">None</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[</span><span class="bp">None</span><span class="p">,:],</span><span class="nb">int</span><span class="p">,</span><span class="n">order</span> <span class="o">=</span> <span class="s">&quot;C&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">assert</span> <span class="n">mask</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">values</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>  <span class="c">#check that mask has correct shape </span>
                <span class="n">_mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span><span class="n">dtype</span> <span class="o">=</span> <span class="nb">int</span><span class="p">,</span> <span class="n">order</span> <span class="o">=</span> <span class="s">&quot;C&quot;</span><span class="p">)</span>
                <span class="n">_mask</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[:,</span><span class="bp">None</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[</span><span class="bp">None</span><span class="p">,:]]</span> <span class="o">=</span> <span class="mi">2</span>   <span class="c">#do not fake with cis counts                                             </span>
            <span class="n">s</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">))</span> <span class="o">&lt;=</span> <span class="mf">1e-10</span> 
            <span class="n">_mask</span><span class="p">[:,</span><span class="n">s</span><span class="p">]</span><span class="o">=</span> <span class="mi">2</span>
            <span class="n">_mask</span><span class="p">[</span><span class="n">s</span><span class="p">,:]</span> <span class="o">=</span> <span class="mi">2</span>              
            <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
            <span class="n">N</span>
            <span class="n">code</span> <span class="o">=</span> <span class="s">r&quot;&quot;&quot;</span>
<span class="s">            #line 310 &quot;binnedData.py&quot;</span>
<span class="s">            using namespace std; </span>
<span class="s">            for (int i = 0; i &lt; N; i++)    </span>
<span class="s">            {    </span>
<span class="s">                for (int j = i; j&lt;N; j++)</span>
<span class="s">                {</span>
<span class="s">                    if (_mask[i* N + j] == 1)                    </span>
<span class="s">                    {</span>
<span class="s">                    while (true) </span>
<span class="s">                        {</span>
<span class="s">                        int r = rand() % 2;                         </span>
<span class="s">                        if (r == 0)</span>
<span class="s">                            {                            </span>
<span class="s">                            int s = rand() % N;</span>
<span class="s">                            if (_mask[i * N + s] == 0)</span>
<span class="s">                                {                                </span>
<span class="s">                                data[i * N + j] = data[i * N + s];</span>
<span class="s">                                data[j * N + i] = data[i * N + s];</span>
<span class="s">                                break;</span>
<span class="s">                                 </span>
<span class="s">                                }</span>
<span class="s">                            }</span>
<span class="s">                        else</span>
<span class="s">                            {</span>
<span class="s">                            int s = rand() % N;</span>
<span class="s">                            if (_mask[j * N + s] == 0)</span>
<span class="s">                                {</span>
<span class="s">                                data[i * N + j] = data[j * N + s];</span>
<span class="s">                                data[j * N + i] = data[j * N + s]; </span>
<span class="s">                                break;</span>
<span class="s">                                }                            </span>
<span class="s">                            }                        </span>
<span class="s">                        }</span>
<span class="s">                    }</span>
<span class="s">                }</span>
<span class="s">            } </span>
<span class="s">            &quot;&quot;&quot;</span>
            <span class="n">support</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">            #include &lt;math.h&gt;</span>
<span class="s">            &quot;&quot;&quot;</span>
            <span class="n">weave</span><span class="o">.</span><span class="n">inline</span><span class="p">(</span><span class="n">code</span><span class="p">,</span> <span class="p">[</span><span class="s">&#39;_mask&#39;</span><span class="p">,</span><span class="s">&#39;data&#39;</span><span class="p">,</span><span class="s">&quot;N&quot;</span><span class="p">],</span> <span class="n">extra_compile_args</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;-march=native -malign-double -O3&#39;</span><span class="p">],</span><span class="n">support_code</span> <span class="o">=</span><span class="n">support</span> <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span><span class="p">[</span><span class="s">&quot;RemovedCis&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">True</span> 
            <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span><span class="p">[</span><span class="s">&quot;FakedCis&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">True</span> 
            <span class="c">#mat_img(data)</span>
</div>
<div class="viewcode-block" id="binnedData.fakeCis"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.fakeCis">[docs]</a>    <span class="k">def</span> <span class="nf">fakeCis</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;This method fakes cis contacts in an interative way</span>
<span class="sd">        It is done to achieve faking cis contacts that is independent of normalization of the data. </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">removeCis</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">iterativeCorrectWithoutSS</span><span class="p">(</span><span class="n">M</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fakeCisOnce</span><span class="p">(</span><span class="n">silent</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">iterativeCorrectWithoutSS</span><span class="p">(</span><span class="n">M</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fakeCisOnce</span><span class="p">(</span><span class="n">silent</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">iterativeCorrectWithoutSS</span><span class="p">(</span><span class="n">M</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span> 
        <span class="k">print</span><span class="p">(</span><span class="s">&quot;All cis counts are substituted with faked counts&quot;</span><span class="p">)</span>
        <span class="k">print</span><span class="p">(</span><span class="s">&quot;Data is iteratively corrected as a part of faking cis counts&quot;</span><span class="p">)</span>        
        </div>
<div class="viewcode-block" id="binnedData.fakeTranslocations"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.fakeTranslocations">[docs]</a>    <span class="k">def</span> <span class="nf">fakeTranslocations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">translocationRegions</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method fakes reads corresponding to a translocation.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        </span>
<span class="sd">        translocationRegions: list of tuples</span>
<span class="sd">            List of tuples (chr1,start1,end1,chr2,start2,end2), masking a high-count region around visible translocation.</span>
<span class="sd">            If chromosome end is None, it is treated as length of chromosome. So, use (chr1,0,None,chr2,0,None) to remove inter-chromosomal interaction entirely.          </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_checkAppliedOperations</span><span class="p">(</span><span class="n">excludedKeys</span> <span class="o">=</span> <span class="s">&quot;RemovedZeros&quot;</span><span class="p">)</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span><span class="p">),</span><span class="nb">int</span><span class="p">)</span>
        <span class="n">resolution</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">resolution</span>           
        
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">translocationRegions</span><span class="p">:</span>
            <span class="n">st1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmStartsBinCont</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
            <span class="n">st2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmStartsBinCont</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">3</span><span class="p">]]</span>
            <span class="n">beg1</span> <span class="o">=</span> <span class="n">st1</span> <span class="o">+</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">/</span><span class="n">resolution</span>
            <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">None</span><span class="p">:</span> 
                <span class="n">end1</span> <span class="o">=</span> <span class="n">st1</span> <span class="o">+</span> <span class="n">i</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">/</span><span class="n">resolution</span><span class="o">+</span><span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">end1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmEndsBinCont</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
            <span class="n">beg2</span> <span class="o">=</span> <span class="n">st2</span> <span class="o">+</span> <span class="n">i</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">/</span><span class="n">resolution</span>
            <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">None</span><span class="p">:</span> 
                <span class="n">end2</span> <span class="o">=</span> <span class="n">st2</span> <span class="o">+</span> <span class="n">i</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">/</span><span class="n">resolution</span><span class="o">+</span><span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">end2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmEndsBinCont</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">3</span><span class="p">]]</span>

            <span class="n">mask</span><span class="p">[</span><span class="n">beg1</span><span class="p">:</span><span class="n">end1</span><span class="p">,</span><span class="n">beg2</span><span class="p">:</span><span class="n">end2</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">mask</span><span class="p">[</span><span class="n">beg2</span><span class="p">:</span><span class="n">end2</span><span class="p">,</span><span class="n">beg1</span><span class="p">:</span><span class="n">end1</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fakeCisOnce</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span>
         
            
            

</div>
<div class="viewcode-block" id="binnedData.correct"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.correct">[docs]</a>    <span class="k">def</span> <span class="nf">correct</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">names</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;performs single correction without SS</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        names : list of str or None</span>
<span class="sd">            Keys of datasets to be corrected. If none, all are corrected. </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">iterativeCorrectWithoutSS</span><span class="p">(</span><span class="n">names</span><span class="p">,</span> <span class="n">M</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                </div>
<div class="viewcode-block" id="binnedData.iterativeCorrectWithoutSS"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.iterativeCorrectWithoutSS">[docs]</a>    <span class="k">def</span> <span class="nf">iterativeCorrectWithoutSS</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">names</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span><span class="n">M</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span><span class="n">force</span> <span class="o">=</span> <span class="bp">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;performs iterative correction without SS</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        names : list of str or None, optional</span>
<span class="sd">            Keys of datasets to be corrected. By default, all are corrected.</span>
<span class="sd">        M : int, optional</span>
<span class="sd">            Number of iterations to perform. </span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        
        
        
        <span class="k">if</span> <span class="n">force</span> <span class="o">==</span> <span class="bp">False</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_checkItertiveCorrectionError</span><span class="p">()</span>             
            <span class="bp">self</span><span class="o">.</span><span class="n">_checkAppliedOperations</span><span class="p">(</span><span class="n">advicedKeys</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;RemovedDiagonal&quot;</span><span class="p">,</span><span class="s">&quot;RemovedPoor&quot;</span><span class="p">])</span>        
                    
        <span class="k">if</span> <span class="n">names</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span> <span class="n">names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">names</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">ultracorrectSymmetricWithVector</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="n">M</span><span class="o">=</span><span class="n">M</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>         
          
        <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span><span class="p">[</span><span class="s">&quot;Corrected&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">True</span>
</div>
<div class="viewcode-block" id="binnedData.iterativeCorrectWithSS"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.iterativeCorrectWithSS">[docs]</a>    <span class="k">def</span> <span class="nf">iterativeCorrectWithSS</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">names</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span><span class="n">M</span> <span class="o">=</span> <span class="mi">55</span><span class="p">,</span><span class="n">force</span> <span class="o">=</span> <span class="bp">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;performs iterative correction with SS</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        names : list of str or None, optional</span>
<span class="sd">            Keys of datasets to be corrected. By default, all are corrected.</span>
<span class="sd">        M : int, optional</span>
<span class="sd">            Number of iterations to perform. </span>
<span class="sd">        force : bool, optional </span>
<span class="sd">            Force current operation </span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="k">if</span> <span class="n">force</span> <span class="o">==</span> <span class="bp">False</span><span class="p">:</span> 
            <span class="bp">self</span><span class="o">.</span><span class="n">_checkAppliedOperations</span><span class="p">(</span><span class="n">advicedKeys</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;RemovedDiagonal&quot;</span><span class="p">,</span><span class="s">&quot;RemovedPoor&quot;</span><span class="p">],</span> 
                                        <span class="n">excludedKeys</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;Corrected&quot;</span><span class="p">,</span><span class="s">&quot;RemovedCis&quot;</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_checkItertiveCorrectionError</span><span class="p">()</span>
                
        <span class="k">if</span> <span class="n">names</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span> <span class="n">names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">names</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">vec</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">singlesDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">ndata</span><span class="p">,</span><span class="n">nvec</span> <span class="o">=</span> <span class="n">ultracorrectSymmetricWithVector</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">vec</span><span class="p">,</span><span class="n">M</span><span class="o">=</span><span class="n">M</span><span class="p">)</span>                         
            <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">ndata</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">singlesDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">nvec</span>
            <span class="n">vec</span><span class="p">[</span><span class="n">nvec</span><span class="o">==</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">nvec</span><span class="p">[</span><span class="n">nvec</span><span class="o">==</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">biasDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">vec</span> <span class="o">/</span> <span class="n">nvec</span><span class="p">)</span>

        
        <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span><span class="p">[</span><span class="s">&quot;Corrected&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">True</span>
        </div>
<div class="viewcode-block" id="binnedData.removeChromosome"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.removeChromosome">[docs]</a>    <span class="k">def</span> <span class="nf">removeChromosome</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">chromNum</span><span class="p">):</span>
        <span class="s">&quot;removes certain chromosome from all tracks and heatmaps, setting all values to zero &quot;</span>
        <span class="n">beg</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmStartsBinCont</span><span class="p">[</span><span class="n">chromNum</span><span class="p">]</span>
        <span class="n">end</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmEndsBinCont</span><span class="p">[</span><span class="n">chromNum</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">i</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">i</span><span class="p">[:,</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span> 
        
        <span class="k">for</span> <span class="n">mydict</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dicts</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">mydict</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="n">value</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span> 
        
        <span class="k">for</span> <span class="n">mydict</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">eigDicts</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">mydict</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="n">value</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span> 
                

</div>
<div class="viewcode-block" id="binnedData.removeZeros"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.removeZeros">[docs]</a>    <span class="k">def</span> <span class="nf">removeZeros</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">zerosMask</span> <span class="o">=</span> <span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;removes bins with zero counts</span>
<span class="sd">        keeps chromosome starts, ends, etc. consistent&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">zerosMask</span> <span class="o">!=</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">s</span> <span class="o">=</span> <span class="n">zerosMask</span>
        <span class="k">else</span><span class="p">:</span>          
            <span class="n">s</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_giveMask2D</span><span class="p">(),</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="n">s</span> <span class="o">*=</span> <span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">indices</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">s</span><span class="p">),</span><span class="nb">int</span><span class="p">)</span>
        <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span> 
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">indices</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">s</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>
                <span class="n">indices</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">count</span>
                <span class="n">count</span> <span class="o">+=</span><span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span> 
                <span class="n">indices</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">count</span>
        <span class="n">indices</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">r_</span><span class="p">[</span><span class="n">indices</span><span class="p">,</span><span class="n">indices</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>  
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">a</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">b</span> <span class="o">=</span> <span class="n">a</span><span class="p">[:,</span><span class="n">s</span><span class="p">]</span>
            <span class="n">c</span> <span class="o">=</span> <span class="n">b</span><span class="p">[</span><span class="n">s</span><span class="p">,:]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">c</span>            
        
        <span class="k">for</span> <span class="n">mydict</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dicts</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">mydict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">mydict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">mydict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="n">s</span><span class="p">]</span>

        
        <span class="k">for</span> <span class="n">mydict</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">eigDicts</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">mydict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                
                <span class="n">mydict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">mydict</span><span class="p">[</span><span class="n">key</span><span class="p">][:,</span><span class="n">s</span><span class="p">]</span>  
        
        <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[</span><span class="n">s</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">armIndex</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">armIndex</span><span class="p">[</span><span class="n">s</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeEnds</span> <span class="o">=</span> <span class="n">indices</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeEnds</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeStarts</span> <span class="o">=</span> <span class="n">indices</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">centromerePositions</span> <span class="o">=</span> <span class="n">indices</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">centromerePositions</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">removeZerosMask</span> <span class="o">=</span> <span class="n">s</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;RemovedZeros&quot;</span><span class="p">,</span><span class="bp">False</span><span class="p">)</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">You&#39;re removing zeros twice. You can&#39;t restore zeros now!&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span><span class="p">[</span><span class="s">&quot;RemovedZeros&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">True</span>  
        <span class="k">return</span> <span class="n">s</span> 
</div>
<div class="viewcode-block" id="binnedData.restoreZeros"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.restoreZeros">[docs]</a>    <span class="k">def</span> <span class="nf">restoreZeros</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">NAN</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Restores zeros that were removed by removeZeros command. </span>
<span class="sd">        </span>
<span class="sd">        .. warning:: You can restore zeros only if you used removeZeros once.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        value : number-like, optional. </span>
<span class="sd">            Value to fill in missing regions. By default, NAN. </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="s">&quot;removeZerosMask&quot;</span><span class="p">):</span> <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Zeros have not been removed!&quot;</span><span class="p">)</span>        
        
        <span class="n">s</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">removeZerosMask</span>
        <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">a</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">N</span><span class="p">,</span><span class="n">N</span><span class="p">),</span><span class="n">dtype</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span> <span class="o">*</span> <span class="n">value</span> 
            <span class="n">tmp</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">N</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">a</span><span class="p">)),</span><span class="n">dtype</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span> <span class="o">*</span> <span class="n">value</span> 
            <span class="n">tmp</span><span class="p">[</span><span class="n">s</span><span class="p">,:]</span> <span class="o">=</span> <span class="n">a</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">][:,</span><span class="n">s</span><span class="p">]</span> <span class="o">=</span> <span class="n">tmp</span>                     
        <span class="k">for</span> <span class="n">mydict</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dicts</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">mydict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">a</span> <span class="o">=</span> <span class="n">mydict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
                <span class="n">mydict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">N</span><span class="p">,</span><span class="n">dtype</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span> <span class="o">*</span> <span class="n">value</span> 
                <span class="n">mydict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="n">s</span><span class="p">]</span> <span class="o">=</span> <span class="n">a</span>
                        
        <span class="k">for</span> <span class="n">mydict</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">eigDicts</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">mydict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">a</span> <span class="o">=</span> <span class="n">mydict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
                <span class="n">mydict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">a</span><span class="p">),</span><span class="n">N</span><span class="p">),</span><span class="n">dtype</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span> <span class="o">*</span> <span class="n">value</span> 
                <span class="n">mydict</span><span class="p">[</span><span class="n">key</span><span class="p">][:,</span><span class="n">s</span><span class="p">]</span> <span class="o">=</span> <span class="n">a</span>
                                        
        <span class="bp">self</span><span class="o">.</span><span class="n">_initChromosomes</span><span class="p">()</span>     
        <span class="bp">self</span><span class="o">.</span><span class="n">appliedOperations</span><span class="p">[</span><span class="s">&quot;RemovedZeros&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">False</span>           
        
    
            </div>
<div class="viewcode-block" id="binnedData.doPCA"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.doPCA">[docs]</a>    <span class="k">def</span> <span class="nf">doPCA</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">force</span> <span class="o">=</span> <span class="bp">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;performs PCA on the data</span>
<span class="sd">        creates dictionary self.PCADict with results</span>
<span class="sd">        Last column of PC matrix is first PC, second to last - second, etc. </span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Dictionary of principal component matrices for different datasets</span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        
        <span class="n">neededKeys</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;RemovedZeros&quot;</span><span class="p">,</span><span class="s">&quot;Corrected&quot;</span><span class="p">,</span><span class="s">&quot;FakedCis&quot;</span><span class="p">]</span>
        <span class="n">advicedKeys</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;TruncedTrans&quot;</span><span class="p">,</span><span class="s">&quot;RemovedPoor&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">force</span> <span class="o">==</span> <span class="bp">False</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_checkAppliedOperations</span><span class="p">(</span><span class="n">neededKeys</span><span class="p">,</span> <span class="n">advicedKeys</span><span class="p">)</span>
        
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">currentPCA</span><span class="p">,</span><span class="n">eigenvalues</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">PCAEigenvalueDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">eigenvalues</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">currentPCA</span><span class="p">)):</span>
                <span class="k">if</span> <span class="n">spearmanr</span><span class="p">(</span><span class="n">currentPCA</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="bp">self</span><span class="o">.</span><span class="n">trackDict</span><span class="p">[</span><span class="s">&quot;GC&quot;</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">currentPCA</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">currentPCA</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">PCDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">currentPCA</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">PCDict</span> 
    
            </div>
<div class="viewcode-block" id="binnedData.doEig"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.doEig">[docs]</a>    <span class="k">def</span> <span class="nf">doEig</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">numPCs</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span><span class="n">force</span> <span class="o">=</span> <span class="bp">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;performs eigenvector expansion on the data</span>
<span class="sd">        creates dictionary self.EigDict with results</span>
<span class="sd">        Last row of the eigenvector matrix is the largest eigenvector, etc. </span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Dictionary of eigenvector matrices for different datasets</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">neededKeys</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;RemovedZeros&quot;</span><span class="p">,</span><span class="s">&quot;Corrected&quot;</span><span class="p">,</span><span class="s">&quot;FakedCis&quot;</span><span class="p">]</span>
        <span class="n">advicedKeys</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;TruncedTrans&quot;</span><span class="p">,</span><span class="s">&quot;RemovedPoor&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">force</span> <span class="o">==</span> <span class="bp">False</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_checkAppliedOperations</span><span class="p">(</span><span class="n">neededKeys</span><span class="p">,</span> <span class="n">advicedKeys</span><span class="p">)</span>

        
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>             
            <span class="n">currentEIG</span><span class="p">,</span><span class="n">eigenvalues</span> <span class="o">=</span> <span class="n">EIG</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="n">numPCs</span> <span class="o">=</span> <span class="n">numPCs</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">eigEigenvalueDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">eigenvalues</span> 
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">currentEIG</span><span class="p">)):</span>
                <span class="k">if</span> <span class="n">spearmanr</span><span class="p">(</span><span class="n">currentEIG</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="bp">self</span><span class="o">.</span><span class="n">trackDict</span><span class="p">[</span><span class="s">&quot;GC&quot;</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">currentEIG</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">currentEIG</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">EigDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">currentEIG</span>                                      
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">EigDict</span>
    
    </div>
<div class="viewcode-block" id="binnedData.cisToTrans"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedData.cisToTrans">[docs]</a>    <span class="k">def</span> <span class="nf">cisToTrans</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">mode</span> <span class="o">=</span> <span class="s">&quot;All&quot;</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="s">&quot;GM-all&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculates cis-to-trans ratio.</span>
<span class="sd">        &quot;All&quot; - treating SS as trans reads</span>
<span class="sd">        &quot;Dummy&quot; - fake SS reads proportional to cis reads with the same total sum</span>
<span class="sd">        &quot;Matrix&quot; - use heatmap only        </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">filename</span><span class="p">]</span>
        <span class="n">cismap</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[:,</span><span class="bp">None</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[</span><span class="bp">None</span><span class="p">,:]</span>        
        <span class="n">cissums</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">cismap</span> <span class="o">*</span> <span class="n">data</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span> 
        <span class="n">allsums</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>    
        <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s">&quot;All&quot;</span><span class="p">:</span>
            <span class="n">cissums</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">singlesDict</span><span class="p">[</span><span class="n">filename</span><span class="p">]</span>
            <span class="n">allsums</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">singlesDict</span><span class="p">[</span><span class="n">filename</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s">&quot;Dummy&quot;</span><span class="p">:</span>
            <span class="n">sm</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">singlesDict</span><span class="p">[</span><span class="n">filename</span><span class="p">])</span>
            <span class="n">fakesm</span> <span class="o">=</span> <span class="n">cissums</span> <span class="o">*</span> <span class="n">sm</span> <span class="o">/</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">cissums</span><span class="p">)</span>
            <span class="n">cissums</span> <span class="o">+=</span> <span class="n">fakesm</span>
            <span class="n">allsums</span> <span class="o">+=</span> <span class="n">fakesm</span>
        <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s">&quot;Matrix&quot;</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> 
        <span class="k">return</span> <span class="n">cissums</span> <span class="o">/</span> <span class="n">allsums</span>
    

</div></div>
<div class="viewcode-block" id="binnedDataAnalysis"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedDataAnalysis">[docs]</a><span class="k">class</span> <span class="nc">binnedDataAnalysis</span><span class="p">(</span><span class="n">binnedData</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class containing experimental features and data analysis scripts</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
<div class="viewcode-block" id="binnedDataAnalysis.plotScaling"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedDataAnalysis.plotScaling">[docs]</a>    <span class="k">def</span> <span class="nf">plotScaling</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">name</span><span class="p">,</span><span class="n">label</span> <span class="o">=</span> <span class="s">&quot;BLA&quot;</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="bp">None</span><span class="p">):</span>
        <span class="s">&quot;plots scaling of a heatmap,treating arms separately&quot;</span>
        <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
        <span class="n">bins</span> <span class="o">=</span> <span class="n">numutils</span><span class="o">.</span><span class="n">logbins</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">maxChrmArm</span><span class="o">/</span><span class="bp">self</span><span class="o">.</span><span class="n">resolution</span><span class="p">,</span><span class="mf">1.17</span><span class="p">)</span>
        <span class="n">s</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> 
        <span class="n">mask</span> <span class="o">=</span> <span class="n">s</span><span class="p">[:,</span><span class="bp">None</span><span class="p">]</span> <span class="o">*</span> <span class="n">s</span><span class="p">[</span><span class="bp">None</span><span class="p">,:]</span>
        <span class="n">chroms</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">masks</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span><span class="p">):</span>
            <span class="n">beg</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">end</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">centromerePositions</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">chroms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">,</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">])</span>
            <span class="n">masks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mask</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">,</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">])</span>
            <span class="n">beg</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">centromerePositions</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">end</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeEnds</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">chroms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">,</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">])</span>
            <span class="n">masks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mask</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">,</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">])</span>         
        <span class="n">observed</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">expected</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">bins</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
            <span class="n">low</span> <span class="o">=</span> <span class="n">bins</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">high</span> <span class="o">=</span> <span class="n">bins</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">obs</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">exp</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">chroms</span><span class="p">)):</span>
                <span class="k">if</span> <span class="n">low</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">chroms</span><span class="p">[</span><span class="n">j</span><span class="p">]):</span> <span class="k">continue</span> 
                <span class="n">high2</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">high</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">chroms</span><span class="p">[</span><span class="n">j</span><span class="p">]))</span>
                <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="n">low</span><span class="p">,</span><span class="n">high2</span><span class="p">):</span>
                    <span class="n">obs</span> <span class="o">+=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">diag</span><span class="p">(</span><span class="n">chroms</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">k</span><span class="p">))</span>
                    <span class="n">exp</span> <span class="o">+=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">diag</span><span class="p">(</span><span class="n">masks</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">k</span><span class="p">))</span>                    
            <span class="n">observed</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">obs</span><span class="p">)</span>
            <span class="n">expected</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">exp</span><span class="p">)</span>
        <span class="n">observed</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">observed</span><span class="p">,</span><span class="nb">float</span><span class="p">)</span>
        <span class="n">expected</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">expected</span><span class="p">,</span><span class="nb">float</span><span class="p">)</span>
        <span class="n">values</span> <span class="o">=</span> <span class="n">observed</span><span class="o">/</span><span class="n">expected</span>
        <span class="n">bins</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">bins</span><span class="p">,</span><span class="nb">float</span><span class="p">)</span>
        <span class="n">bins2</span> <span class="o">=</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="p">(</span><span class="n">bins</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">bins</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
        <span class="n">norm</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">values</span> <span class="o">*</span> <span class="p">(</span><span class="n">bins</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="o">-</span> <span class="n">bins</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>
        <span class="n">args</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">resolution</span> <span class="o">*</span> <span class="n">bins2</span> <span class="o">/</span> <span class="mi">1000000</span><span class="p">,</span><span class="n">values</span><span class="o">/</span><span class="p">(</span><span class="mf">1.</span> <span class="o">*</span> <span class="n">norm</span><span class="p">)]</span>
        <span class="k">if</span> <span class="n">color</span> <span class="o">!=</span> <span class="bp">None</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">color</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="n">label</span><span class="p">,</span><span class="n">linewidth</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span>
        


        </div>
<div class="viewcode-block" id="binnedDataAnalysis.averageTransMap"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedDataAnalysis.averageTransMap">[docs]</a>    <span class="k">def</span> <span class="nf">averageTransMap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">name</span> <span class="p">,</span> <span class="n">mycmap</span> <span class="o">=</span> <span class="s">&quot;hot_r&quot;</span><span class="p">,</span><span class="n">vmin</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span><span class="n">vmax</span> <span class="o">=</span> <span class="bp">None</span><span class="p">):</span>
        <span class="s">&quot;plots and returns average inter-chromosomal inter-arm map&quot;</span>
        <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>                 
        <span class="n">avarms</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">80</span><span class="p">,</span><span class="mi">80</span><span class="p">))</span>
        <span class="n">avmasks</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">80</span><span class="p">,</span><span class="mi">80</span><span class="p">))</span>
        <span class="n">discardCutoff</span> <span class="o">=</span> <span class="mi">10</span>
                
        
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span><span class="p">):</span>
            <span class="k">print</span> <span class="n">i</span> 
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">]:</span>
                    <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">]:</span>
                        <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="n">j</span><span class="p">:</span> <span class="k">continue</span> 
                                                
                        <span class="n">cenbeg1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">cntrStarts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">resolution</span>
                        <span class="n">cenbeg2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">cntrStarts</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">resolution</span>
                        <span class="n">cenend1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">cntrEnds</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">resolution</span>
                        <span class="n">cenend2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">cntrEnds</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">resolution</span>
                        
                        <span class="n">beg1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                        <span class="n">beg2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
                        <span class="n">end1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeEnds</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                        <span class="n">end2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeEnds</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
                        <span class="k">if</span> <span class="n">k</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                            <span class="n">bx</span> <span class="o">=</span> <span class="n">cenbeg1</span>
                            <span class="n">ex</span> <span class="o">=</span> <span class="n">beg1</span><span class="o">-</span><span class="mi">1</span>
                            <span class="n">dx</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>                            
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">bx</span> <span class="o">=</span> <span class="n">cenend1</span>
                            <span class="n">ex</span> <span class="o">=</span> <span class="n">end1</span>
                            <span class="n">dx</span> <span class="o">=</span> <span class="mi">1</span>
                        <span class="k">if</span> <span class="n">l</span> <span class="o">==</span><span class="mi">1</span><span class="p">:</span>
                            <span class="n">by</span> <span class="o">=</span> <span class="n">cenbeg2</span>
                            <span class="n">ey</span> <span class="o">=</span> <span class="n">beg2</span><span class="o">-</span><span class="mi">1</span>
                            <span class="n">dy</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">by</span> <span class="o">=</span> <span class="n">cenend2</span>
                            <span class="n">ey</span> <span class="o">=</span> <span class="n">end2</span>
                            <span class="n">dy</span> <span class="o">=</span> <span class="mi">1</span>
                                                    
                        <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">bx</span> <span class="o">-</span> <span class="n">ex</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">discardCutoff</span><span class="p">:</span> <span class="k">continue</span>
                        <span class="k">if</span> <span class="n">bx</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span> <span class="n">bx</span> <span class="o">=</span> <span class="bp">None</span>
                        <span class="k">if</span> <span class="n">ex</span> <span class="o">&lt;</span><span class="mi">0</span><span class="p">:</span> <span class="n">ex</span> <span class="o">=</span> <span class="bp">None</span>                          
                        <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">by</span> <span class="o">-</span> <span class="n">ey</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">discardCutoff</span><span class="p">:</span> <span class="k">continue</span>
                        <span class="k">if</span> <span class="n">by</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span> <span class="n">by</span> <span class="o">=</span> <span class="bp">None</span>
                        <span class="k">if</span> <span class="n">ey</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span> <span class="n">ey</span> <span class="o">=</span> <span class="bp">None</span> 
                                                 
                                                    
                        <span class="n">arms</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">bx</span><span class="p">:</span><span class="n">ex</span><span class="p">:</span><span class="n">dx</span><span class="p">,</span><span class="n">by</span><span class="p">:</span><span class="n">ey</span><span class="p">:</span><span class="n">dy</span><span class="p">]</span>                                                 
                        <span class="k">assert</span> <span class="nb">max</span><span class="p">(</span><span class="n">arms</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">maxChrmArm</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">resolution</span> <span class="o">+</span> <span class="mi">2</span>
                                                
                        <span class="n">mx</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">arms</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
                        <span class="n">my</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">arms</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
                        <span class="n">maskx</span> <span class="o">=</span> <span class="n">mx</span> <span class="o">==</span> <span class="mi">0</span> 
                        <span class="n">masky</span> <span class="o">=</span> <span class="n">my</span> <span class="o">==</span> <span class="mi">0</span>
                        <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">maskx</span><span class="p">[</span><span class="bp">None</span><span class="p">,:]</span> <span class="o">+</span> <span class="n">masky</span><span class="p">[:,</span><span class="bp">None</span><span class="p">])</span> <span class="o">==</span> <span class="bp">False</span>                    
                        <span class="n">maskf</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span><span class="nb">float</span><span class="p">)</span>
                        <span class="n">mlenx</span> <span class="o">=</span> <span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> 
                        <span class="n">mleny</span> <span class="o">=</span> <span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
                        
                        <span class="k">if</span> <span class="nb">min</span><span class="p">(</span><span class="n">mlenx</span><span class="p">,</span> <span class="n">mleny</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">discardCutoff</span><span class="p">:</span> <span class="k">continue</span>
                                                 
                        
                        <span class="n">add</span> <span class="o">=</span> <span class="n">numutils</span><span class="o">.</span><span class="n">zoomOut</span><span class="p">(</span><span class="n">arms</span><span class="p">,</span><span class="n">avarms</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
                        <span class="k">assert</span> <span class="n">numpy</span><span class="o">.</span><span class="n">abs</span><span class="p">((</span><span class="n">arms</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">-</span> <span class="n">add</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span> <span class="o">/</span> <span class="n">arms</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span> <span class="o">&lt;</span> <span class="mf">0.02</span>
                        
                         
                                                                        
                        <span class="n">addmask</span> <span class="o">=</span> <span class="n">numutils</span><span class="o">.</span><span class="n">zoomOut</span><span class="p">(</span><span class="n">maskf</span><span class="p">,</span><span class="n">avarms</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
                        <span class="n">avarms</span> <span class="o">+=</span> <span class="n">add</span> 
                        <span class="n">avmasks</span> <span class="o">+=</span> <span class="n">addmask</span>
                          
        
        <span class="n">avarms</span> <span class="o">/=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">avarms</span><span class="p">)</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">avarms</span> <span class="o">/</span> <span class="n">avmasks</span>
        <span class="n">data</span> <span class="o">/=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> 
        <span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">numutils</span><span class="o">.</span><span class="n">trunc</span><span class="p">(</span><span class="n">data</span><span class="p">)),</span><span class="n">cmap</span> <span class="o">=</span> <span class="s">&quot;jet&quot;</span><span class="p">,</span><span class="n">interpolation</span> <span class="o">=</span> <span class="s">&quot;nearest&quot;</span><span class="p">,</span><span class="n">vmin</span> <span class="o">=</span> <span class="n">vmin</span><span class="p">,</span> <span class="n">vmax</span> <span class="o">=</span> <span class="n">vmax</span><span class="p">)</span>
        <span class="n">removeBorder</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">numutils</span><span class="o">.</span><span class="n">trunc</span><span class="p">(</span><span class="n">data</span><span class="p">))</span>
            </div>
<div class="viewcode-block" id="binnedDataAnalysis.perArmCorrelation"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedDataAnalysis.perArmCorrelation">[docs]</a>    <span class="k">def</span> <span class="nf">perArmCorrelation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">data1</span><span class="p">,</span><span class="n">data2</span><span class="p">,</span><span class="n">doByArms</span> <span class="o">=</span> <span class="p">[]):</span>
        <span class="sd">&quot;&quot;&quot;does inter-chromosomal spearman correlation </span>
<span class="sd">        of two vectors for each chromosomes separately.</span>
<span class="sd">        Averages over chromosomes with weight of chromosomal length</span>
<span class="sd">        For chromosomes in &quot;doByArms&quot; treats arms as separatre chromosomes</span>
<span class="sd">        returns average Spearman r correlation</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">cr</span> <span class="o">=</span> <span class="mi">0</span> 
        <span class="n">ln</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">doByArms</span><span class="p">:</span>
                <span class="n">beg</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="n">end</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">centromerePositions</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="n">beg</span><span class="p">:</span> 
                    <span class="n">cr</span> <span class="o">+=</span> <span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">spearmanr</span><span class="p">(</span><span class="n">data1</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">],</span><span class="n">data2</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">])[</span><span class="mi">0</span><span class="p">]))</span> <span class="o">*</span> <span class="p">(</span><span class="n">end</span> <span class="o">-</span> <span class="n">beg</span><span class="p">)</span>            
                    <span class="n">ln</span> <span class="o">+=</span> <span class="p">(</span><span class="n">end</span><span class="o">-</span><span class="n">beg</span><span class="p">)</span>
                    <span class="k">print</span> <span class="n">spearmanr</span><span class="p">(</span><span class="n">data1</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">],</span><span class="n">data2</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">beg</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">centromerePositions</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="n">end</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeEnds</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="n">beg</span><span class="p">:</span> 
                    <span class="n">cr</span> <span class="o">+=</span> <span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">spearmanr</span><span class="p">(</span><span class="n">data1</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">],</span><span class="n">data2</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">])[</span><span class="mi">0</span><span class="p">]))</span> <span class="o">*</span> <span class="p">(</span><span class="n">end</span> <span class="o">-</span> <span class="n">beg</span><span class="p">)</span>            
                    <span class="n">ln</span> <span class="o">+=</span> <span class="p">(</span><span class="n">end</span><span class="o">-</span><span class="n">beg</span><span class="p">)</span>
                    <span class="k">print</span> <span class="n">spearmanr</span><span class="p">(</span><span class="n">data1</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">],</span><span class="n">data2</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span> 
                <span class="n">beg</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="n">end</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeEnds</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="n">beg</span><span class="p">:</span> 
                    <span class="n">cr</span> <span class="o">+=</span> <span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">spearmanr</span><span class="p">(</span><span class="n">data1</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">],</span><span class="n">data2</span><span class="p">[</span><span class="n">beg</span><span class="p">:</span><span class="n">end</span><span class="p">])[</span><span class="mi">0</span><span class="p">]))</span> <span class="o">*</span> <span class="p">(</span><span class="n">end</span> <span class="o">-</span> <span class="n">beg</span><span class="p">)</span>            
                    <span class="n">ln</span> <span class="o">+=</span> <span class="p">(</span><span class="n">end</span><span class="o">-</span><span class="n">beg</span><span class="p">)</span>                            
        <span class="k">return</span> <span class="n">cr</span><span class="o">/</span><span class="n">ln</span> 
 
 </div>
<div class="viewcode-block" id="binnedDataAnalysis.divideOutAveragesPerChromosome"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedDataAnalysis.divideOutAveragesPerChromosome">[docs]</a>    <span class="k">def</span> <span class="nf">divideOutAveragesPerChromosome</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="s">&quot;divides each interchromosomal map by it&#39;s mean value&quot;</span>
        <span class="n">mask2D</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_giveMask2D</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">chrom1</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">chrom2</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="n">value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                    <span class="n">submatrix</span> <span class="o">=</span> <span class="n">value</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">[</span><span class="n">chrom1</span><span class="p">]:</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeEnds</span><span class="p">[</span><span class="n">chrom1</span><span class="p">],</span>
                                      <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">[</span><span class="n">chrom2</span><span class="p">]:</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeEnds</span><span class="p">[</span><span class="n">chrom2</span><span class="p">]]</span>
                    <span class="n">masksum</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mask2D</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">[</span><span class="n">chrom1</span><span class="p">]:</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeEnds</span><span class="p">[</span><span class="n">chrom1</span><span class="p">],</span>
                                      <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">[</span><span class="n">chrom2</span><span class="p">]:</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeEnds</span><span class="p">[</span><span class="n">chrom2</span><span class="p">]])</span>
                    <span class="n">valuesum</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">submatrix</span><span class="p">)</span>
                    <span class="n">mean</span> <span class="o">=</span> <span class="n">valuesum</span> <span class="o">/</span> <span class="n">masksum</span>
                    <span class="n">submatrix</span> <span class="o">/=</span> <span class="n">mean</span>
 
</div>
<div class="viewcode-block" id="binnedDataAnalysis.interchromosomalValues"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.binnedDataAnalysis.interchromosomalValues">[docs]</a>    <span class="k">def</span> <span class="nf">interchromosomalValues</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="s">&quot;GM-all&quot;</span><span class="p">,</span><span class="n">returnAll</span> <span class="o">=</span> <span class="bp">False</span><span class="p">):</span>
        <span class="s">&quot;returns average inter-chromosome-interaction values, ordered always the same way&quot;</span>
        <span class="n">values</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[:,</span><span class="bp">None</span><span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[</span><span class="bp">None</span><span class="p">,:]</span>
        <span class="n">values</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[:,</span><span class="bp">None</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[</span><span class="bp">None</span><span class="p">,:]]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span> <span class="o">-</span> <span class="mi">1</span> 
        <span class="c">#mat_img(values)</span>
        <span class="n">uv</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">values</span><span class="p">))[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> 
        <span class="n">probs</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">values</span><span class="o">.</span><span class="n">ravel</span><span class="p">(),</span><span class="n">weights</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">filename</span><span class="p">]</span><span class="o">.</span><span class="n">ravel</span><span class="p">())</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">values</span><span class="o">.</span><span class="n">ravel</span><span class="p">())</span>
        <span class="k">if</span> <span class="n">returnAll</span> <span class="o">==</span> <span class="bp">False</span><span class="p">:</span> <span class="k">return</span> <span class="n">probs</span><span class="p">[</span><span class="n">uv</span><span class="p">]</span> <span class="o">/</span> <span class="n">counts</span><span class="p">[</span><span class="n">uv</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span> 
            <span class="n">probs</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span> 
            <span class="n">values</span> <span class="o">=</span> <span class="n">probs</span><span class="o">/</span><span class="n">counts</span>
            <span class="n">values</span><span class="p">[</span><span class="n">counts</span><span class="o">==</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>             
            <span class="c">#mat_img(values.reshape((22,22)))</span>
            <span class="k">return</span> <span class="n">values</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span><span class="p">))</span>


    </div></div>
<div class="viewcode-block" id="experimentalBinnedData"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.experimentalBinnedData">[docs]</a><span class="k">class</span> <span class="nc">experimentalBinnedData</span><span class="p">(</span><span class="n">binnedData</span><span class="p">):</span>
    <span class="s">&quot;Contains some poorly-implemented new features&quot;</span>        
<div class="viewcode-block" id="experimentalBinnedData.emulateCis"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.experimentalBinnedData.emulateCis">[docs]</a>    <span class="k">def</span> <span class="nf">emulateCis</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;if you want to have fun creating syntetic data, this emulates cis contacts. </span>
<span class="sd">        adjust cis/trans ratio in the C code&quot;&quot;&quot;</span>
        <span class="n">transmap</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[:,</span><span class="bp">None</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[</span><span class="bp">None</span><span class="p">,:]</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">transmap</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="mf">1.</span> 
            <span class="c">#mask = na(self.chromosomeIndex[:,None] == self.chromosomeIndex[None,:],int)</span>
            <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
            <span class="n">N</span>
            <span class="n">code</span> <span class="o">=</span> <span class="s">r&quot;&quot;&quot;</span>
<span class="s">            #line 841 &quot;binary_search.py&quot;</span>
<span class="s">            using namespace std; </span>
<span class="s">            for (int i = 0; i &lt; N; i++)    </span>
<span class="s">            {                 </span>
<span class="s">                for (int j = 0; j&lt;N; j++)</span>
<span class="s">                {</span>
<span class="s">                    if (transmap[N * i + j] == 1)</span>
<span class="s">                    {</span>
<span class="s">                        data[N * i + j] = data[N * i +j] * 300 /(abs(i-j) + 0.5);                     </span>
<span class="s">                    }</span>
<span class="s">                }</span>
<span class="s">            } </span>
<span class="s">            &quot;&quot;&quot;</span>
            <span class="n">support</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">            #include &lt;math.h&gt;</span>
<span class="s">            &quot;&quot;&quot;</span>
            <span class="n">weave</span><span class="o">.</span><span class="n">inline</span><span class="p">(</span><span class="n">code</span><span class="p">,</span> <span class="p">[</span><span class="s">&#39;transmap&#39;</span><span class="p">,</span><span class="s">&#39;data&#39;</span><span class="p">,</span><span class="s">&quot;N&quot;</span><span class="p">],</span> <span class="n">extra_compile_args</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;-march=native -malign-double -O3&#39;</span><span class="p">],</span><span class="n">support_code</span> <span class="o">=</span><span class="n">support</span> <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">removedCis</span> <span class="o">=</span> <span class="bp">False</span> 
        <span class="bp">self</span><span class="o">.</span><span class="n">fakedCis</span> <span class="o">=</span> <span class="bp">False</span>
         

</div>
<div class="viewcode-block" id="experimentalBinnedData.fakeMissing"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.experimentalBinnedData.fakeMissing">[docs]</a>    <span class="k">def</span> <span class="nf">fakeMissing</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">iterative</span> <span class="o">=</span> <span class="bp">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;fakes megabases that have no reads. For cis reads fakes with cis reads at the same distance. For trans fakes with random trans read at the same diagonal. </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="mf">1.</span>
            <span class="n">sm</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span>  
            <span class="n">mask</span> <span class="o">=</span> <span class="n">sm</span><span class="p">[:,</span><span class="bp">None</span><span class="p">]</span> <span class="o">*</span> <span class="n">sm</span><span class="p">[</span><span class="bp">None</span><span class="p">,:]</span>
            <span class="n">transmask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[:,</span><span class="bp">None</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[</span><span class="bp">None</span><span class="p">,:],</span><span class="nb">int</span><span class="p">)</span>
            <span class="c">#mat_img(transmask)</span>
            <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
            <span class="n">N</span><span class="p">,</span><span class="n">transmask</span> <span class="p">,</span><span class="n">mask</span> <span class="c">#to remove warning</span>
            <span class="n">code</span> <span class="o">=</span> <span class="s">r&quot;&quot;&quot;</span>
<span class="s">            #line 841 &quot;binary_search.py&quot;</span>
<span class="s">            using namespace std; </span>
<span class="s">            for (int i = 0; i &lt; N; i++)    </span>
<span class="s">            {    </span>
<span class="s">                for (int j = i; j&lt;N; j++)</span>
<span class="s">                {</span>
<span class="s">                    if ((MASK2(i,j) == 0) )                    </span>
<span class="s">                    {</span>
<span class="s">                    for (int ss = 0; ss &lt; 401; ss++) </span>
<span class="s">                        {</span>
<span class="s">                        int k = 0;                            </span>
<span class="s">                        int s = rand() % (N - (j-i));                        </span>
<span class="s">                        if ((mask[s * N + s + j - i] == 1) &amp;&amp; ((transmask[s * N + s + j - i] == transmask[i * N + j]) || (ss &gt; 200)) )</span>
<span class="s">                                {                                </span>
<span class="s">                                data[i * N + j] = data[s * N + s + j - i];</span>
<span class="s">                                data[j * N + i] = data[s * N + s + j - i];</span>
<span class="s">                                break;</span>
<span class="s">                                }</span>
<span class="s">                        if (ss == 400) {printf(&quot;Cannot fake one point... skipping </span><span class="si">%d</span><span class="s"> </span><span class="si">%d</span><span class="s"> \n&quot;,i,j);}</span>
<span class="s">                            </span>
<span class="s">                            </span>
<span class="s">                        }</span>
<span class="s">                    }</span>
<span class="s">                }</span>
<span class="s">            } </span>
<span class="s">            &quot;&quot;&quot;</span>
            <span class="n">support</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">            #include &lt;math.h&gt;</span>
<span class="s">            &quot;&quot;&quot;</span>
            <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="mi">5</span><span class="p">):</span>
                <span class="n">s</span> <span class="c">#to remove warning</span>
                <span class="n">weave</span><span class="o">.</span><span class="n">inline</span><span class="p">(</span><span class="n">code</span><span class="p">,</span> <span class="p">[</span><span class="s">&#39;transmask&#39;</span><span class="p">,</span><span class="s">&#39;mask&#39;</span><span class="p">,</span><span class="s">&#39;data&#39;</span><span class="p">,</span><span class="s">&quot;N&quot;</span><span class="p">],</span> <span class="n">extra_compile_args</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;-march=native -malign-double -O3&#39;</span><span class="p">],</span><span class="n">support_code</span> <span class="o">=</span><span class="n">support</span> <span class="p">)</span>                
                <span class="n">data</span> <span class="o">=</span> <span class="n">correct</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>              
            <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span>
             
            
            <span class="c">#mat_img(self.dataDict[i]&gt;0)</span>
            

    
</div>
<div class="viewcode-block" id="experimentalBinnedData.iterativeCorrectByTrans"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.experimentalBinnedData.iterativeCorrectByTrans">[docs]</a>    <span class="k">def</span> <span class="nf">iterativeCorrectByTrans</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">names</span> <span class="o">=</span> <span class="bp">None</span><span class="p">):</span>        
        <span class="sd">&quot;&quot;&quot;performs iterative correction by trans data only, corrects cis also</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        names : list of str or None, optional</span>
<span class="sd">            Keys of datasets to be corrected. By default, all are corrected.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">names</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span> <span class="n">names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>        
        <span class="bp">self</span><span class="o">.</span><span class="n">transmap</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[:,</span><span class="bp">None</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeIndex</span><span class="p">[</span><span class="bp">None</span><span class="p">,:]</span>
        <span class="c">#mat_img(self.transmap)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">names</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">dataDict</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="bp">self</span><span class="o">.</span><span class="n">biasDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">numutils</span><span class="o">.</span><span class="n">ultracorrectSymmetricByMask</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">transmap</span><span class="p">,</span><span class="mi">50</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">singlesDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">/=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biasDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="k">except</span><span class="p">:</span> <span class="k">print</span> <span class="s">&quot;bla&quot;</span>
    </div>
<div class="viewcode-block" id="experimentalBinnedData.loadWigFile"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.experimentalBinnedData.loadWigFile">[docs]</a>    <span class="k">def</span> <span class="nf">loadWigFile</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">filename</span><span class="p">,</span><span class="n">label</span><span class="p">,</span><span class="n">control</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span><span class="n">wigFileType</span> <span class="o">=</span> <span class="s">&quot;Auto&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Currently only fixed-step wig files and bigWig files are supported!!!</span>
<span class="sd">        Import from fixedStep wig files is very fast, however is not the most reliable.</span>
<span class="sd">        </span>
<span class="sd">        TODO: rewrite parser to support any resolution.  </span>
<span class="sd">        </span>
<span class="sd">        for VariableStep files use wigToBigWig utility to convert them to bigWig format first. </span>
<span class="sd">        To use it you will also need to have fetchChromSizes script. </span>
<span class="sd">        </span>
<span class="sd">        Then you just run </span>
<span class="sd">        $bash fetchChromSizes hg18 &gt; hg18.chrom.sizes</span>
<span class="sd">        $./wigToBigWig myWig.wig hg18.chrom.sizes myWig.bigWig</span>
<span class="sd">        </span>
<span class="sd">        And you enjoy your favourite bigWig. </span>
<span class="sd">        </span>
<span class="sd">        BigWig import is implemented using bx-python module.</span>
<span class="sd">        It is normally very fast; however, it has a bug at low resolutions. </span>
<span class="sd">        I have an ugly workaround for it (chopping the quiery into many smaller pieces), but I hope </span>
<span class="sd">        that they actually fix this bug. </span>
<span class="sd">        </span>
<span class="sd">        Anyway, check their repo on BitBucket, maybe they&#39;ve fixed my issue # 39 :) </span>
<span class="sd">        https://bitbucket.org/james_taylor/bx-python/overview</span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
        
        <span class="k">def</span> <span class="nf">loadFile</span><span class="p">(</span><span class="n">name</span><span class="p">,</span><span class="n">wigFileType</span> <span class="o">=</span> <span class="n">wigFileType</span><span class="p">):</span>
            <span class="sd">&quot;&quot;&quot;Choosing right method to load wigfile&quot;&quot;&quot;</span>
            
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="o">==</span> <span class="bp">False</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s"> Wig file not found : </span><span class="si">%s</span><span class="s"> &quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">name</span><span class="p">)))</span>
                        
            <span class="k">if</span> <span class="n">wigFileType</span> <span class="o">==</span> <span class="s">&quot;Auto&quot;</span><span class="p">:</span>
                <span class="n">ext</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">name</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">ext</span> <span class="o">==</span> <span class="s">&quot;&quot;</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Wig file has no extension. Please specify it&#39;s type&quot;</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">ext</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>  <span class="o">==</span> <span class="s">&quot;.wig&quot;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span><span class="o">.</span><span class="n">readline</span><span class="p">()[:</span><span class="mi">2</span><span class="p">]</span> <span class="o">!=</span> <span class="s">&quot;fi&quot;</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Cannot read non fixed-step wig files! Please use wigToBigWig utility. See docstring of this method.&quot;</span><span class="p">)</span>
                    <span class="n">wigFileType</span> <span class="o">=</span> <span class="s">&quot;wig&quot;</span>
                <span class="k">elif</span> <span class="n">ext</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s">&quot;.bigwig&quot;</span><span class="p">:</span>
                    <span class="n">wigFileType</span> <span class="o">=</span> <span class="s">&quot;bigwig&quot;</span>
                <span class="k">else</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Unknown extension of wig file: </span><span class="si">%s</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">ext</span><span class="p">)</span>
                             
            <span class="k">if</span> <span class="n">wigFileType</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s">&quot;wig&quot;</span><span class="p">:</span> 
                <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">parseFixedStepWigAtKbResolution</span><span class="p">(</span><span class="n">name</span><span class="p">,</span><span class="n">resolution</span> <span class="o">=</span> <span class="mi">5000</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">wigFileType</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s">&quot;bigwig&quot;</span><span class="p">:</span> 
                <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">parseBigWigFile</span><span class="p">(</span><span class="n">name</span><span class="p">,</span><span class="n">resolution</span> <span class="o">=</span> <span class="mi">5000</span><span class="p">,</span><span class="n">divideByValidCounts</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Wrong type of wig file : </span><span class="si">%s</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">wigFileType</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">data</span>
        
        <span class="s">&quot;Loading the data files&quot;</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">loadFile</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">control</span> <span class="o">!=</span> <span class="bp">None</span><span class="p">:</span> 
            <span class="n">controlData</span> <span class="o">=</span> <span class="n">loadFile</span><span class="p">(</span><span class="n">control</span><span class="p">)</span> 
        
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">resolution</span> <span class="o">%</span> <span class="mi">5000</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Cannot parse wig file at resolution that is not a multiply of 5 kb&quot;</span><span class="p">)</span>
                 
        <span class="n">vector</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span><span class="p">,</span><span class="nb">float</span><span class="p">)</span>  <span class="c">#resulting array</span>
        <span class="k">for</span> <span class="n">chrom</span><span class="p">,</span><span class="n">value</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>            
            <span class="n">value</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>            
            <span class="k">if</span> <span class="n">control</span> <span class="o">!=</span> <span class="bp">None</span><span class="p">:</span> 
                <span class="n">chromControl</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">controlData</span><span class="p">[</span><span class="n">chrom</span><span class="p">])</span>
                <span class="n">vmask</span> <span class="o">=</span> <span class="n">value</span> <span class="o">!=</span> <span class="mi">0</span> 
                <span class="n">cmask</span> <span class="o">=</span> <span class="n">chromControl</span> <span class="o">!=</span> <span class="mi">0</span>
                <span class="n">keepmask</span> <span class="o">=</span> <span class="n">vmask</span> <span class="o">*</span> <span class="n">cmask</span>
                <span class="n">vmasksum</span><span class="p">,</span><span class="n">cmasksum</span> <span class="o">=</span> <span class="n">vmask</span><span class="o">.</span><span class="n">sum</span><span class="p">(),</span><span class="n">cmask</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
                <span class="k">if</span> <span class="nb">max</span><span class="p">(</span><span class="n">vmasksum</span><span class="p">,</span><span class="n">cmasksum</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="mf">1.</span> <span class="o">*</span> <span class="nb">min</span><span class="p">(</span><span class="n">vmasksum</span><span class="p">,</span><span class="n">cmasksum</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mf">1.3</span><span class="p">:</span>
                    <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">Big deviation: number of non-zero data points: </span><span class="si">%s</span><span class="s">, control points:</span><span class="si">%s</span><span class="s">.&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">vmasksum</span><span class="p">,</span><span class="n">cmasksum</span><span class="p">))</span>    
                <span class="n">value</span><span class="p">[</span><span class="o">-</span><span class="n">keepmask</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span> 
                <span class="n">value</span><span class="p">[</span><span class="n">keepmask</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span><span class="p">[</span><span class="n">keepmask</span><span class="p">]</span> <span class="o">/</span> <span class="n">chromControl</span><span class="p">[</span><span class="n">keepmask</span><span class="p">]</span>                    
                                          
            <span class="n">value</span><span class="o">.</span><span class="n">resize</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmLensBin</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">resolution</span><span class="o">/</span><span class="mi">5000</span><span class="p">))</span>            
            <span class="n">value</span><span class="o">.</span><span class="n">shape</span> <span class="o">=</span> <span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">resolution</span> <span class="o">/</span> <span class="mi">5000</span> <span class="p">)</span>
            <span class="k">if</span> <span class="n">value</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Chromosome </span><span class="si">%s</span><span class="s"> contains zero data in wig file </span><span class="si">%s</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">idx2label</span><span class="p">[</span><span class="n">chrom</span><span class="p">],</span><span class="n">filename</span><span class="p">))</span>
            <span class="n">mask</span> <span class="o">=</span> <span class="n">value</span> <span class="o">==</span> <span class="mi">0</span>
            <span class="n">value</span><span class="p">[</span><span class="o">-</span><span class="n">mask</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">value</span><span class="p">[</span><span class="o">-</span><span class="n">mask</span><span class="p">])</span> 
              
            <span class="n">valuesum</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">value</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">masksum</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mask</span> <span class="o">==</span> <span class="bp">False</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">valuesum</span><span class="p">[</span><span class="n">masksum</span><span class="o">==</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span> 
            <span class="n">vmask</span> <span class="o">=</span> <span class="n">valuesum</span> <span class="o">!=</span> <span class="mi">0</span> 
            <span class="n">valuesum</span><span class="p">[</span><span class="n">vmask</span><span class="p">]</span> <span class="o">/=</span> <span class="n">masksum</span><span class="p">[</span><span class="n">vmask</span><span class="p">]</span>
            <span class="n">valuesum</span><span class="p">[</span><span class="o">-</span><span class="n">vmask</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">valuesum</span><span class="p">[</span><span class="n">vmask</span><span class="p">])</span>  <span class="c">#setting all unknown points to the median of known points             </span>
                        
            <span class="n">vector</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmStartsBinCont</span><span class="p">[</span><span class="n">chrom</span><span class="p">]:</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmEndsBinCont</span><span class="p">[</span><span class="n">chrom</span><span class="p">]]</span> <span class="o">=</span> <span class="n">valuesum</span>        
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">vector</span><span class="p">)</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span><span class="p">:</span> 
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Length mismatch. Length of vector: </span><span class="si">%d</span><span class="s">, length of genome:</span><span class="si">%d</span><span class="s">&quot;</span> <span class="o">%</span> 
                             <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">vector</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span><span class="p">))</span> 
        <span class="bp">self</span><span class="o">.</span><span class="n">trackDict</span><span class="p">[</span><span class="n">label</span><span class="p">]</span> <span class="o">=</span> <span class="n">vector</span>
            
             </div>
<div class="viewcode-block" id="experimentalBinnedData.loadErezEigenvector1MB"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.experimentalBinnedData.loadErezEigenvector1MB">[docs]</a>    <span class="k">def</span> <span class="nf">loadErezEigenvector1MB</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">erezFolder</span><span class="p">):</span>
        <span class="s">&quot;Loads Erez chromatin domain eigenvector for HindIII&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">resolution</span> <span class="o">!=</span> <span class="mi">1000000</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Erez eigenvector is only at 1MB resolution&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">folderName</span> <span class="o">!=</span> <span class="s">&quot;hg18&quot;</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Erez eigenvector is for hg18 only!&quot;</span><span class="p">)</span>
        <span class="n">folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">erezFolder</span><span class="p">,</span><span class="s">&quot;GM-combined.ctgDATA1.ctgDATA1.1000000bp.hm.eigenvector.tab&quot;</span><span class="p">)</span>
        <span class="n">folder2</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">erezFolder</span><span class="p">,</span><span class="s">&quot;GM-combined.ctgDATA1.ctgDATA1.1000000bp.hm.eigenvector2.tab&quot;</span><span class="p">)</span>
        <span class="n">eigenvector</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span><span class="p">,</span><span class="nb">float</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">24</span><span class="p">):</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="n">folder</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&quot;DATA1&quot;</span><span class="p">,</span><span class="nb">str</span><span class="p">(</span><span class="n">chrom</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">]:</span> 
                <span class="n">filename</span> <span class="o">=</span> <span class="n">folder2</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&quot;DATA1&quot;</span><span class="p">,</span><span class="nb">str</span><span class="p">(</span><span class="n">chrom</span><span class="p">))</span>
            <span class="n">mydata</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="nb">float</span><span class="p">(</span><span class="n">j</span><span class="p">)</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">()]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()])</span>
            <span class="n">eigenvector</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmStartsBinCont</span><span class="p">[</span><span class="n">chrom</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mydata</span><span class="p">[:,</span><span class="mi">1</span><span class="p">],</span><span class="nb">int</span><span class="p">)]</span> <span class="o">=</span> <span class="n">mydata</span><span class="p">[:,</span><span class="mi">2</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">trackDict</span><span class="p">[</span><span class="s">&quot;Erez&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">eigenvector</span>            

            </div>
<div class="viewcode-block" id="experimentalBinnedData.loadTanayDomains"><a class="viewcode-back" href="../../binneddata.html#hiclib.binnedData.experimentalBinnedData.loadTanayDomains">[docs]</a>    <span class="k">def</span> <span class="nf">loadTanayDomains</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>             
        <span class="s">&quot;domains, extracted from Tanay paper image&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">folderName</span> <span class="o">!=</span> <span class="s">&quot;hg18&quot;</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Tanay domains work only with hg18&quot;</span><span class="p">)</span>
        <span class="n">data</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;0 - 17, 1 - 13.5, 2 - 6.5, 0 - 2, 2 - 2; x - 6.5, 0 - 6, 1 - 13.5, 0 - 1.5, 1 - 14.5 </span>
<span class="s">    1 - 8.5, 0 - 2.5, 1 - 14, 2 - 6; 0 - 1.5, 2 - 11.5, 1 - 35</span>
<span class="s">    1 - 14, 0-6, 2 - 11; 2 - 4.5, 1 - 5, 0 - 4, 1 -20.5, 0 - 2</span>
<span class="s">    0 - 3, 2 - 14; 2 - 5, 1 - 42</span>
<span class="s">    2 - 16; 2 - 7, 0 - 3, 1 - 18.5, 0 - 1, 1 - 13, 0 - 2.5 </span>
<span class="s">    0 - 2, 1 - 6.5, 0 - 7.5, 2 - 4; 2 - 6, 1 - 31</span>
<span class="s">    0 - 2, 1 - 11, 2 - 7; 2 - 7.5, 1 - 5, 0 - 3, 1 - 19 </span>
<span class="s">    2 - 9.5, 0 - 1, 2 - 5; 2 - 4, 1 - 27.5, 0 - 2.5</span>
<span class="s">    2 - 11.5, 0 - 2.5, x - 2.5; x - 5, 2 - 8, 0 - 3.5, 1 - 9, 0 - 6</span>
<span class="s">    2 - 13.5; 2 - 9, 0 - 3, 1 - 6, 0 - 3.5, 1 - 10.5</span>
<span class="s">    0 - 3.5, 2 - 15; 2 - 1, 0 - 7.5, 1 - 13, 0 - 1.5, 1 - 4</span>
<span class="s">    0 - 4, 2 - 8; 2 - 2, 0 - 5, 2 - 2.5, 1 - 13, 0 - 6.5, 1 - 3.5 </span>
<span class="s">    x - 5.5; 2 - 8.5, 0 - 1, 2 - 7, 1 - 16</span>
<span class="s">    x - 5.5; 2 - 14.5, 0 - 6, 2 - 3, 1 - 2.5, 2 - 1, 0 - 3</span>
<span class="s">    x - 5.5; 2 - 6, 0 - 3.5, 2 - 1.5, 0 - 11.5, 2 - 5.5</span>
<span class="s">    0 - 11, 2 - 1; x - 2.5, 2 - 6.5, 0 - 3, 2 - 2, 0 - 3.5 </span>
<span class="s">    0 - 4, 2 - 1.5, 0 - 1.5; 0 - 19</span>
<span class="s">    2 - 5; 2 - 20</span>
<span class="s">    0 - 9.5, x - 1.5; x - 1, 2 - 2, 0 - 8.5</span>
<span class="s">    0 - 2, 2 - 7; 0 - 8, 2 - 2, 0 - 1</span>
<span class="s">    x - 0.5; 2 - 8.5, 0 - 3</span>
<span class="s">    x - 4; 0 -12 </span>
<span class="s">    x - 1.5, 1 - 13, 2 - 5.5; 2 - 2, 1 - 29&quot;&quot;&quot;</span>
        <span class="n">chroms</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)]</span>
        <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>     
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chroms</span><span class="p">:</span>
            <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">([])</span>
            <span class="n">cur</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> 
            <span class="k">for</span> <span class="n">arm</span> <span class="ow">in</span> <span class="n">chrom</span><span class="p">:</span>
                
                <span class="k">for</span> <span class="n">enrty</span> <span class="ow">in</span> <span class="n">arm</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;,&quot;</span><span class="p">):</span>
                    <span class="n">spentry</span> <span class="o">=</span> <span class="n">enrty</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;-&quot;</span><span class="p">)</span>
                    <span class="k">if</span> <span class="s">&quot;x&quot;</span> <span class="ow">in</span> <span class="n">spentry</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span> <span class="n">value</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> 
                    <span class="k">else</span><span class="p">:</span> <span class="n">value</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">spentry</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                    <span class="n">cur</span> <span class="o">+=</span> <span class="p">([</span><span class="n">value</span><span class="p">]</span> <span class="o">*</span> <span class="nb">int</span><span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="nb">float</span><span class="p">(</span><span class="n">spentry</span><span class="p">[</span><span class="mi">1</span><span class="p">])))</span>
                <span class="n">cur</span> <span class="o">+=</span> <span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span>
        <span class="c">#lenses = [len(i) for i in result]</span>
            
        <span class="n">domains</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span><span class="p">,</span><span class="nb">int</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmCount</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmLens</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">resolution</span><span class="p">)):</span>
                <span class="n">domains</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmStartsBinCont</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="n">i</span><span class="p">][(</span><span class="n">j</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">/</span> <span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmLens</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">/</span><span class="bp">self</span><span class="o">.</span><span class="n">resolution</span><span class="p">)))</span> <span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">trackDict</span><span class="p">[</span><span class="s">&#39;TanayDomains&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">domains</span> 
            
            
            
         </div></div>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" size="18" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../../np-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../../index.html">mirnylab v0.1 documentation</a> &raquo;</li>
          <li><a href="../index.html" >Module code</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2012, the Leonids.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.1.
    </div>
  </body>
</html>