

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>hiclib.fragmentHiC &mdash; mirnylab v0.1 documentation</title>
    <link rel="stylesheet" href="../../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <link rel="top" title="mirnylab v0.1 documentation" href="../../index.html" />
    <link rel="up" title="Module code" href="../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../../np-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../../index.html">mirnylab v0.1 documentation</a> &raquo;</li>
          <li><a href="../index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <h1>Source code for hiclib.fragmentHiC</h1><div class="highlight"><pre>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This is a module class for fragment-level Hi-C data analysis.</span>
<span class="sd">The base class &quot;HiCdataset&quot; can load, save and merge Hi-C datasets, perform certain filters, and save binned heatmaps.   </span>

<span class="sd">Additional class HiCStatistics contains methods to analyze HiC data on a fragment level. </span>
<span class="sd">This includes read statistics, scalings, etc.</span>

<span class="sd">Input data</span>
<span class="sd">----------</span>

<span class="sd">When used together with iterative mapping, this class can load files from h5dicts created by iterative mapping. </span>

<span class="sd">This method can also input any dictionary-like structure, such as a dictionary, numpy.savez, etc. </span>
<span class="sd">The minimal subset of information are positions of two reads, but providing strand informations is adviced.</span>

<span class="sd">If restriction fragment assignment is not provided, it will be automatically recalculated.  </span>

<span class="sd">.. warning :: 1-bp difference in positions of restriction sites will force certain algorithms, such as scaling calculations, to throw an exception.    </span>
<span class="sd">    It is adviced to supply restriction site data only if it was generated by iterative mapping code. </span>
<span class="sd">    </span>

<span class="sd">Concepts</span>
<span class="sd">--------</span>

<span class="sd">All read data is stored in a synchronized h5dict-based dictionary of arrays. </span>
<span class="sd">Each variable has a fixed name and type, as specified in the self.vectors variable. </span>
<span class="sd">Whenever the variable is accessed from the program, it is loaded from the h5dict.</span>

<span class="sd">All fragment data is stored in RAM, and can be re-created at any point using </span>
<span class="sd">:py:func:`rebuildFragments &lt;HiCdataset.rebuildFragments&gt;`  </span>

<span class="sd">Whenever a set of reads needs to be excluded from the dataset, a </span>
<span class="sd">:py:func:`maskFilter  &lt;HiCdataset.maskFilter&gt;` method is called, that goes over all datasets and overrides them. </span>
<span class="sd">This method automatically rebuilds fragments.</span>

<span class="sd">Filtering the data</span>
<span class="sd">------------------</span>

<span class="sd">This class has many build-in methods for filtering the data. </span>
<span class="sd">However, one can easily construct another filter as presented in multiple one-liner examples below</span>

<span class="sd">.. code-block:: python</span>

<span class="sd">    &gt;&gt;&gt; Dset = HiCdataset(**kwargs) </span>
<span class="sd">    &gt;&gt;&gt; Dset.fragmentFilter((Dset.ufragmentlen &gt;1000) * (Dset.ufragmentlen &lt; 4000)) #filter reads from fragments between 1kb and 4kb long. </span>
<span class="sd">    &gt;&gt;&gt; Dset.maskFilter(Dset.DS)   #Keep only DS reads</span>
<span class="sd">    &gt;&gt;&gt; Dset.maskFilter(Dset.chrms1 == Dset.chrms2)  #keep only cis reads</span>
<span class="sd">    &gt;&gt;&gt; Dset.maskFilter((Dset.chrms1 !=14) + (Dset.chrms2 !=14))  #Exclude all reads from chromosome 14</span>
<span class="sd">    &gt;&gt;&gt; Dset.maskFilter(Dset.dist1 + Dset.dist2 &gt; 500)  #Keep only random breaks, if 500 is maximum molecule length</span>


<span class="sd">-------------------------------------------------------------------------------</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">warnings</span> 
<span class="kn">import</span> <span class="nn">os</span><span class="o">,</span><span class="nn">cPickle</span>
<span class="kn">from</span> <span class="nn">mirnylab.genome</span> <span class="kn">import</span> <span class="n">Genome</span> 
<span class="kn">import</span> <span class="nn">mirnylab.hic.mapping</span> 
<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">from</span> <span class="nn">numpy</span> <span class="kn">import</span> <span class="n">array</span> <span class="k">as</span> <span class="n">na</span>  
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">stats</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">math</span> <span class="kn">import</span> <span class="n">sqrt</span> 
<span class="kn">from</span> <span class="nn">mirnylab.h5dict</span> <span class="kn">import</span> <span class="n">h5dict</span> 
<span class="kn">from</span> <span class="nn">mirnylab</span> <span class="kn">import</span> <span class="n">plotting</span> 
<span class="kn">from</span> <span class="nn">mirnylab.plotting</span> <span class="kn">import</span> <span class="n">mat_img</span><span class="p">,</span><span class="n">removeAxes</span>

<span class="kn">from</span> <span class="nn">mirnylab</span> <span class="kn">import</span> <span class="n">numutils</span>  
<span class="kn">from</span> <span class="nn">mirnylab.numutils</span> <span class="kn">import</span> <span class="n">arrayInArray</span><span class="p">,</span>  <span class="n">sumByArray</span><span class="p">,</span> <span class="n">correct</span><span class="p">,</span> <span class="n">ultracorrect</span>

<span class="n">r_</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">r_</span>

<span class="k">def</span> <span class="nf">corr</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">):</span> <span class="k">return</span> <span class="n">stats</span><span class="o">.</span><span class="n">spearmanr</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>        


<div class="viewcode-block" id="HiCdataset"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset">[docs]</a><span class="k">class</span> <span class="nc">HiCdataset</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Base class to operate on HiC dataset. </span>
<span class="sd">    </span>
<span class="sd">    This class stores all information about HiC reads on a hard drive. </span>
<span class="sd">    Whenever a variable corresponding to any record is used, it is loaded/saved from/to the HDD. </span>
<span class="sd">    </span>
<span class="sd">    If you apply any filters to a dataset, it will actually modify the content of the current working file.  </span>
<span class="sd">    Thus, to preserve the data, loading datasets is advised. &quot;&quot;&quot;</span>
    
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span> <span class="p">,</span> <span class="n">genome</span> <span class="p">,</span> <span class="n">maximumMoleculeLength</span> <span class="o">=</span> <span class="mi">500</span> <span class="p">,</span> <span class="n">override</span> <span class="o">=</span> <span class="bp">True</span> <span class="p">,</span> <span class="n">autoFlush</span> <span class="o">=</span> <span class="bp">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;        </span>
<span class="sd">        __init__ method </span>
<span class="sd">        </span>
<span class="sd">        Initializes empty dataset by default. </span>
<span class="sd">        If &quot;override&quot; is False, works with existing dataset.   </span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : string </span>
<span class="sd">            A filename to store HiC dataset in an HDF5 file.  </span>
<span class="sd">        genome : folder with genome, or Genome object </span>
<span class="sd">            A folder with fastq files of the genome and gap table from Genome browser.</span>
<span class="sd">            Alternatively, mirnylab.genome.Genome object.               </span>
<span class="sd">        maximumMoleculeLength : int, optional </span>
<span class="sd">            Maximum length of molecules in the HiC library, used as a cutoff for dangling ends filter</span>
<span class="sd">        override : bool, optional</span>
<span class="sd">            Use specified dataset, do not remove it&#39;s contents. </span>
<span class="sd">            By default, if filename exists, it is deleted upon initialization.</span>
<span class="sd">        autoFlush : bool, optional</span>
<span class="sd">            Set to True to disable autoflush - possibly speeds up read/write operations. </span>
<span class="sd">            Don&#39;t forget to run flush then!             </span>
<span class="sd">        </span>
<span class="sd">        &quot;&quot;&quot;</span>                
        <span class="c">#----------&gt;&gt;&gt; Important::: do not define any variables before vectors!!! &lt;&lt;&lt;-------- </span>
        <span class="c">#These are fields that will be kept on a hard drive </span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vectors</span> <span class="o">=</span> <span class="p">{</span><span class="s">&quot;chrms1&quot;</span><span class="p">:</span><span class="s">&quot;int8&quot;</span><span class="p">,</span><span class="s">&quot;chrms2&quot;</span><span class="p">:</span><span class="s">&quot;int8&quot;</span><span class="p">,</span> <span class="c">#chromosomes. If &gt;chromosomeCount, then it&#39;s second chromosome arm! </span>
                        <span class="s">&quot;mids1&quot;</span><span class="p">:</span><span class="s">&quot;int32&quot;</span><span class="p">,</span><span class="s">&quot;mids2&quot;</span><span class="p">:</span><span class="s">&quot;int32&quot;</span><span class="p">,</span>  <span class="c">#midpoint of a fragment, determined as &quot;(start+end)/2&quot;</span>
                        <span class="s">&quot;fraglens1&quot;</span><span class="p">:</span><span class="s">&quot;int32&quot;</span><span class="p">,</span><span class="s">&quot;fraglens2&quot;</span><span class="p">:</span><span class="s">&quot;int32&quot;</span><span class="p">,</span> <span class="c">#fragment lengthes                        </span>
                        <span class="s">&quot;distances&quot;</span><span class="p">:</span><span class="s">&quot;int32&quot;</span><span class="p">,</span> <span class="c">#distance between fragments. If -1, different chromosomes. If -2, different arms.                         </span>
                        <span class="s">&quot;fragids1&quot;</span><span class="p">:</span><span class="s">&quot;int64&quot;</span><span class="p">,</span><span class="s">&quot;fragids2&quot;</span><span class="p">:</span><span class="s">&quot;int64&quot;</span><span class="p">,</span>  <span class="c">#IDs of fragments. fragIDmult * chromosome + location                          </span>
                        <span class="s">&quot;dists1&quot;</span><span class="p">:</span><span class="s">&quot;int32&quot;</span><span class="p">,</span><span class="s">&quot;dists2&quot;</span><span class="p">:</span><span class="s">&quot;int32&quot;</span><span class="p">,</span>        <span class="c">#distance to rsite</span>
                        <span class="s">&quot;cuts1&quot;</span><span class="p">:</span><span class="s">&quot;int32&quot;</span><span class="p">,</span><span class="s">&quot;cuts2&quot;</span><span class="p">:</span><span class="s">&quot;int32&quot;</span><span class="p">,</span>           <span class="c">#precise location of cut-site </span>
                        <span class="s">&quot;strands1&quot;</span><span class="p">:</span><span class="s">&quot;bool&quot;</span><span class="p">,</span><span class="s">&quot;strands2&quot;</span><span class="p">:</span><span class="s">&quot;bool&quot;</span><span class="p">,</span>
                        <span class="s">&quot;DS&quot;</span><span class="p">:</span><span class="s">&quot;bool&quot;</span><span class="p">,</span><span class="s">&quot;SS&quot;</span><span class="p">:</span><span class="s">&quot;bool&quot;</span><span class="p">}</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">autoFlush</span> <span class="o">=</span> <span class="n">autoFlush</span>
                       
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span> 
            <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">Genome</span><span class="p">(</span><span class="n">genomePath</span> <span class="o">=</span> <span class="n">genome</span><span class="p">,</span> <span class="n">readChrms</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;#&quot;</span><span class="p">,</span><span class="s">&quot;X&quot;</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">genome</span>             
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="n">Genome</span><span class="p">)</span>        
        
        <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmCount</span>  <span class="c">#used for building heatmaps</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">fragIDmult</span>
        <span class="k">print</span> <span class="s">&quot;----&gt; New dataset opened, genome </span><span class="si">%s</span><span class="s">, folder = </span><span class="si">%s</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">folderName</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">maximumMoleculeLength</span> <span class="o">=</span> <span class="n">maximumMoleculeLength</span>  <span class="c">#maximum length of a molecule for SS reads        </span>
        <span class="bp">self</span><span class="o">.</span><span class="n">filename</span> <span class="o">=</span> <span class="n">filename</span> <span class="c">#File to save the data                             </span>
                 
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filename</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">override</span> <span class="o">==</span> <span class="bp">False</span><span class="p">:</span> 
                <span class="k">print</span> <span class="s">&quot;-----&gt;!!!File already exists! It will be opened in the &#39;append&#39; mode.&quot;</span>  
                <span class="k">print</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filename</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filename</span><span class="p">)[</span><span class="mi">0</span><span class="p">]):</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;Folder in which you want to create file do not exist: </span><span class="si">%s</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filename</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">try</span><span class="p">:</span> 
                <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filename</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">except</span><span class="p">:</span> 
                <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s">&quot;Failed to create directory: </span><span class="si">%s</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filename</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
            
        <span class="bp">self</span><span class="o">.</span><span class="n">h5dict</span> <span class="o">=</span> <span class="n">h5dict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filename</span><span class="p">,</span><span class="n">autoflush</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">autoFlush</span> <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_setData</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">name</span><span class="p">,</span><span class="n">data</span><span class="p">):</span>
        <span class="s">&quot;an internal method to save numpy arrays to HDD quickly&quot;</span>
        <span class="k">if</span> <span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectors</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Attept to save data not specified in self.vectors&quot;</span><span class="p">)</span>
        <span class="n">dtype</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">dtype</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vectors</span><span class="p">[</span><span class="n">name</span><span class="p">])</span>         
        <span class="n">data</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">dtype</span><span class="o">=</span><span class="n">dtype</span><span class="p">)</span>        
        <span class="bp">self</span><span class="o">.</span><span class="n">h5dict</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span>

    
    <span class="k">def</span> <span class="nf">_getData</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">name</span><span class="p">):</span>
        <span class="s">&quot;an internal method to load numpy arrays from HDD quickly&quot;</span>         
        <span class="k">if</span> <span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectors</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Attept to load data not specified in self.vectors&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">h5dict</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>                
    
    <span class="k">def</span> <span class="nf">__getattribute__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">x</span><span class="p">):</span>
        <span class="s">&quot;a method that overrides set/get operation for self.vectors so that they&#39;re always on HDD&quot;</span>
        <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="s">&quot;vectors&quot;</span><span class="p">:</span> <span class="k">return</span> <span class="nb">object</span><span class="o">.</span><span class="n">__getattribute__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">x</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectors</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>            
            <span class="n">a</span> <span class="o">=</span>  <span class="bp">self</span><span class="o">.</span><span class="n">_getData</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>            
            <span class="k">return</span> <span class="n">a</span>                 
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">object</span><span class="o">.</span><span class="n">__getattribute__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">x</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__setattr__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">x</span><span class="p">,</span><span class="n">value</span><span class="p">):</span>
        <span class="s">&quot;a method that overrides set/get operation for self.vectors so that they&#39;re always on HDD&quot;</span>
        <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="s">&quot;vectors&quot;</span><span class="p">:</span> <span class="k">return</span> <span class="nb">object</span><span class="o">.</span><span class="n">__setattr__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">x</span><span class="p">,</span><span class="n">value</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectors</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>        
            <span class="bp">self</span><span class="o">.</span><span class="n">_setData</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">value</span><span class="p">)</span>                        
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">object</span><span class="o">.</span><span class="n">__setattr__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">x</span><span class="p">,</span><span class="n">value</span><span class="p">)</span>
        
    <span class="k">def</span> <span class="nf">_buildFragments</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="s">&quot;ufragments&quot;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">rebuildFragments</span><span class="p">()</span>
        
<div class="viewcode-block" id="HiCdataset.flush"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.flush">[docs]</a>    <span class="k">def</span> <span class="nf">flush</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="s">&quot;Flushes h5dict if used in autoFlush = False mode&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">h5dict</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
        
    </div>
<div class="viewcode-block" id="HiCdataset.merge"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.merge">[docs]</a>    <span class="k">def</span> <span class="nf">merge</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">filenames</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;combines data from multiple datasets</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">            filenames : list of strings</span>
<span class="sd">                List of folders to merge to current working folder                </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">filename</span> <span class="ow">in</span> <span class="n">filenames</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;----&gt; Cannot merge folder into itself! Create a new folder&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">filenames</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">Cannot open file: </span><span class="si">%s</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">filename</span><span class="p">)</span>        
            
        <span class="n">h5dicts</span> <span class="o">=</span> <span class="p">[</span><span class="n">h5dict</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="n">mode</span> <span class="o">=</span> <span class="s">&#39;r&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">filenames</span><span class="p">]</span>        
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectors</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">res</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">mydict</span> <span class="ow">in</span> <span class="n">h5dicts</span><span class="p">:</span>
                <span class="n">res</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mydict</span><span class="p">[</span><span class="n">name</span><span class="p">])</span>
            <span class="n">res</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_setData</span><span class="p">(</span><span class="n">name</span><span class="p">,</span><span class="n">res</span><span class="p">)</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">rebuildFragments</span><span class="p">()</span>

    </div>
<div class="viewcode-block" id="HiCdataset.parseInputData"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.parseInputData">[docs]</a>    <span class="k">def</span> <span class="nf">parseInputData</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">dictLike</span><span class="p">,</span><span class="n">zeroBaseChrom</span> <span class="o">=</span> <span class="bp">True</span> <span class="p">,</span><span class="n">enzymeToFillRsites</span> <span class="o">=</span> <span class="bp">None</span><span class="p">):</span>        
        <span class="sd">&quot;&quot;&quot;Inputs data from a dictionary-like object, containing coordinates of the reads. </span>
<span class="sd">        Performs filtering of the reads.   </span>
<span class="sd">        </span>
<span class="sd">        A good example of a dict-like object is a numpy.savez </span>
<span class="sd">        </span>
<span class="sd">        .. warning::  Restriction fragments MUST be specified exactly as in the Genome class.</span>
<span class="sd">        </span>
<span class="sd">        .. warning:: Strand information is needed for proper scaling calculations, but will be imitated if not provided</span>
<span class="sd">        </span>
<span class="sd">           </span>
<span class="sd">                    </span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        dictLike[&quot;chrms1,2&quot;] : Chromosomes of 2 sides of the read         </span>
<span class="sd">        dictLike[&quot;cuts1,2&quot;] : Exact position of cuts        </span>
<span class="sd">        dictLike[&quot;strands1,2&quot;], essential : Direction of the read        </span>
<span class="sd">        dictLike[&quot;rsites1,2&quot;], optional : Position of rsite to which the read is pointing                                 </span>
<span class="sd">        dictLike[&quot;uprsites1,2&quot;] , optional : rsite upstream (larger genomic coordinate) of the cut position        </span>
<span class="sd">        dictLike[&quot;downrsites1,2&quot;] , optional  : rsite downstream (smaller genomic coordinate) of the cut position</span>
<span class="sd">        </span>
<span class="sd">        zeroBaseChrom : bool , optional</span>
<span class="sd">            Use zero-base chromosome counting if True, one-base if False</span>
<span class="sd">        enzymeToFillRsites = None or str, optional if rsites are specified</span>
<span class="sd">            Enzyme name to use with Bio.restriction</span>
<span class="sd">                 </span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="n">rsite_related</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;rsites1&quot;</span><span class="p">,</span><span class="s">&quot;rsites2&quot;</span><span class="p">,</span><span class="s">&quot;uprsites1&quot;</span><span class="p">,</span><span class="s">&quot;uprsites2&quot;</span><span class="p">,</span><span class="s">&quot;downrsites1&quot;</span><span class="p">,</span><span class="s">&quot;downrsites2&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">dictLike</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span>            
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">dictLike</span><span class="p">):</span> <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s">&quot;File not found: </span><span class="si">%s</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">dictLike</span><span class="p">)</span>
            <span class="k">print</span> <span class="s">&quot;     loading data from file </span><span class="si">%s</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">dictLike</span>
            <span class="n">dictLike</span> <span class="o">=</span> <span class="n">h5dict</span><span class="p">(</span><span class="n">dictLike</span><span class="p">,</span><span class="s">&#39;r&#39;</span><span class="p">)</span> <span class="c">#attempting to open h5dict</span>
        
        <span class="k">if</span> <span class="bp">False</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="n">i</span> <span class="ow">in</span> <span class="n">dictLike</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">rsite_related</span><span class="p">]:</span>
            <span class="n">noRsites</span> <span class="o">=</span> <span class="bp">False</span>            
        <span class="k">else</span><span class="p">:</span>
            <span class="n">noRsites</span> <span class="o">=</span> <span class="bp">True</span>
        
        <span class="s">&quot;Filling in chromosomes and positions - mandatory objects&quot;</span>
        <span class="n">a</span> <span class="o">=</span> <span class="n">dictLike</span><span class="p">[</span><span class="s">&quot;chrms1&quot;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">trackLen</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>         
        
        <span class="k">if</span> <span class="n">zeroBaseChrom</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span> <span class="o">=</span> <span class="n">a</span>        
            <span class="bp">self</span><span class="o">.</span><span class="n">chrms2</span> <span class="o">=</span> <span class="n">dictLike</span><span class="p">[</span><span class="s">&quot;chrms2&quot;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span> <span class="o">=</span> <span class="n">a</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">chrms2</span> <span class="o">=</span> <span class="n">dictLike</span><span class="p">[</span><span class="s">&quot;chrms2&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span>
            
        <span class="bp">self</span><span class="o">.</span><span class="n">cuts1</span> <span class="o">=</span> <span class="n">dictLike</span><span class="p">[</span><span class="s">&quot;cuts1&quot;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cuts2</span> <span class="o">=</span> <span class="n">dictLike</span><span class="p">[</span><span class="s">&quot;cuts2&quot;</span><span class="p">]</span>    
        

        <span class="k">if</span> <span class="ow">not</span> <span class="p">((</span><span class="s">&quot;strands1&quot;</span> <span class="ow">in</span> <span class="n">dictLike</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="ow">and</span> <span class="p">(</span><span class="s">&quot;strands2&quot;</span> <span class="ow">in</span> <span class="n">dictLike</span><span class="o">.</span><span class="n">keys</span><span class="p">())):</span>
            <span class="k">print</span> <span class="s">&quot;No strand information provided, assigning random strands.&quot;</span>
            <span class="n">t</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">trackLen</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">strands1</span> <span class="o">=</span> <span class="n">t</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">strands2</span> <span class="o">=</span> <span class="mi">1</span><span class="o">-</span><span class="n">t</span>
            <span class="n">noStrand</span> <span class="o">=</span> <span class="bp">True</span> 
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">strands1</span> <span class="o">=</span> <span class="n">dictLike</span><span class="p">[</span><span class="s">&quot;strands1&quot;</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">strands2</span> <span class="o">=</span> <span class="n">dictLike</span><span class="p">[</span><span class="s">&quot;strands2&quot;</span><span class="p">]</span>            
            <span class="n">noStrand</span> <span class="o">=</span> <span class="bp">False</span>   <span class="c">#strand information filled in </span>
        
        <span class="k">if</span> <span class="n">noRsites</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>   <span class="c">#We have to fill rsites ousrlves. Let&#39;s see what enzyme to use! </span>
            <span class="k">if</span> <span class="p">(</span><span class="n">enzymeToFillRsites</span> <span class="o">==</span> <span class="bp">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">hasEnzyme</span><span class="p">()</span> <span class="o">==</span> <span class="bp">False</span><span class="p">)</span> <span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Please specify enzyme if your data has no rsites&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">enzymeToFillRsites</span> <span class="o">!=</span> <span class="bp">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">hasEnzyme</span><span class="p">()</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">enzymeToFillRsites</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">enzymeName</span><span class="p">:</span>
                        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;genome.enzymeName different from supplied enzyme&quot;</span><span class="p">)</span>                                         
                <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">setEnzyme</span><span class="p">(</span><span class="n">enzymeToFillRsites</span><span class="p">)</span>
            <span class="c">#enzymeToFillRsites has preference over self.genome&#39;s enzyme</span>
                
            <span class="k">print</span> <span class="s">&quot;Filling rsites&quot;</span>            
            <span class="n">rsitedict</span> <span class="o">=</span> <span class="n">h5dict</span><span class="p">()</span>  <span class="c">#creating dict to pass to anton&#39;s code </span>
            <span class="n">rsitedict</span><span class="p">[</span><span class="s">&quot;chrms1&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span>
            <span class="n">rsitedict</span><span class="p">[</span><span class="s">&quot;chrms2&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chrms2</span>
            <span class="n">rsitedict</span><span class="p">[</span><span class="s">&quot;cuts1&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cuts1</span>
            <span class="n">rsitedict</span><span class="p">[</span><span class="s">&quot;cuts2&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cuts2</span>            
            <span class="n">rsitedict</span><span class="p">[</span><span class="s">&quot;strands1&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">strands1</span>
            <span class="n">rsitedict</span><span class="p">[</span><span class="s">&quot;strands2&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">strands2</span>            
            <span class="n">mirnylab</span><span class="o">.</span><span class="n">hic</span><span class="o">.</span><span class="n">mapping</span><span class="o">.</span><span class="n">fill_rsites</span><span class="p">(</span><span class="n">lib</span> <span class="o">=</span> <span class="n">rsitedict</span><span class="p">,</span> <span class="n">genome_db</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">rsitedict</span> <span class="o">=</span> <span class="n">dictLike</span> <span class="c">#rsite information is in our dictionary        </span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">DS</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms2</span> <span class="o">&gt;=</span><span class="mi">0</span><span class="p">)</span>   <span class="c">#if we have reads from both chromosomes, we&#39;re a DS read</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">SS</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DS</span> <span class="o">==</span> <span class="bp">False</span><span class="p">)</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">dists1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">rsitedict</span><span class="p">[</span><span class="s">&quot;rsites1&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">cuts1</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">dists2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">rsitedict</span><span class="p">[</span><span class="s">&quot;rsites2&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">cuts2</span><span class="p">)</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">mids1</span> <span class="o">=</span> <span class="p">(</span><span class="n">rsitedict</span><span class="p">[</span><span class="s">&quot;uprsites1&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">rsitedict</span><span class="p">[</span><span class="s">&quot;downrsites1&quot;</span><span class="p">])</span><span class="o">/</span><span class="mi">2</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mids2</span> <span class="o">=</span> <span class="p">(</span><span class="n">rsitedict</span><span class="p">[</span><span class="s">&quot;uprsites2&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">rsitedict</span><span class="p">[</span><span class="s">&quot;downrsites2&quot;</span><span class="p">])</span><span class="o">/</span><span class="mi">2</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">fraglens1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">abs</span><span class="p">((</span><span class="n">rsitedict</span><span class="p">[</span><span class="s">&quot;uprsites1&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">rsitedict</span><span class="p">[</span><span class="s">&quot;downrsites1&quot;</span><span class="p">]))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fraglens2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">abs</span><span class="p">((</span><span class="n">rsitedict</span><span class="p">[</span><span class="s">&quot;uprsites2&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">rsitedict</span><span class="p">[</span><span class="s">&quot;downrsites2&quot;</span><span class="p">]))</span>
        
        <span class="k">del</span> <span class="n">rsitedict</span>   <span class="c">#deletes hdf5 file, so it&#39;s important </span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">fragids1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mids1</span> <span class="o">+</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span><span class="p">,</span><span class="n">dtype</span> <span class="o">=</span> <span class="s">&quot;int64&quot;</span><span class="p">)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fragids2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mids2</span> <span class="o">+</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms2</span><span class="p">,</span><span class="n">dtype</span> <span class="o">=</span> <span class="s">&quot;int64&quot;</span><span class="p">)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span> 
        
        <span class="n">distances</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mids1</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">mids2</span><span class="p">)</span>
        <span class="n">distances</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chrms2</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">distances</span> <span class="o">=</span> <span class="n">distances</span>   <span class="c">#distances between restriction fragments</span>
        
        <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span>  <span class="c">#moving SS reads to the first side</span>
        <span class="n">mask</span> <span class="c">#Eclipse warning removal </span>
        <span class="n">variables</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">i</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectors</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s">&quot;1&quot;</span><span class="p">])</span>  <span class="c">#set of variables to change</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">variables</span><span class="p">:</span>  
            <span class="k">exec</span><span class="p">(</span><span class="s">&quot;a = self.</span><span class="si">%s</span><span class="s">1&quot;</span> <span class="o">%</span> <span class="n">i</span><span class="p">)</span>
            <span class="k">exec</span><span class="p">(</span><span class="s">&quot;b = self.</span><span class="si">%s</span><span class="s">2&quot;</span> <span class="o">%</span> <span class="n">i</span><span class="p">)</span>
            <span class="k">exec</span><span class="p">(</span><span class="s">&quot;a[mask] = b[mask]&quot;</span><span class="p">)</span>
            <span class="k">exec</span><span class="p">(</span><span class="s">&quot;b[mask] = -1&quot;</span><span class="p">)</span>
            <span class="k">exec</span><span class="p">(</span><span class="s">&quot;self.</span><span class="si">%s</span><span class="s">1 = a&quot;</span> <span class="o">%</span> <span class="n">i</span><span class="p">)</span>
            <span class="k">exec</span><span class="p">(</span><span class="s">&quot;self.</span><span class="si">%s</span><span class="s">2 = b&quot;</span> <span class="o">%</span> <span class="n">i</span><span class="p">)</span>
                
        <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragids1</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragids2</span><span class="p">)</span>   <span class="c">#Discard dangling ends and self-circles</span>
        <span class="n">maskLen</span><span class="p">,</span> <span class="n">noSameFrag</span>  <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">mask</span><span class="p">),</span>  <span class="n">mask</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>                         
        <span class="n">mask</span> <span class="o">*=</span>  <span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms2</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomeCount</span><span class="p">))</span>  <span class="c">#Discard unused chromosomes</span>
        <span class="n">noUnusedChroms</span> <span class="o">=</span>  <span class="n">mask</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>                  
        <span class="n">mask</span> <span class="o">*=</span> <span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span> <span class="o">&gt;=</span><span class="mi">0</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms2</span> <span class="o">&gt;=</span><span class="mi">0</span><span class="p">))</span>   <span class="c">#Has to have at least one side mapped</span>
        <span class="n">noUnmapped</span> <span class="o">=</span> <span class="n">mask</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>                 
        <span class="k">if</span> <span class="n">noStrand</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span> 
            <span class="n">dist</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">cuts1</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">cuts2</span><span class="p">)</span>  <span class="c">#Can&#39;t tell if reads point to each other. </span>
        <span class="k">else</span><span class="p">:</span> 
            <span class="n">dist</span> <span class="o">=</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">cuts1</span> <span class="o">*</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">strands1</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">cuts2</span> <span class="o">*</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">strands2</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>  <span class="c">#distance between sites facing each other</span>
                                
                                
        <span class="n">readsMolecules</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">chrms2</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">strands1</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">strands2</span><span class="p">)</span> <span class="o">*</span>  <span class="p">(</span><span class="n">dist</span> <span class="o">&gt;=</span><span class="mi">0</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">dist</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">maximumMoleculeLength</span><span class="p">)</span><span class="c">#filtering out DE</span>
                 
        <span class="n">mask</span> <span class="o">*=</span> <span class="p">(</span><span class="n">readsMolecules</span>  <span class="o">==</span> <span class="bp">False</span><span class="p">)</span>
        <span class="n">extraDE</span> <span class="o">=</span> <span class="n">mask</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>                                 
        
        
                        
        <span class="k">print</span> <span class="s">&quot;     Original reads: {maskLen}  -&gt; No same fragment: {noSameFrag} -&gt; remove unused chrom: {noUnusedChroms} -&gt; ...&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="o">**</span><span class="nb">locals</span><span class="p">())</span>
        <span class="k">print</span> <span class="s">&quot;     ... -&gt; No unmapped reads: {noUnmapped} -&gt; no extra DEs (--&gt; (&lt;500) &lt;--): {extraDE}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="o">**</span><span class="nb">locals</span><span class="p">())</span>        
         
        <span class="bp">self</span><span class="o">.</span><span class="n">maskFilter</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span>
        
            </div>
<div class="viewcode-block" id="HiCdataset.saveFragments"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.saveFragments">[docs]</a>    <span class="k">def</span> <span class="nf">saveFragments</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="s">&quot;saves fragment data to make correct expected estimates after applying a heavy mask&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ufragmentsOriginal</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ufragmentlenOriginal</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragmentlen</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="HiCdataset.originalFragments"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.originalFragments">[docs]</a>    <span class="k">def</span> <span class="nf">originalFragments</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="s">&quot;loads original fragments&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragmentsOriginal</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ufragmentlen</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragmentlenOriginal</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="HiCdataset.calculateWeights"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.calculateWeights">[docs]</a>    <span class="k">def</span> <span class="nf">calculateWeights</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Calculates weights for reads based on fragment length correction similar to Tanay&#39;s;</span>
<span class="sd">         may be used for scalings or creating heatmaps&quot;&quot;&quot;</span>        
        <span class="n">fragmentLength</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ufragmentlen</span>
        <span class="n">pls</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">fragmentLength</span><span class="p">)</span>
        <span class="n">pls</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">r_</span><span class="p">[</span><span class="n">pls</span><span class="p">,</span><span class="n">pls</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">fragmentLength</span><span class="p">)</span>    
        <span class="n">mysum</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragmentSum</span><span class="p">(),</span><span class="nb">float</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">weights</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">N</span><span class="p">,</span><span class="nb">float</span><span class="p">)</span>
        <span class="n">meanSum</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mysum</span><span class="p">,</span><span class="nb">float</span><span class="p">))</span>
        <span class="c">#watch = numpy.zeros(len(mysum),int)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.991</span><span class="p">,</span><span class="mf">0.01</span><span class="p">):</span>
            <span class="n">b1</span><span class="p">,</span><span class="n">b2</span> <span class="o">=</span> <span class="n">pls</span><span class="p">[</span><span class="n">i</span><span class="o">*</span><span class="n">N</span><span class="p">],</span><span class="n">pls</span><span class="p">[(</span><span class="n">i</span><span class="o">+</span><span class="mf">0.01</span><span class="p">)</span><span class="o">*</span><span class="n">N</span><span class="p">]</span>
            <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">b1</span> <span class="o">&lt;=</span> <span class="n">fragmentLength</span><span class="p">)</span><span class="o">*</span> <span class="p">(</span><span class="n">b2</span> <span class="o">&gt;</span> <span class="n">fragmentLength</span><span class="p">)</span>
            <span class="c">#watch[p] += 1                        </span>
            <span class="n">value</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">mysum</span><span class="p">[</span><span class="n">p</span><span class="p">])</span>      
            <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">weights</span><span class="p">[</span><span class="n">p</span><span class="p">]</span> <span class="o">=</span>  <span class="n">value</span> <span class="o">/</span> <span class="n">meanSum</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">print</span> <span class="s">&quot;no weights&quot;</span><span class="p">,</span><span class="n">i</span><span class="p">,</span><span class="n">b1</span><span class="p">,</span><span class="n">b2</span>


</div>
<div class="viewcode-block" id="HiCdataset.buildHeatmap"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.buildHeatmap">[docs]</a>    <span class="k">def</span> <span class="nf">buildHeatmap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">chromosome</span> <span class="o">=</span> <span class="mi">14</span><span class="p">,</span><span class="n">chromosome2</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span><span class="n">resolution</span> <span class="o">=</span> <span class="mi">1000000</span><span class="p">,</span><span class="n">show</span> <span class="o">=</span> <span class="bp">False</span><span class="p">,</span><span class="n">useRsiteDensity</span> <span class="o">=</span> <span class="bp">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Builds heatmaps between any two chromosomes at a given resolution</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        chromosome1 : int</span>
<span class="sd">            First chromosome</span>
<span class="sd">        chromosome2 : int, optional </span>
<span class="sd">            Second chromosome, if the map is trans. </span>
<span class="sd">        resolution : int</span>
<span class="sd">            Resolution of a heatmap. Default is 1M  </span>
<span class="sd">        show : bool , optional</span>
<span class="sd">            Show heatmap, or just output it?</span>
<span class="sd">        useRsiteDensity : bool, optional</span>
<span class="sd">            Correct map by density of rsites.              </span>
<span class="sd">        </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_buildFragments</span><span class="p">()</span>        
        <span class="k">if</span> <span class="n">chromosome2</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span> 
            <span class="n">chromosome2</span> <span class="o">=</span> <span class="n">chromosome</span>                        
            <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span> <span class="o">==</span> <span class="n">chromosome</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms2</span> <span class="o">==</span> <span class="n">chromosome</span><span class="p">)</span>
            <span class="n">p1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mids1</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
            <span class="n">p2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mids2</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
            <span class="n">p1</span> <span class="o">=</span> <span class="n">na</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span><span class="nb">int</span><span class="p">)</span>
            <span class="n">p2</span> <span class="o">=</span> <span class="n">na</span><span class="p">(</span><span class="n">p2</span><span class="p">,</span><span class="nb">int</span><span class="p">)</span>
            <span class="n">b1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">max</span><span class="p">(</span><span class="n">p1</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">+</span><span class="n">resolution</span><span class="p">,</span><span class="n">p2</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">+</span><span class="n">resolution</span><span class="p">),</span><span class="n">resolution</span><span class="p">)</span>            
            <span class="n">hist</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">histogram2d</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span><span class="n">p2</span><span class="p">,(</span><span class="n">b1</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">hist</span> <span class="o">=</span> <span class="n">hist</span> <span class="o">+</span> <span class="n">numpy</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="n">hist</span><span class="p">)</span>
            
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span> <span class="o">==</span> <span class="n">chromosome</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms2</span> <span class="o">==</span> <span class="n">chromosome2</span><span class="p">)</span>
            <span class="n">p11</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mids1</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
            <span class="n">p21</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mids2</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>            
            <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span> <span class="o">==</span> <span class="n">chromosome2</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms2</span> <span class="o">==</span> <span class="n">chromosome</span><span class="p">)</span>
            <span class="n">p12</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mids2</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
            <span class="n">p22</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mids1</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>            
            <span class="n">p1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">r_</span><span class="p">[</span><span class="n">p11</span><span class="p">,</span><span class="n">p12</span><span class="p">]</span>
            <span class="n">p2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">r_</span><span class="p">[</span><span class="n">p21</span><span class="p">,</span><span class="n">p22</span><span class="p">]</span>            
            <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">p1</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">p2</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>                 
                <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">500</span><span class="p">,</span><span class="mi">500</span><span class="p">),</span><span class="nb">float</span><span class="p">)</span> <span class="o">+</span> <span class="mf">0.000001</span>
            <span class="n">b1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,(</span><span class="n">p1</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">+</span><span class="n">resolution</span><span class="p">),</span><span class="n">resolution</span><span class="p">)</span>            
            <span class="n">b2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,(</span><span class="n">p2</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">+</span><span class="n">resolution</span><span class="p">),</span><span class="n">resolution</span><span class="p">)</span>
            <span class="n">l1</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">b1</span><span class="p">)</span>
            <span class="n">l2</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">b2</span><span class="p">)</span>
            <span class="n">hist</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">histogram2d</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span><span class="n">p2</span><span class="p">,(</span><span class="n">b1</span><span class="p">,</span><span class="n">b2</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">chromosome2</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span> <span class="n">chromosome2</span> <span class="o">=</span> <span class="n">chromosome</span>
        <span class="k">if</span> <span class="n">useRsiteDensity</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>            
            <span class="n">m1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span> <span class="o">==</span> <span class="n">chromosome</span>
            <span class="n">m2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span> <span class="o">==</span> <span class="n">chromosome2</span>
            <span class="n">p1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span><span class="p">[</span><span class="n">m1</span><span class="p">]</span>  <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span> 
            <span class="n">p2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span><span class="p">[</span><span class="n">m2</span><span class="p">]</span>  <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span>
            <span class="n">p1mod</span> <span class="o">=</span> <span class="n">p1</span> <span class="o">/</span> <span class="n">resolution</span>
            <span class="n">p2mod</span> <span class="o">=</span> <span class="n">p2</span> <span class="o">/</span> <span class="n">resolution</span>            
            <span class="n">myarray</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">p1mod</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
            <span class="n">vec1</span> <span class="o">=</span>  <span class="n">sumByArray</span><span class="p">(</span><span class="n">p1mod</span><span class="p">,</span><span class="n">myarray</span><span class="p">)</span>
            <span class="n">myarray</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">p2mod</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
            <span class="n">vec2</span> <span class="o">=</span> <span class="n">sumByArray</span><span class="p">(</span><span class="n">p2mod</span><span class="p">,</span><span class="n">myarray</span><span class="p">)</span>
            <span class="n">vec1</span> <span class="o">=</span> <span class="n">vec1</span><span class="p">[:</span><span class="n">l1</span><span class="p">]</span>
            <span class="n">vec2</span> <span class="o">=</span> <span class="n">vec2</span><span class="p">[:</span><span class="n">l2</span><span class="p">]</span>
            <span class="n">correction</span> <span class="o">=</span> <span class="n">na</span><span class="p">(</span><span class="n">vec1</span><span class="p">[:,</span><span class="bp">None</span><span class="p">]</span> <span class="o">*</span> <span class="n">vec2</span><span class="p">[</span><span class="bp">None</span><span class="p">,:],</span><span class="nb">float</span><span class="p">)</span>
            <span class="n">mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">logical_or</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">correction</span><span class="p">),</span> <span class="n">correction</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> 
            <span class="n">correction</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span> 
            <span class="n">hist2</span> <span class="o">=</span> <span class="n">hist</span> <span class="o">/</span> <span class="n">correction</span>
            <span class="n">hist2</span> <span class="o">/=</span> <span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">hist2</span><span class="p">[</span><span class="n">mask</span><span class="o">==</span><span class="bp">False</span><span class="p">])</span> <span class="o">/</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">hist</span><span class="p">[</span><span class="n">mask</span><span class="o">==</span><span class="bp">False</span><span class="p">]))</span>
            <span class="k">if</span> <span class="n">show</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span> <span class="n">mat_img</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">hist2</span><span class="p">,</span><span class="nb">float</span><span class="p">)</span> <span class="p">))</span>
            <span class="k">else</span><span class="p">:</span> <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">hist2</span> <span class="p">)</span>
        <span class="k">if</span> <span class="n">show</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span> <span class="n">mat_img</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">hist</span><span class="p">,</span><span class="nb">float</span><span class="p">)</span> <span class="p">))</span>
        <span class="k">else</span><span class="p">:</span> <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">hist</span><span class="p">)</span>
    </div>
<div class="viewcode-block" id="HiCdataset.buildAllHeatmap"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.buildAllHeatmap">[docs]</a>    <span class="k">def</span> <span class="nf">buildAllHeatmap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">resolution</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates an all-by-all heatmap in accordance with mapping provided by &#39;genome&#39; class</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        Resolution : int</span>
<span class="sd">            Resolution of a heatmap </span>
<span class="sd">        &quot;&quot;&quot;</span> 
        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">setResolution</span><span class="p">(</span><span class="n">resolution</span><span class="p">)</span>
        <span class="n">dr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DS</span> 
        
        <span class="n">label1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmStartsBinCont</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span><span class="p">[</span><span class="n">dr</span><span class="p">]</span> <span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">mids1</span><span class="p">[</span><span class="n">dr</span><span class="p">]</span> <span class="o">/</span> <span class="n">resolution</span>
        <span class="n">label1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">label1</span><span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="s">&quot;uint32&quot;</span><span class="p">)</span>
        <span class="n">label2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmStartsBinCont</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms2</span><span class="p">[</span><span class="n">dr</span><span class="p">]</span> <span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">mids2</span><span class="p">[</span><span class="n">dr</span><span class="p">]</span> <span class="o">/</span> <span class="n">resolution</span>
        <span class="n">label2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">label2</span><span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="s">&quot;uint32&quot;</span><span class="p">)</span>       
        <span class="n">label</span> <span class="o">=</span> <span class="n">label1</span> <span class="o">*</span> <span class="n">numpy</span><span class="o">.</span><span class="n">int64</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span><span class="p">)</span> <span class="o">+</span> <span class="n">label2</span>
        <span class="k">del</span> <span class="n">label1</span>
        <span class="k">del</span> <span class="n">label2</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span> <span class="o">&lt;</span> <span class="mi">65000</span><span class="p">:</span>
            <span class="n">label</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="s">&quot;uint32&quot;</span><span class="p">)</span>                    
         
        <span class="n">counts</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">minlength</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span><span class="o">**</span><span class="mi">2</span><span class="p">:</span>
            <span class="k">print</span> <span class="s">&quot;heatmap exceed length of the genome!!! Check genome&quot;</span>
            <span class="nb">exit</span><span class="p">()</span>
            
        <span class="n">counts</span><span class="o">.</span><span class="n">shape</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">counts</span><span class="p">)):</span>
            <span class="n">counts</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="n">i</span><span class="p">:]</span> <span class="o">+=</span> <span class="n">counts</span><span class="p">[</span><span class="n">i</span><span class="p">:,</span><span class="n">i</span><span class="p">]</span>
            <span class="n">counts</span><span class="p">[</span><span class="n">i</span><span class="p">:,</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">counts</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="n">i</span><span class="p">:]</span>        
        <span class="k">return</span> <span class="n">counts</span> 
    </div>
<div class="viewcode-block" id="HiCdataset.buildSinglesCoverage"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.buildSinglesCoverage">[docs]</a>    <span class="k">def</span> <span class="nf">buildSinglesCoverage</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">resolution</span><span class="p">):</span>
        <span class="s">&quot;creates an SS coverage vector heatmap in accordance with the output of the &#39;genome&#39; class&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">setResolution</span><span class="p">(</span><span class="n">resolution</span><span class="p">)</span>
        <span class="n">ds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DS</span> <span class="o">==</span> <span class="bp">False</span>        
        <span class="n">label</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmStartsBinCont</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span><span class="p">[</span><span class="n">ds</span><span class="p">]</span> <span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">mids1</span><span class="p">[</span><span class="n">ds</span><span class="p">]</span> <span class="o">/</span> <span class="n">resolution</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="n">sumByArray</span><span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">counts</span>
    </div>
<div class="viewcode-block" id="HiCdataset.buildFragmetCoverage"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.buildFragmetCoverage">[docs]</a>    <span class="k">def</span> <span class="nf">buildFragmetCoverage</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">resolution</span><span class="p">):</span>
        <span class="s">&quot;creates restriction site density vector (visible sites only) in accordance with the &#39;genome&#39; class&quot;</span>         
        <span class="bp">self</span><span class="o">.</span><span class="n">_buildFragments</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">setResolution</span><span class="p">(</span><span class="n">resolution</span><span class="p">)</span>
        <span class="n">chroms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span>
        <span class="n">label</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmStartsBinCont</span><span class="p">[</span><span class="n">chroms</span> <span class="p">]</span> <span class="o">+</span> <span class="n">positions</span> <span class="o">/</span> <span class="n">resolution</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="n">sumByArray</span><span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">counts</span>
        

    
        
    </div>
<div class="viewcode-block" id="HiCdataset.fragmentFilter"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.fragmentFilter">[docs]</a>    <span class="k">def</span> <span class="nf">fragmentFilter</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">fragments</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;keeps only reads that originate from fragments in &#39;fragments&#39; variable, for DS - on both sides</span>
<span class="sd">        </span>
<span class="sd">        Parameters </span>
<span class="sd">        ----------</span>
<span class="sd">        fragments : numpy.array of fragment IDs or bools        </span>
<span class="sd">            List of fragments to keep, or their indexes in self.ufragments        </span>
<span class="sd">        &quot;&quot;&quot;</span>        
        <span class="k">if</span> <span class="n">fragments</span><span class="o">.</span><span class="n">dtype</span> <span class="o">==</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bool</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_buildFragments</span><span class="p">()</span>
            <span class="n">fragments</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span><span class="p">[</span><span class="n">fragments</span><span class="p">]</span>        
        <span class="n">m1</span> <span class="o">=</span> <span class="n">arrayInArray</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragids1</span><span class="p">,</span><span class="n">fragments</span><span class="p">)</span>
        <span class="n">m2</span> <span class="o">=</span> <span class="n">arrayInArray</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragids2</span><span class="p">,</span><span class="n">fragments</span><span class="p">)</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">SS</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span><span class="n">m1</span><span class="p">,</span><span class="n">m2</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">maskFilter</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span>

    </div>
<div class="viewcode-block" id="HiCdataset.maskFilter"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.maskFilter">[docs]</a>    <span class="k">def</span> <span class="nf">maskFilter</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">mask</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;keeps only reads designated by mask</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        mask : array of bools</span>
<span class="sd">            Indexes of reads to keep </span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="k">print</span> <span class="s">&quot;          Number of reads changed  </span><span class="si">%d</span><span class="s"> ---&gt; </span><span class="si">%d</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">mask</span><span class="p">),</span><span class="n">mask</span><span class="o">.</span><span class="n">sum</span><span class="p">()),</span>
        <span class="n">length</span> <span class="o">=</span> <span class="mi">0</span> 
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectors</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_getData</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
            <span class="n">ld</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">length</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> 
                <span class="n">length</span> <span class="o">=</span> <span class="n">ld</span>
                
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ld</span> <span class="o">!=</span> <span class="n">length</span><span class="p">:</span> 
                    <span class="bp">self</span><span class="o">.</span><span class="n">delete</span><span class="p">()</span>             
            <span class="bp">self</span><span class="o">.</span><span class="n">_setData</span><span class="p">(</span><span class="n">name</span><span class="p">,</span><span class="n">data</span><span class="p">[</span><span class="n">mask</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">N</span> <span class="o">=</span> <span class="n">mask</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>   
        <span class="bp">self</span><span class="o">.</span><span class="n">rebuildFragments</span><span class="p">()</span>
        </div>
<div class="viewcode-block" id="HiCdataset.rebuildFragments"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.rebuildFragments">[docs]</a>    <span class="k">def</span> <span class="nf">rebuildFragments</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="s">&quot;recreates a set of fragments - runs when reads have changed&quot;</span>
        <span class="k">try</span><span class="p">:</span> 
            <span class="n">past</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span><span class="p">)</span>        
        <span class="k">except</span><span class="p">:</span>
            <span class="n">past</span> <span class="o">=</span> <span class="mi">0</span>        
        <span class="bp">self</span><span class="o">.</span><span class="n">ufragids1</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragids1ind</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragids1</span><span class="p">,</span><span class="n">return_index</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ufragids2</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragids2ind</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragids2</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">DS</span><span class="p">],</span><span class="n">return_index</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
                
        <span class="c">#Funding unique fragments and unique fragment IDs</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ufragment1len</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fraglens1</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragids1ind</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ufragment2len</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fraglens2</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">DS</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragids2ind</span><span class="p">]</span>
 
        <span class="n">uall</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">r_</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragids1</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragids2</span><span class="p">]</span>
        <span class="n">ulen</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">r_</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragment1len</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragment2len</span><span class="p">]</span>
         
        <span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span><span class="p">,</span><span class="n">ind</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">uall</span><span class="p">,</span> <span class="bp">True</span><span class="p">)</span>        
        <span class="bp">self</span><span class="o">.</span><span class="n">ufragmentlen</span> <span class="o">=</span> <span class="n">ulen</span><span class="p">[</span><span class="n">ind</span><span class="p">]</span>            
        <span class="k">print</span> <span class="s">&quot;          Fragments number changed -   </span><span class="si">%d</span><span class="s"> ---&gt;  </span><span class="si">%d</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">past</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span><span class="p">))</span>
        
            
</div>
<div class="viewcode-block" id="HiCdataset.filterExtreme"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.filterExtreme">[docs]</a>    <span class="k">def</span> <span class="nf">filterExtreme</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">cutH</span> <span class="o">=</span> <span class="mf">0.005</span><span class="p">,</span><span class="n">cutL</span> <span class="o">=</span> <span class="mi">0</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;removes fragments with most and/or least # counts</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        cutH : float, 0&lt;=cutH &lt; 1, optional</span>
<span class="sd">            Fraction of the most-counts fragments to be removed</span>
<span class="sd">        cutL : float, 0&lt;=cutL&lt;1, optional </span>
<span class="sd">            Fraction of the least-counts fragments to be removed</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_buildFragments</span><span class="p">()</span>
        <span class="k">print</span> <span class="s">&quot;-----&gt;Extreme fragments filter: remove top </span><span class="si">%lf</span><span class="s">, bottom </span><span class="si">%lf</span><span class="s"> fragments&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">cutH</span><span class="p">,</span> <span class="n">cutL</span><span class="p">)</span>
        <span class="n">s</span>  <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragmentSum</span><span class="p">()</span>
        <span class="n">ss</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
                         
        <span class="n">valueL</span><span class="p">,</span> <span class="n">valueH</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">ss</span><span class="p">,</span> <span class="p">[</span><span class="mf">100.</span> <span class="o">*</span> <span class="n">cutL</span><span class="p">,</span> <span class="mi">100</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.</span> <span class="o">-</span> <span class="n">cutH</span><span class="p">)])</span>
        <span class="n">news</span> <span class="o">=</span> <span class="p">(</span><span class="n">s</span> <span class="o">&gt;=</span> <span class="n">valueL</span> <span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">s</span> <span class="o">&lt;=</span> <span class="n">valueH</span><span class="p">)</span>        
        <span class="bp">self</span><span class="o">.</span><span class="n">fragmentFilter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span><span class="p">[</span><span class="n">news</span><span class="p">])</span>
        
        <span class="k">print</span> <span class="s">&quot;     #Top fragments are: &quot;</span><span class="p">,</span><span class="n">ss</span><span class="p">[</span><span class="o">-</span><span class="mi">10</span><span class="p">:]</span>
        <span class="k">print</span> <span class="s">&quot;     # Cutoff for low # counts is (counts): &quot;</span><span class="p">,</span><span class="n">valueL</span> <span class="p">,</span><span class="s">&quot;; cutoff for large # counts is: &quot;</span><span class="p">,</span><span class="n">valueH</span><span class="p">,</span> <span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span>         
        
        </div>
<div class="viewcode-block" id="HiCdataset.filterLarge"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.filterLarge">[docs]</a>    <span class="k">def</span> <span class="nf">filterLarge</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">cutlarge</span> <span class="o">=</span> <span class="mi">100000</span><span class="p">,</span><span class="n">cutsmall</span> <span class="o">=</span> <span class="mi">100</span> <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;removes very large and small fragments</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        cutlarge : int </span>
<span class="sd">            remove fragments larger than it</span>
<span class="sd">        cutsmall : int</span>
<span class="sd">            remove fragments smaller than it</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_buildFragments</span><span class="p">()</span>
        <span class="k">print</span> <span class="s">&quot;-----&gt;Small/large fragments filter: keep strictly less than </span><span class="si">%d</span><span class="s">, strictly more than </span><span class="si">%d</span><span class="s"> bp&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">cutlarge</span><span class="p">,</span> <span class="n">cutsmall</span><span class="p">)</span>                
        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragmentlen</span> <span class="o">&lt;</span> <span class="p">(</span><span class="n">cutlarge</span> <span class="p">)</span> <span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragmentlen</span> <span class="o">&gt;</span> <span class="n">cutsmall</span><span class="p">)</span>                      
        <span class="bp">self</span><span class="o">.</span><span class="n">fragmentFilter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span><span class="p">[</span><span class="n">p</span><span class="p">])</span>
        <span class="k">print</span> 

    </div>
<div class="viewcode-block" id="HiCdataset.filterRsiteStart"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.filterRsiteStart">[docs]</a>    <span class="k">def</span> <span class="nf">filterRsiteStart</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">offset</span> <span class="o">=</span> <span class="mi">5</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;removes reads that start within x bp near rsite</span>
<span class="sd">        TODO: fix this so that it agrees with the definition. </span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        offset : int </span>
<span class="sd">            Number of bp to exclude next to rsite, not including offset</span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="k">print</span> <span class="s">&quot;-----&gt;Semi-dangling end filter: remove guys who start </span><span class="si">%d</span><span class="s"> bp near the rsite&quot;</span> <span class="o">%</span> <span class="n">offset</span>
                        
        <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dists1</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">fraglens1</span><span class="p">)</span> <span class="o">&gt;=</span><span class="n">offset</span><span class="p">)</span> <span class="o">*</span> <span class="p">((</span><span class="n">numpy</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dists2</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">fraglens2</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">offset</span> <span class="p">)</span><span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">DS</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">SS</span><span class="p">)</span>    
                 
        <span class="bp">self</span><span class="o">.</span><span class="n">maskFilter</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span>
        <span class="k">print</span>
        </div>
<div class="viewcode-block" id="HiCdataset.filterDuplicates"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.filterDuplicates">[docs]</a>    <span class="k">def</span> <span class="nf">filterDuplicates</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="s">&quot;removes duplicate molecules in DS reads&quot;</span>
        <span class="k">print</span> <span class="s">&quot;-----&gt;Filtering duplicates in DS reads: &quot;</span>
        
        <span class="n">dups</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">N</span><span class="p">,</span><span class="mi">2</span><span class="p">),</span><span class="n">dtype</span> <span class="o">=</span> <span class="s">&quot;int64&quot;</span><span class="p">,</span><span class="n">order</span> <span class="o">=</span> <span class="s">&quot;C&quot;</span><span class="p">)</span>
        <span class="n">dups</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">cuts1</span> <span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="s">&quot;int64&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span> <span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="s">&quot;int64&quot;</span><span class="p">)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span> 
        <span class="n">dups</span><span class="p">[:,</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">cuts2</span> <span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="s">&quot;int64&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms2</span> <span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="s">&quot;int64&quot;</span><span class="p">)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span>
        <span class="n">dups</span><span class="o">.</span><span class="n">shape</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">N</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">strings</span> <span class="o">=</span> <span class="n">dups</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="s">&quot;|S16&quot;</span><span class="p">)</span>   <span class="c">#Converting two indices to a single string to run unique</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">strings</span><span class="p">)</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">N</span>
        <span class="n">uids</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">strings</span><span class="p">,</span><span class="n">return_index</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">del</span> <span class="n">strings</span><span class="p">,</span> <span class="n">dups</span> 
        <span class="n">stay</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SS</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">stay</span><span class="p">[</span><span class="n">uids</span><span class="p">]</span> <span class="o">=</span> <span class="bp">True</span>
        <span class="n">ds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DS</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="k">print</span> <span class="s">&quot;     Number of DS reads changed - </span><span class="si">%d</span><span class="s"> ---&gt; </span><span class="si">%d</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">ds</span><span class="p">,</span><span class="n">ds</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">DS</span><span class="p">)</span> <span class="o">+</span> <span class="n">stay</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span> 
        <span class="k">del</span> <span class="n">uids</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">maskFilter</span><span class="p">(</span><span class="n">stay</span><span class="p">)</span>
        <span class="k">print</span>
        
</div>
<div class="viewcode-block" id="HiCdataset.fragmentSum"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.fragmentSum">[docs]</a>    <span class="k">def</span> <span class="nf">fragmentSum</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">fragments</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span> <span class="n">strands</span> <span class="o">=</span> <span class="s">&quot;both&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;returns sum of all counts for a set or subset of fragments</span>

<span class="sd">        Parameters</span>
<span class="sd">        ---------- </span>
<span class="sd">        fragments : list of fragment IDs, optional</span>
<span class="sd">            Use only this fragments. By default all fragments are used</span>
<span class="sd">        strands : 1,2 or &quot;both&quot; (default) </span>
<span class="sd">            Use only first or second side of the read (first has SS, second - doesn&#39;t) </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_buildFragments</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">fragments</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span> <span class="n">fragments</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span>                
        <span class="k">if</span> <span class="n">strands</span> <span class="o">==</span> <span class="s">&quot;both&quot;</span><span class="p">:</span>  
            <span class="k">return</span> <span class="n">sumByArray</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragids1</span><span class="p">,</span><span class="n">fragments</span><span class="p">)</span> <span class="o">+</span> <span class="n">sumByArray</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragids2</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">DS</span><span class="p">],</span><span class="n">fragments</span><span class="p">)</span> 
        <span class="k">if</span> <span class="n">strands</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span> <span class="k">return</span> <span class="n">sumByArray</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragids1</span><span class="p">,</span><span class="n">fragments</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">strands</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span><span class="k">return</span> <span class="n">sumByArray</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragids2</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">DS</span><span class="p">],</span><span class="n">fragments</span><span class="p">)</span>
        
        </div>
    <span class="k">def</span> <span class="nf">printStats</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">print</span> <span class="s">&quot;-----&gt; Statistics for the file  </span><span class="si">%s</span><span class="s">!&quot;</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">filename</span>
        <span class="k">print</span> <span class="s">&quot;     Single sided reads: &quot;</span> <span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">SS</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="k">print</span> <span class="s">&quot;     Double sided reads: &quot;</span> <span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">DS</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="n">ss1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">strands1</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span><span class="o">&gt;=</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">ss2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">strands2</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms2</span><span class="o">&gt;=</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">sf</span> <span class="o">=</span> <span class="n">ss1</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">+</span> <span class="n">ss2</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="n">sr</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ss1</span><span class="p">)</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">ss2</span><span class="p">)</span> <span class="o">-</span> <span class="n">sf</span>
        <span class="k">print</span> <span class="s">&quot;     reverse/forward bias&quot;</span><span class="p">,</span><span class="nb">float</span><span class="p">(</span><span class="n">sr</span><span class="p">)</span><span class="o">/</span><span class="n">sf</span>
                    
        
<div class="viewcode-block" id="HiCdataset.save"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.save">[docs]</a>    <span class="k">def</span> <span class="nf">save</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">filename</span><span class="p">):</span>
        <span class="s">&quot;Saves dataset to filename, does not change the working file.&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">filename</span> <span class="o">==</span> <span class="n">filename</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">StandardError</span><span class="p">(</span><span class="s">&quot;Cannot save to the working file&quot;</span><span class="p">)</span>
        <span class="n">newh5dict</span> <span class="o">=</span> <span class="n">h5dict</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span><span class="n">mode</span> <span class="o">=</span> <span class="s">&#39;w&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectors</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span> <span class="n">newh5dict</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">h5dict</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
        <span class="k">print</span> <span class="s">&quot;----&gt; Data saved to file </span><span class="si">%s</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">filename</span><span class="p">,)</span>
    </div>
<div class="viewcode-block" id="HiCdataset.load"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.load">[docs]</a>    <span class="k">def</span> <span class="nf">load</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">filename</span><span class="p">):</span>
        <span class="s">&quot;Loads dataset from file to working file; check for inconsistency&quot;</span>
        <span class="n">otherh5dict</span> <span class="o">=</span> <span class="n">h5dict</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span><span class="s">&#39;r&#39;</span><span class="p">)</span>
        <span class="n">length</span> <span class="o">=</span> <span class="mi">0</span> 
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectors</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">otherh5dict</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
            <span class="n">ld</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">length</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> 
                <span class="n">length</span> <span class="o">=</span> <span class="n">ld</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ld</span> <span class="o">!=</span> <span class="n">length</span><span class="p">:</span> 
                    <span class="k">print</span><span class="p">(</span><span class="s">&quot;----&gt;!!!!!File </span><span class="si">%s</span><span class="s"> contains inconsistend data&lt;----&quot;</span> <span class="o">%</span> <span class="n">filename</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">exitProgram</span><span class="p">(</span><span class="s">&quot;----&gt; Sorry...&quot;</span><span class="p">)</span>
                    
            <span class="bp">self</span><span class="o">.</span><span class="n">_setData</span><span class="p">(</span><span class="n">name</span><span class="p">,</span><span class="n">data</span><span class="p">)</span> 
        <span class="k">print</span> <span class="s">&quot;----&gt;Loaded data from file </span><span class="si">%s</span><span class="s">, contains </span><span class="si">%d</span><span class="s"> reads&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">length</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">N</span> <span class="o">=</span> <span class="n">length</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">rebuildFragments</span><span class="p">()</span>
            

        </div>
<div class="viewcode-block" id="HiCdataset.saveHeatmap"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCdataset.saveHeatmap">[docs]</a>    <span class="k">def</span> <span class="nf">saveHeatmap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">filename</span><span class="p">,</span><span class="n">resolution</span> <span class="o">=</span> <span class="mi">1000000</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Saves heatmap to filename at given resolution. </span>
<span class="sd">                </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename : str</span>
<span class="sd">        resolution : int              </span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="k">try</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span> <span class="k">pass</span>        
        <span class="n">heatmap</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">buildAllHeatmap</span><span class="p">(</span><span class="n">resolution</span><span class="p">)</span>        
        <span class="n">singles</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">buildSinglesCoverage</span><span class="p">(</span><span class="n">resolution</span><span class="p">)</span>        
        <span class="n">frags</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">buildFragmetCoverage</span><span class="p">(</span><span class="n">resolution</span><span class="p">)</span>
        <span class="n">chromosomeStarts</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmStartsBinCont</span><span class="p">)</span>
        <span class="n">tosave</span> <span class="o">=</span> <span class="n">h5dict</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="n">filename</span><span class="p">,</span><span class="n">mode</span> <span class="o">=</span> <span class="s">&quot;w&quot;</span><span class="p">)</span>
        <span class="n">tosave</span><span class="p">[</span><span class="s">&quot;resolution&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">resolution</span>
        <span class="n">tosave</span><span class="p">[</span><span class="s">&quot;heatmap&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">heatmap</span>
        <span class="n">tosave</span><span class="p">[</span><span class="s">&quot;singles&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">singles</span>
        <span class="n">tosave</span><span class="p">[</span><span class="s">&quot;frags&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">frags</span>
        <span class="n">tosave</span><span class="p">[</span><span class="s">&quot;genomeBinNum&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">numBins</span>
        <span class="n">tosave</span><span class="p">[</span><span class="s">&quot;chromosomeSTarts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">chromosomeStarts</span>
        

        
        <span class="k">print</span> <span class="s">&quot;----&gt; Heatmap saved to &#39;</span><span class="si">%s</span><span class="s">&#39; at </span><span class="si">%d</span><span class="s"> resolution&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">filename</span><span class="p">,</span><span class="n">resolution</span><span class="p">)</span>
        </div>
    <span class="k">def</span> <span class="nf">exitProgram</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">a</span><span class="p">):</span>
        <span class="k">print</span> <span class="n">a</span>
        <span class="k">print</span> <span class="s">&quot;     ----&gt; Bye! :) &lt;----&quot;</span>
        <span class="nb">exit</span><span class="p">()</span>
</div>
<div class="viewcode-block" id="HiCStatistics"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCStatistics">[docs]</a><span class="k">class</span> <span class="nc">HiCStatistics</span><span class="p">(</span><span class="n">HiCdataset</span><span class="p">):</span>
    <span class="s">&quot;a sub-class of a &#39;HiCdataset&#39; class used to do statistics on Hi-C reads&quot;</span> 

<div class="viewcode-block" id="HiCStatistics.multiplyForTesting"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCStatistics.multiplyForTesting">[docs]</a>    <span class="k">def</span> <span class="nf">multiplyForTesting</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">N</span> <span class="o">=</span> <span class="mi">100</span><span class="p">):</span>
        <span class="s">&quot;used for heavy load testing only&quot;</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">vectors</span><span class="p">:</span>
            <span class="k">print</span> <span class="s">&quot;multipliing&quot;</span><span class="p">,</span><span class="n">name</span>
            <span class="n">one</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_getData</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">name</span> <span class="ow">in</span> <span class="p">[</span><span class="s">&quot;mids1&quot;</span><span class="p">,</span><span class="s">&quot;mids2&quot;</span><span class="p">]:</span>
                <span class="n">one</span> <span class="o">+=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="o">*</span><span class="n">N</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">one</span><span class="p">))</span>            
            <span class="n">blowup</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">hstack</span><span class="p">(</span><span class="nb">tuple</span><span class="p">([</span><span class="n">one</span><span class="p">]</span><span class="o">*</span><span class="n">N</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_setData</span><span class="p">(</span><span class="n">name</span><span class="p">,</span><span class="n">blowup</span><span class="p">)</span>
            
    </div>
<div class="viewcode-block" id="HiCStatistics.buildLengthDependencePlot"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCStatistics.buildLengthDependencePlot">[docs]</a>    <span class="k">def</span> <span class="nf">buildLengthDependencePlot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">label</span> <span class="o">=</span> <span class="s">&quot;plot&quot;</span><span class="p">,</span> <span class="n">strands</span> <span class="o">=</span> <span class="s">&quot;both&quot;</span><span class="p">,</span><span class="n">color</span> <span class="o">=</span> <span class="bp">None</span><span class="p">):</span>
        <span class="s">&quot;plots dependence of counts on fragment length. May do based on one strands only&quot;</span>
        <span class="s">&quot;please run  plt.legend &amp; plt.show() after calling this for all datasets you want to consider&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_buildFragments</span><span class="p">()</span>     
        <span class="n">fragmentLength</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ufragmentlen</span>
        <span class="n">pls</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">fragmentLength</span><span class="p">)</span>
        <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">fragmentLength</span><span class="p">)</span>
        <span class="n">sums</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">sizes</span> <span class="o">=</span> <span class="p">[]</span>            
        <span class="n">mysum</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragmentSum</span><span class="p">(</span><span class="bp">None</span><span class="p">,</span><span class="n">strands</span><span class="p">)</span>
    
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.98</span><span class="p">,</span><span class="mf">0.015</span><span class="p">):</span>
            <span class="n">b1</span><span class="p">,</span><span class="n">b2</span> <span class="o">=</span> <span class="n">pls</span><span class="p">[</span><span class="n">i</span><span class="o">*</span><span class="n">N</span><span class="p">],</span><span class="n">pls</span><span class="p">[(</span><span class="n">i</span><span class="o">+</span><span class="mf">0.015</span><span class="p">)</span><span class="o">*</span><span class="n">N</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">b1</span> <span class="o">&lt;</span> <span class="n">fragmentLength</span><span class="p">)</span><span class="o">*</span> <span class="p">(</span><span class="n">b2</span> <span class="o">&gt;</span> <span class="n">fragmentLength</span><span class="p">)</span>
            <span class="n">value</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">mysum</span><span class="p">[</span><span class="n">p</span><span class="p">])</span>            
            <span class="n">sums</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
            <span class="n">sizes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">fragmentLength</span><span class="p">[</span><span class="n">p</span><span class="p">]))</span>    
        <span class="k">if</span> <span class="n">color</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span> <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">sizes</span><span class="p">,</span><span class="n">sums</span><span class="p">,</span><span class="s">&#39;x-&#39;</span><span class="p">,</span><span class="n">markersize</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span><span class="n">linewidth</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span><span class="n">label</span> <span class="o">=</span> <span class="n">label</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>  <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">sizes</span><span class="p">,</span><span class="n">sums</span><span class="p">,</span><span class="n">color</span><span class="p">,</span><span class="n">linewidth</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span><span class="n">label</span> <span class="o">=</span> <span class="n">label</span><span class="p">)</span>
    
        </div>
<div class="viewcode-block" id="HiCStatistics.plotScaling"><a class="viewcode-back" href="../../fragment.html#hiclib.fragmentHiC.HiCStatistics.plotScaling">[docs]</a>    <span class="k">def</span> <span class="nf">plotScaling</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">fragids1</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span><span class="n">fragids2</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span>   <span class="c">#IDs of fragments for which to plot scaling. </span>
                    <span class="c">#One can, for example, limit oneself to only fragments shorter than 1000 bp </span>
                    <span class="c">#Or calculate scaling only between different arms                    </span>
                    <span class="n">useWeights</span> <span class="o">=</span> <span class="bp">False</span> <span class="p">,</span>        <span class="c">#use weights associated with fragment length                    </span>
                    <span class="n">excludeNeighbors</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span> <span class="n">enzyme</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span>   <span class="c">#number of neighboring fragments to exclude. Enzyme is needed for that!   </span>
                    <span class="n">normalize</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span>          <span class="c">#normalize the final plot to sum to one</span>
                    <span class="n">withinArms</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span>                <span class="c">#Treat chromosomal arms separately</span>
                    <span class="n">mindist</span> <span class="o">=</span> <span class="mi">10000</span><span class="p">,</span>  <span class="c">#Scaling was proved to be unreliable under 10000 bp for 6-cutter enzymes</span>
                    <span class="n">maxdist</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span>
                    <span class="c">#----------Calculating scaling within a set of regions only------</span>
                    <span class="n">regions</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span>   <span class="c">#Array of tuples (chrom, start, end) for which scaling should be calculated</span>
                    <span class="c">#Note that calculation might be extremely long (it might be proportional to # of regions for # &gt; 100)</span>
                    <span class="n">appendReadCount</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span>   <span class="c">#Append read count to the plot label </span>
                    <span class="o">**</span><span class="n">kwargs</span> <span class="c">#kwargs to be passed to plotting                       </span>
                    <span class="p">):</span>               <span class="c">#Sad smiley, because this method is very painful and complicated</span>
        <span class="sd">&quot;&quot;&quot;plots scaling over, possibly uses subset of fragmetns, or weigts, possibly normalizes after plotting</span>
<span class="sd">        </span>
<span class="sd">        Plan of scaling calculation: </span>
<span class="sd">        </span>
<span class="sd">        1. Subdivide all genome into regions. \n         </span>
<span class="sd">            a. Different chromosomes \n            </span>
<span class="sd">            b. Different arms \n            </span>
<span class="sd">            c. User defined squares/rectangles on a contact map \n            </span>
<span class="sd">               -(chromosome, start,end) square around the diagonal \n                </span>
<span class="sd">               -(chr, st1, end1, st2, end2) rectangle \n</span>
<span class="sd">                       </span>
<span class="sd">        2. Use either all fragments, or only interactions between two groups of fragments \n         </span>
<span class="sd">            e.g. you can calculate how scaling for small fragments is different from that for large \n            </span>
<span class="sd">            It can be possibly used for testing Hi-C protocol issues. \n             </span>
<span class="sd">            One can see effect of weights by doing this \n        </span>
<span class="sd">            </span>
<span class="sd">        3. (optional) Calculate correction associated with fragment length dependence </span>
<span class="sd">                         </span>
<span class="sd">        4. Subdivide all possible genomic separation into log-spaced bins</span>
<span class="sd">                 </span>
<span class="sd">        5. Calculate expected number of fragment pairs within each bin (possibly with weights from step 3).</span>
<span class="sd">         </span>
<span class="sd">            If exclusion of neighbors is specificed, expected number of fragments knows about this</span>
<span class="sd">            </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        fragids1, fragids2 : numpy.array of fragment IDs, optional </span>
<span class="sd">            Scaling is calculated only for interactions between fragids1 and fragids2</span>
<span class="sd">            If omitted, all fragments are used</span>
<span class="sd">            If boolean array is supplied, it serves as a mask for fragments. </span>
<span class="sd">        useWeights : bool, optional </span>
<span class="sd">            Use weights calculated from fragment length</span>
<span class="sd">        excludeNeighbors : int or None, optional</span>
<span class="sd">            If None, all fragment pairs are considered. </span>
<span class="sd">            If integer, only fragment pairs separated by this rsites are considered.</span>
<span class="sd">        enzyme : string (&quot;HindIII&quot;,&quot;NcoI&quot;)</span>
<span class="sd">            If excludeNeighbors is used, you have to specify restriction enzyme</span>
<span class="sd">        normalize : bool, optional</span>
<span class="sd">            Do an overall normalization of the answer, by default True. </span>
<span class="sd">        withinArms : bool, optional</span>
<span class="sd">            Set to false to use whole chromosomes instead of arms</span>
<span class="sd">        mindist, maxdist : int, optional</span>
<span class="sd">            Use lengthes from mindist to maxdist</span>
<span class="sd">        regions : list of (chrom, start,end) or (ch,st1,end1,st2,end2), optional</span>
<span class="sd">            Restrict scaling calculation to only certain squares of the map</span>
<span class="sd">        appendReadCount : bool, optional </span>
<span class="sd">            Append read count to the plot label</span>
<span class="sd">        kwargs : dict, optional</span>
<span class="sd">            Dictionary of kw args to be passed to plt.plot</span>
<span class="sd">            </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        (bins,probabilities) - values to plot on the scaling plot        </span>
<span class="sd">         </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_buildFragments</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">excludeNeighbors</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span> <span class="n">excludeNeighbors</span> <span class="o">=</span> <span class="bp">None</span>   <span class="c">#Not excluding neighbors  </span>
        <span class="c">#use all fragments if they&#39;re not specified </span>
        <span class="k">if</span> <span class="n">fragids1</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span> <span class="n">fragids1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span>
        <span class="k">if</span> <span class="n">fragids2</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span> <span class="n">fragids2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span>
        <span class="k">if</span> <span class="n">fragids1</span><span class="o">.</span><span class="n">dtype</span> <span class="o">==</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bool</span><span class="p">:</span> 
            <span class="n">fragids1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span><span class="p">[</span><span class="n">fragids1</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">fragids2</span><span class="o">.</span><span class="n">dtype</span> <span class="o">==</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bool</span><span class="p">:</span>
            <span class="n">fragids2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span><span class="p">[</span><span class="n">fragids2</span><span class="p">]</span>
        
        <span class="c">#Calculate regions if not specified </span>
        <span class="k">if</span> <span class="n">regions</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span> 
            <span class="k">if</span> <span class="n">withinArms</span> <span class="o">==</span> <span class="bp">False</span><span class="p">:</span> 
                <span class="n">regions</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmLens</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmCount</span><span class="p">)]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">regions</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">cntrMids</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmCount</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[(</span><span class="n">i</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">cntrMids</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmLens</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">chrmCount</span><span class="p">)]</span>
                
        
        <span class="k">if</span> <span class="n">maxdist</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span> <span class="n">maxdist</span> <span class="o">=</span> <span class="nb">max</span> <span class="p">(</span> <span class="nb">max</span><span class="p">([</span><span class="n">i</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">-</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">regions</span><span class="p">]),</span>  <span class="c">#normal regions </span>
                        <span class="nb">max</span><span class="p">([</span><span class="nb">abs</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">-</span> <span class="n">i</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">regions</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">3</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="mi">0</span><span class="p">]),</span>   <span class="c">#rectangular regions</span>
                        <span class="nb">max</span><span class="p">([</span><span class="nb">abs</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">i</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">regions</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">3</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="mi">0</span><span class="p">]))</span>

        <span class="n">regionID</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span><span class="p">),</span><span class="n">numpy</span><span class="o">.</span><span class="n">int16</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>  <span class="c">#Region to which a read belongs             </span>
        <span class="n">chr1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chrms1</span>
        <span class="n">chr2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chrms2</span>
        <span class="n">pos1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mids1</span>
        <span class="n">pos2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mids2</span>
        <span class="n">fragRegions1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">fragids1</span><span class="p">),</span><span class="nb">int</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="n">fragRegions2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">fragids2</span><span class="p">),</span><span class="nb">int</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span> 
        <span class="n">fragch1</span> <span class="o">=</span>  <span class="n">fragids1</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span>
        <span class="n">fragch2</span> <span class="o">=</span>  <span class="n">fragids2</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span>
        <span class="n">fragpos1</span> <span class="o">=</span> <span class="n">fragids1</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span>
        <span class="n">fragpos2</span> <span class="o">=</span> <span class="n">fragids2</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span>            
        
        <span class="k">for</span> <span class="n">regionNum</span><span class="p">,</span><span class="n">region</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">regions</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">region</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span> 
                <span class="n">chrom</span><span class="p">,</span><span class="n">start1</span><span class="p">,</span><span class="n">end1</span>  <span class="o">=</span> <span class="n">region</span>
                <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">chr1</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">pos1</span> <span class="o">&gt;</span> <span class="n">start1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">pos1</span> <span class="o">&lt;</span> <span class="n">end1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">chr2</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">pos2</span> <span class="o">&gt;</span> <span class="n">start1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">pos2</span> <span class="o">&lt;</span> <span class="n">end1</span><span class="p">)</span>                                     
                <span class="n">regionID</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span> <span class="o">=</span> <span class="n">regionNum</span> 
                <span class="n">mask1</span> <span class="o">=</span> <span class="p">(</span><span class="n">fragch1</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">fragpos1</span> <span class="o">&gt;</span> <span class="n">start1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">fragpos1</span> <span class="o">&lt;</span> <span class="n">end1</span><span class="p">)</span>
                <span class="n">mask2</span> <span class="o">=</span> <span class="p">(</span><span class="n">fragch2</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">fragpos2</span> <span class="o">&gt;</span> <span class="n">start1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">fragpos2</span> <span class="o">&lt;</span> <span class="n">end1</span><span class="p">)</span>   
                <span class="n">fragRegions1</span><span class="p">[</span><span class="n">mask1</span><span class="p">]</span> <span class="o">=</span> <span class="n">regionNum</span>
                <span class="n">fragRegions2</span><span class="p">[</span><span class="n">mask2</span><span class="p">]</span> <span class="o">=</span> <span class="n">regionNum</span>  

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">region</span><span class="p">)</span> <span class="o">==</span> <span class="mi">5</span><span class="p">:</span> 
                <span class="n">chrom</span><span class="p">,</span><span class="n">start1</span><span class="p">,</span><span class="n">end1</span><span class="p">,</span><span class="n">start2</span><span class="p">,</span><span class="n">end2</span> <span class="o">=</span> <span class="n">region</span>
                <span class="n">mask1</span> <span class="o">=</span> <span class="p">(</span><span class="n">chr1</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">chr2</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">pos1</span> <span class="o">&gt;</span> <span class="n">start1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">pos1</span> <span class="o">&lt;</span> <span class="n">end1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">pos2</span> <span class="o">&gt;</span> <span class="n">start2</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">pos2</span>  <span class="o">&lt;</span> <span class="n">end2</span><span class="p">)</span>
                <span class="n">mask2</span> <span class="o">=</span> <span class="p">(</span><span class="n">chr1</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">chr2</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">pos1</span> <span class="o">&gt;</span> <span class="n">start2</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">pos1</span> <span class="o">&lt;</span> <span class="n">end2</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">pos2</span> <span class="o">&gt;</span> <span class="n">start1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">pos2</span>  <span class="o">&lt;</span> <span class="n">end1</span><span class="p">)</span>
                <span class="n">mask</span> <span class="o">=</span> <span class="n">mask1</span> <span class="o">+</span> <span class="n">mask2</span>
                <span class="n">regionID</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span> <span class="o">=</span> <span class="n">regionNum</span>                    
                <span class="n">mask1</span> <span class="o">=</span> <span class="p">(</span><span class="n">fragch1</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">fragpos1</span> <span class="o">&gt;</span> <span class="n">start1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">fragpos1</span> <span class="o">&lt;</span> <span class="n">end1</span><span class="p">)</span>
                <span class="n">mask2</span> <span class="o">=</span> <span class="p">(</span><span class="n">fragch2</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">fragpos2</span> <span class="o">&gt;</span> <span class="n">start2</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">fragpos2</span> <span class="o">&lt;</span> <span class="n">end2</span><span class="p">)</span>   
                <span class="n">fragRegions1</span><span class="p">[</span><span class="n">mask1</span><span class="p">]</span> <span class="o">=</span> <span class="n">regionNum</span>
                <span class="n">fragRegions2</span><span class="p">[</span><span class="n">mask2</span><span class="p">]</span> <span class="o">=</span> <span class="n">regionNum</span>
        <span class="k">del</span> <span class="n">chr1</span><span class="p">,</span><span class="n">chr2</span><span class="p">,</span><span class="n">pos1</span><span class="p">,</span><span class="n">pos2</span>

                                        
        <span class="n">lens</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">numutils</span><span class="o">.</span><span class="n">logbins</span><span class="p">(</span><span class="n">mindist</span><span class="p">,</span><span class="n">maxdist</span><span class="p">,</span><span class="mf">1.25</span><span class="p">),</span><span class="nb">float</span><span class="p">)</span><span class="o">+</span><span class="mf">0.1</span>   <span class="c">#bins of lengths        </span>
        
        <span class="n">positions</span> <span class="o">=</span> <span class="p">[]</span>   <span class="c">#arrays to write the result </span>
        <span class="n">values</span> <span class="o">=</span> <span class="p">[]</span>
        
        <span class="k">if</span> <span class="n">excludeNeighbors</span> <span class="o">!=</span> <span class="bp">None</span><span class="p">:</span> 
            <span class="k">if</span> <span class="n">enzyme</span> <span class="o">==</span> <span class="bp">None</span><span class="p">:</span> <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Please specify enzyme if you&#39;re excluding Neighbors&quot;</span><span class="p">)</span>
            <span class="n">fragmentDists</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">getFragmentDistance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragids1</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">fragids2</span><span class="p">,</span> <span class="n">enzyme</span><span class="p">)</span>
            <span class="n">mask3</span> <span class="o">=</span> <span class="n">fragmentDists</span> <span class="o">&gt;</span> <span class="n">excludeNeighbors</span>   <span class="c">#keep only guys more than excludeNeighbors fragments apart </span>
        
        <span class="c">#Keeping reads for fragments in use</span>
        <span class="n">p11</span> <span class="o">=</span> <span class="n">arrayInArray</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragids1</span><span class="p">,</span><span class="n">fragids1</span><span class="p">)</span>
        <span class="n">p12</span> <span class="o">=</span> <span class="n">arrayInArray</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragids1</span><span class="p">,</span><span class="n">fragids2</span><span class="p">)</span>
        <span class="n">p21</span> <span class="o">=</span> <span class="n">arrayInArray</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragids2</span><span class="p">,</span><span class="n">fragids1</span><span class="p">)</span>
        <span class="n">p22</span> <span class="o">=</span> <span class="n">arrayInArray</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragids2</span><span class="p">,</span><span class="n">fragids2</span><span class="p">)</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">p11</span><span class="o">*</span> <span class="n">p22</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">p12</span> <span class="o">*</span> <span class="n">p21</span><span class="p">)</span>   <span class="c">#reads between fragids1 and fragids2 </span>
        <span class="n">mask2</span> <span class="o">=</span> <span class="p">(</span><span class="n">mask</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">regionID</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">strands1</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">strands2</span><span class="p">)</span>
        <span class="c">#Reads only between fragments</span>
        <span class="c">#Reads from the same region</span>
        <span class="c">#Reads like --&gt; --&gt;    or &lt;-- &lt;--, discarding --&gt; &lt;-- and &lt;-- --&gt;             </span>
        <span class="k">if</span> <span class="n">excludeNeighbors</span> <span class="o">!=</span> <span class="bp">None</span><span class="p">:</span> <span class="n">mask2</span> <span class="o">=</span> <span class="n">mask2</span> <span class="o">*</span> <span class="n">mask3</span>    <span class="c">#remove all neighbors </span>
                    
        <span class="n">distances</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">distances</span><span class="p">[</span><span class="n">mask2</span><span class="p">])</span>            
                
        <span class="s">&quot;calculating fragments lengths for exclusions to expected # of counts&quot;</span>        
        <span class="c">#sorted fragment IDs and lengthes</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span><span class="p">)</span>
        <span class="n">usort</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ufragments</span><span class="p">[</span><span class="n">args</span><span class="p">]</span>        

        <span class="k">if</span> <span class="n">useWeights</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>   <span class="c">#calculating weights if needed  </span>
            <span class="k">try</span><span class="p">:</span>      <span class="bp">self</span><span class="o">.</span><span class="n">weights</span>
            <span class="k">except</span><span class="p">:</span>   <span class="bp">self</span><span class="o">.</span><span class="n">calculateWeights</span><span class="p">()</span>                
            <span class="n">uweights</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">weights</span><span class="p">[</span><span class="n">args</span><span class="p">]</span>   <span class="c">#weights for sorted fragment IDs </span>
            <span class="n">weights1</span> <span class="o">=</span> <span class="n">uweights</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="n">usort</span><span class="p">,</span><span class="n">fragids1</span><span class="p">)]</span>
            <span class="n">weights2</span> <span class="o">=</span> <span class="n">uweights</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="n">usort</span><span class="p">,</span><span class="n">fragids2</span><span class="p">)]</span>   <span class="c">#weghts for fragment IDs under  consideration        </span>
        
        <span class="n">lenmins</span><span class="p">,</span><span class="n">lenmaxs</span>  <span class="o">=</span> <span class="n">lens</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">lens</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>        
                 
        <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">lenmins</span><span class="p">)</span>   <span class="c">#number of bins </span>
        <span class="n">count</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="n">N</span><span class="p">)]</span>      <span class="c">#count of reads in each min </span>
        <span class="n">chr1</span> <span class="o">=</span> <span class="n">fragids1</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span>    
        <span class="n">chr2</span> <span class="o">=</span> <span class="n">fragids2</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span>
        <span class="n">pos1</span> <span class="o">=</span> <span class="n">fragids1</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span>        
        <span class="n">pos2</span> <span class="o">=</span> <span class="n">fragids2</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span>
        
        <span class="k">for</span> <span class="n">regionNumber</span><span class="p">,</span><span class="n">region</span>  <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">regions</span><span class="p">):</span>  

            <span class="n">mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">fragRegions1</span>  <span class="o">==</span> <span class="n">regionNumber</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">mask2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">fragRegions2</span>  <span class="o">==</span> <span class="n">regionNumber</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>  <span class="c">#filtering fragments that correspont to current region             </span>
            <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">mask</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">mask2</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span> <span class="k">continue</span>                             
            <span class="n">bp1</span><span class="p">,</span> <span class="n">bp2</span> <span class="o">=</span> <span class="n">pos1</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span> <span class="n">pos2</span><span class="p">[</span><span class="n">mask2</span><span class="p">]</span>         <span class="c">#positions of fragments on chromosome</span>
             
            <span class="n">p2arg</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">bp2</span><span class="p">)</span>   
            <span class="n">p2</span> <span class="o">=</span> <span class="n">bp2</span><span class="p">[</span><span class="n">p2arg</span><span class="p">]</span>           <span class="c">#sorted positions on the second fragment</span>

            <span class="k">if</span> <span class="n">excludeNeighbors</span> <span class="o">!=</span> <span class="bp">None</span><span class="p">:</span>
                <span class="s">&quot;calculating excluded fragments (neighbors) and their weights to subtract them later&quot;</span> 
                <span class="n">excFrag1</span><span class="p">,</span> <span class="n">excFrag2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">getPairsLessThanDistance</span><span class="p">(</span><span class="n">fragids1</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span> <span class="p">,</span> <span class="n">fragids2</span><span class="p">[</span><span class="n">mask2</span><span class="p">],</span> <span class="n">excludeNeighbors</span><span class="p">,</span> <span class="n">enzyme</span><span class="p">)</span>                                                
                <span class="n">excDists</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">excFrag2</span> <span class="o">-</span> <span class="n">excFrag1</span><span class="p">)</span>  <span class="c">#distances between excluded fragment pairs</span>
                <span class="k">if</span> <span class="n">useWeights</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>        
                    <span class="n">correctionWeights</span> <span class="o">=</span> <span class="n">weights1</span><span class="p">[</span><span class="n">numutils</span><span class="o">.</span><span class="n">arraySearch</span><span class="p">(</span><span class="n">fragids1</span><span class="p">,</span><span class="n">excFrag1</span><span class="p">)]</span>
                    <span class="n">correctionWeights</span> <span class="o">=</span> <span class="n">correctionWeights</span> <span class="o">*</span>  <span class="n">weights2</span><span class="p">[</span><span class="n">numutils</span><span class="o">.</span><span class="n">arraySearch</span><span class="p">(</span><span class="n">fragids2</span><span class="p">,</span><span class="n">excFrag2</span><span class="p">)]</span>   <span class="c">#weights for excluded fragment pairs                        </span>
            <span class="k">if</span> <span class="n">useWeights</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span> 
                <span class="n">w1</span><span class="p">,</span><span class="n">w2</span> <span class="o">=</span> <span class="n">weights1</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span> <span class="n">weights2</span><span class="p">[</span><span class="n">mask2</span><span class="p">]</span>                                  
                <span class="n">sw2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">r_</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">w2</span><span class="p">[</span><span class="n">p2arg</span><span class="p">])]</span>    <span class="c">#cumsum for sorted weights on 2 strand                         </span>
            <span class="k">for</span> <span class="n">lenmin</span><span class="p">,</span><span class="n">lenmax</span><span class="p">,</span><span class="n">index</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">lenmins</span><span class="p">,</span><span class="n">lenmaxs</span><span class="p">,</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">lenmins</span><span class="p">))):</span>
                <span class="s">&quot;Now calculating actual number of fragment pairs for a length-bin, or weight of all these pairs&quot;</span>                
                <span class="n">mymin</span><span class="p">,</span><span class="n">mymax</span> <span class="o">=</span> <span class="n">bp1</span> <span class="o">-</span> <span class="n">lenmax</span><span class="p">,</span> <span class="n">bp1</span> <span class="o">-</span> <span class="n">lenmin</span>   <span class="c">#calculating boundaries for fragments on a second strand                  </span>
                <span class="n">val1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="n">p2</span><span class="p">,</span><span class="n">mymin</span><span class="p">)</span>  <span class="c">#Calculating indexes on the second strand </span>
                <span class="n">val2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="n">p2</span><span class="p">,</span><span class="n">mymax</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">useWeights</span> <span class="o">==</span> <span class="bp">False</span><span class="p">:</span> <span class="n">curcount</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">val1</span> <span class="o">-</span> <span class="n">val2</span><span class="p">))</span>   <span class="c">#just # of fragments </span>
                <span class="k">else</span><span class="p">:</span> <span class="n">curcount</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">w1</span> <span class="o">*</span> <span class="n">numpy</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">sw2</span><span class="p">[</span><span class="n">val1</span><span class="p">]</span> <span class="o">-</span> <span class="n">sw2</span><span class="p">[</span><span class="n">val2</span><span class="p">]))</span>      <span class="c">#(difference in cumsum of weights) * my weight </span>
                
                <span class="n">mymin</span><span class="p">,</span><span class="n">mymax</span> <span class="o">=</span> <span class="n">bp1</span> <span class="o">+</span> <span class="n">lenmax</span><span class="p">,</span> <span class="n">bp1</span> <span class="o">+</span> <span class="n">lenmin</span>   <span class="c">#repeating the same for </span>
                <span class="n">val1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="n">p2</span><span class="p">,</span><span class="n">mymin</span><span class="p">)</span>
                <span class="n">val2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="n">p2</span><span class="p">,</span><span class="n">mymax</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">useWeights</span> <span class="o">==</span> <span class="bp">False</span><span class="p">:</span> <span class="n">curcount</span> <span class="o">+=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">val1</span> <span class="o">-</span> <span class="n">val2</span><span class="p">))</span>
                <span class="k">else</span><span class="p">:</span> <span class="n">curcount</span> <span class="o">+=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">w1</span> <span class="o">*</span> <span class="n">numpy</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">sw2</span><span class="p">[</span><span class="n">val1</span><span class="p">]</span> <span class="o">-</span> <span class="n">sw2</span><span class="p">[</span><span class="n">val2</span><span class="p">]))</span>
                
                <span class="k">if</span> <span class="n">excludeNeighbors</span> <span class="o">!=</span> <span class="bp">None</span><span class="p">:</span> <span class="c">#now modifying expected count because of excluded fragments </span>
                    <span class="k">if</span> <span class="n">useWeights</span> <span class="o">==</span> <span class="bp">False</span><span class="p">:</span> 
                        <span class="n">ignore</span> <span class="o">=</span> <span class="p">((</span><span class="n">excDists</span> <span class="o">&gt;</span> <span class="n">lenmin</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">excDists</span> <span class="o">&lt;</span> <span class="n">lenmax</span><span class="p">))</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
                    <span class="k">else</span><span class="p">:</span>                        
                        <span class="n">ignore</span> <span class="o">=</span> <span class="p">(</span><span class="n">correctionWeights</span><span class="p">[((</span><span class="n">excDists</span> <span class="o">&gt;</span> <span class="n">lenmin</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">excDists</span> <span class="o">&lt;</span> <span class="n">lenmax</span><span class="p">))])</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
                    
                    <span class="k">if</span> <span class="p">(</span><span class="n">ignore</span> <span class="o">&gt;=</span> <span class="n">curcount</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">ignore</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">):</span>
                        <span class="k">if</span> <span class="n">ignore</span> <span class="o">&lt;</span> <span class="n">curcount</span> <span class="o">*</span> <span class="mf">1.0001</span><span class="p">:</span>
                            <span class="n">curcount</span> <span class="o">=</span> <span class="n">ignore</span> <span class="o">=</span> <span class="mi">0</span>   <span class="c">#Check for error and numerical instabilities  </span>
                        <span class="k">else</span><span class="p">:</span>   <span class="k">print</span> <span class="s">&quot;error found&quot;</span> <span class="p">,</span> <span class="s">&quot;lenmin:&quot;</span><span class="p">,</span><span class="n">lenmin</span><span class="p">,</span> <span class="s">&quot;  curcount:&quot;</span><span class="p">,</span><span class="n">curcount</span><span class="p">,</span> <span class="s">&quot;  ignore:&quot;</span><span class="p">,</span>  <span class="n">ignore</span>
                    <span class="k">else</span><span class="p">:</span> <span class="c">#Everything is all right                                                                      </span>
                        <span class="n">curcount</span>  <span class="o">-=</span> <span class="n">ignore</span>
                <span class="n">count</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="o">+=</span> <span class="n">curcount</span>
                <span class="c">#print index,count[index]</span>
        <span class="n">maxcountsall</span> <span class="o">=</span> <span class="n">count</span>        
        <span class="n">rawValues</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">lens</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>   <span class="c">#Dividing observed by expected</span>
            <span class="n">beg</span><span class="p">,</span><span class="n">end</span>  <span class="o">=</span> <span class="n">lens</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">lens</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span>            
            <span class="n">first</span><span class="p">,</span><span class="n">last</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="n">distances</span><span class="p">,[</span><span class="n">beg</span><span class="p">,</span><span class="n">end</span><span class="p">]))</span>             
            <span class="n">mycounts</span> <span class="o">=</span> <span class="n">last</span> <span class="o">-</span> <span class="n">first</span>
            <span class="n">maxcounts</span> <span class="o">=</span> <span class="n">maxcountsall</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>                                     
            <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sqrt</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">beg</span><span class="p">)</span><span class="o">*</span><span class="nb">float</span><span class="p">(</span><span class="n">end</span><span class="p">)))</span>
            <span class="n">values</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mycounts</span><span class="o">/</span><span class="nb">float</span><span class="p">(</span><span class="n">maxcounts</span><span class="p">))</span>
            <span class="n">rawValues</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mycounts</span><span class="p">)</span>
            
        <span class="n">positions</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span>
        <span class="n">values</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">values</span><span class="p">)</span>
         
        <span class="k">if</span> <span class="n">normalize</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span> <span class="n">values</span> <span class="o">/=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mf">1.</span> <span class="o">*</span> <span class="p">(</span><span class="n">positions</span> <span class="o">*</span> <span class="n">values</span><span class="p">)</span> <span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">logical_not</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">positions</span> <span class="o">*</span> <span class="n">values</span><span class="p">))])</span>        
        
        <span class="k">if</span> <span class="n">appendReadCount</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span> 
            <span class="k">if</span> <span class="s">&quot;label&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span> 
                <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;label&quot;</span><span class="p">]</span>  <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;label&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="s">&quot;, </span><span class="si">%d</span><span class="s"> reads&quot;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">distances</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span><span class="n">values</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="n">values</span><span class="p">)</span> 
        
    </div>
    <span class="k">def</span> <span class="nf">plotRsiteStartDistribution</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">useSSReadsOnly</span> <span class="o">=</span> <span class="bp">False</span><span class="p">,</span><span class="n">offset</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span><span class="n">length</span> <span class="o">=</span> <span class="mi">200</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">useSSReadsOnly</span> <span class="o">==</span> <span class="bp">True</span><span class="p">:</span>
            <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">SS</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DS</span>
        <span class="n">dists1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fraglens1</span> <span class="o">-</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dists1</span><span class="p">,</span><span class="n">dtype</span> <span class="o">=</span> <span class="s">&quot;int32&quot;</span><span class="p">)</span> 
        <span class="n">dists2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fraglens2</span> <span class="o">-</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dists2</span><span class="p">,</span><span class="n">dtype</span> <span class="o">=</span> <span class="s">&quot;int32&quot;</span><span class="p">)</span> 
        <span class="n">m</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">dists1</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span><span class="n">dists2</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="k">if</span> <span class="n">offset</span> <span class="o">&lt;</span> <span class="o">-</span><span class="n">m</span><span class="p">:</span> 
            <span class="n">offset</span> <span class="o">=</span> <span class="o">-</span><span class="n">m</span>
            <span class="k">print</span> <span class="s">&quot;minimum negative distance is </span><span class="si">%d</span><span class="s">, larger than offset; offset set to </span><span class="si">%d</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">m</span><span class="p">,</span><span class="o">-</span><span class="n">m</span><span class="p">)</span>
        <span class="n">dists1</span> <span class="o">+=</span> <span class="n">offset</span>
        <span class="n">dists2</span> <span class="o">+=</span> <span class="n">offset</span>          
        <span class="n">myrange</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="n">offset</span><span class="p">,</span><span class="n">length</span><span class="o">-</span><span class="n">offset</span><span class="p">)</span>
        
        <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">141</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s">&quot;strands1, side 1&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">myrange</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="mi">5</span><span class="o">+</span><span class="n">dists1</span><span class="p">[</span><span class="n">mask</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">strands1</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span> <span class="o">==</span> <span class="bp">True</span> <span class="p">])[:</span><span class="n">length</span><span class="p">])</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">142</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s">&quot;strands1, side 2&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">myrange</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">dists2</span><span class="p">[</span><span class="n">mask</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">strands1</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span> <span class="o">==</span> <span class="bp">True</span> <span class="p">])[:</span><span class="n">length</span><span class="p">])</span>
        
        <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">143</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s">&quot;strands2, side 1&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">myrange</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">dists1</span><span class="p">[</span><span class="n">mask</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">strands1</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span> <span class="o">==</span> <span class="bp">False</span> <span class="p">])[:</span><span class="n">length</span><span class="p">])</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">144</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s">&quot;strands2, side 2&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">myrange</span><span class="p">,</span><span class="n">numpy</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">dists2</span><span class="p">[</span><span class="n">mask</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">strands1</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span> <span class="o">==</span> <span class="bp">False</span> <span class="p">])[:</span><span class="n">length</span><span class="p">])</span>
        
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>





</div>
<span class="k">class</span> <span class="nc">experimentalFeatures</span><span class="p">(</span><span class="n">HiCdataset</span><span class="p">):</span>
    <span class="k">def</span> <span class="nf">splitFragmentsBystrands</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="s">&quot;Splits fragments: those with strands = 1 gets location += 100. This is totally safe!  &quot;</span>
        <span class="s">&quot;This might be fun if you want to analyze two sides of the fragment separately, but unnecessary otherwise&quot;</span>
        <span class="n">f1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragids1</span>
        <span class="n">f1</span> <span class="o">+=</span> <span class="p">((</span><span class="n">f1</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span><span class="p">)</span> <span class="o">%</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">f1</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">strands1</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span> 
        <span class="n">f2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragids2</span>
        <span class="n">f1</span> <span class="o">+=</span> <span class="p">((</span><span class="n">f2</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragIDmult</span><span class="p">)</span> <span class="o">%</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">f2</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">strands1</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">rebuildFragments</span><span class="p">()</span>
    



<span class="k">def</span> <span class="nf">plotFigure2c</span><span class="p">():</span>
    <span class="n">TR</span> <span class="o">=</span> <span class="n">HiCdataset</span><span class="p">()</span>
    <span class="n">TR</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&quot;GM-all.refined&quot;</span><span class="p">)</span>
    <span class="n">hm</span> <span class="o">=</span> <span class="n">TR</span><span class="o">.</span><span class="n">buildHeatmap</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1000000</span><span class="p">,</span> <span class="bp">False</span><span class="p">,</span><span class="bp">False</span><span class="p">)</span>
    <span class="n">TR</span><span class="o">.</span><span class="n">calculateWeights</span><span class="p">()</span>    
    <span class="n">TR</span><span class="o">.</span><span class="n">weights</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">TR</span><span class="o">.</span><span class="n">weights</span><span class="p">),</span><span class="nb">float</span><span class="p">)</span>   <span class="c">#if you want to correct just by fragment density, not by length dependence </span>
    <span class="n">hm2</span> <span class="o">=</span> <span class="n">TR</span><span class="o">.</span><span class="n">buildHeatmap</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1000000</span><span class="p">,</span> <span class="bp">False</span><span class="p">,</span><span class="n">weights</span> <span class="o">=</span> <span class="bp">True</span> <span class="p">)</span>
    <span class="n">hm2</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">hm2</span><span class="p">)]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">hm</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span>
    <span class="sd">&quot;&quot;&quot;p1-6 are 6 lines to be plotted, below is plotting only&quot;&quot;&quot;</span>
    <span class="n">p1</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">hm</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)[</span><span class="n">mask</span><span class="p">]</span>
    <span class="n">p3</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">correct</span><span class="p">(</span><span class="n">hm</span><span class="p">),</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)[</span><span class="n">mask</span><span class="p">]</span>
    <span class="n">p5</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">ultracorrect</span><span class="p">(</span><span class="n">hm</span><span class="p">,</span><span class="mi">40</span><span class="p">),</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)[</span><span class="n">mask</span><span class="p">]</span>
    <span class="n">p4</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">correct</span><span class="p">(</span><span class="n">hm2</span><span class="p">),</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)[</span><span class="n">mask</span><span class="p">]</span>
    <span class="n">p2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">hm2</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)[</span><span class="n">mask</span><span class="p">]</span>
    <span class="n">p6</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">ultracorrect</span><span class="p">(</span><span class="n">hm2</span><span class="p">,</span><span class="mi">40</span><span class="p">),</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)[</span><span class="n">mask</span><span class="p">]</span>
    <span class="n">matplotlib</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s">&#39;font.sans-serif&#39;</span><span class="p">]</span><span class="o">=</span><span class="s">&#39;Arial&#39;</span>        
    <span class="n">dashstyle</span> <span class="o">=</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>    
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
    
    <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="mi">80</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s">&quot;Total coverage&quot;</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">8</span><span class="p">)</span>
        
    <span class="n">line21</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">p1</span> <span class="o">/</span> <span class="n">p1</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span><span class="s">&quot;-&quot;</span><span class="p">,</span><span class="n">linewidth</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span><span class="n">color</span> <span class="o">=</span> <span class="s">&quot;#e5a826&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">line22</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">p3</span><span class="o">/</span><span class="n">p3</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span><span class="s">&quot;--&quot;</span><span class="p">,</span><span class="n">linewidth</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span><span class="n">color</span> <span class="o">=</span> <span class="s">&quot;#e5a826&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">line22</span><span class="o">.</span><span class="n">set_dashes</span><span class="p">(</span><span class="n">dashstyle</span><span class="p">)</span>
    <span class="n">line23</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">p5</span><span class="o">/</span><span class="n">p5</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span> <span class="n">linewidth</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s">&quot;grey&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        
    <span class="k">for</span> <span class="n">xlabel_i</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">():</span> <span class="n">xlabel_i</span><span class="o">.</span><span class="n">set_fontsize</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">xlabel_i</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_yticklabels</span><span class="p">():</span> <span class="n">xlabel_i</span><span class="o">.</span><span class="n">set_fontsize</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span>                    
    <span class="n">legend</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">([</span><span class="n">line21</span><span class="p">,</span> <span class="n">line22</span><span class="p">,</span> <span class="n">line23</span><span class="p">],</span>
                        <span class="p">[</span><span class="s">&quot;Raw data&quot;</span><span class="p">,</span><span class="s">&quot;Single correction&quot;</span><span class="p">,</span><span class="s">&quot;Iterative correction&quot;</span><span class="p">],</span><span class="n">prop</span><span class="o">=</span><span class="p">{</span><span class="s">&quot;size&quot;</span><span class="p">:</span><span class="mi">6</span><span class="p">},</span><span class="n">loc</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span><span class="n">handlelength</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">legend</span><span class="o">.</span><span class="n">draw_frame</span><span class="p">(</span><span class="bp">False</span><span class="p">)</span>
    <span class="n">removeAxes</span><span class="p">(</span><span class="n">shift</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span> <span class="n">ax</span><span class="p">)</span>
    
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="o">.</span><span class="n">values</span><span class="p">():</span> <span class="n">i</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s">&#39;none&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">&#39;black&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">&#39;black&#39;</span><span class="p">)</span>
    

    <span class="n">ax2</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="n">sharex</span> <span class="o">=</span> <span class="n">ax</span> <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="mi">80</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s">&quot;Position on chom 1 (MB)&quot;</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s">&quot;Total coverage&quot;</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">8</span><span class="p">)</span>

    <span class="n">line1</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">p4</span><span class="o">/</span><span class="n">p4</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span><span class="s">&quot;--&quot;</span><span class="p">,</span><span class="n">color</span> <span class="o">=</span> <span class="s">&quot;#9b3811&quot;</span><span class="p">,</span><span class="n">linewidth</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>    
    <span class="n">line1</span><span class="o">.</span><span class="n">set_dashes</span><span class="p">(</span><span class="n">dashstyle</span><span class="p">)</span>
    <span class="n">line2</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">p2</span> <span class="o">/</span> <span class="n">p2</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span><span class="s">&quot;-&quot;</span><span class="p">,</span><span class="n">color</span> <span class="o">=</span> <span class="s">&quot;#9b3811&quot;</span><span class="p">,</span><span class="n">linewidth</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>    
    <span class="n">line3</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">p6</span><span class="o">/</span><span class="n">p6</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span> <span class="n">linewidth</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s">&quot;grey&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">xlabel_i</span> <span class="ow">in</span> <span class="n">ax2</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">():</span> <span class="n">xlabel_i</span><span class="o">.</span><span class="n">set_fontsize</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">xlabel_i</span> <span class="ow">in</span> <span class="n">ax2</span><span class="o">.</span><span class="n">get_yticklabels</span><span class="p">():</span> <span class="n">xlabel_i</span><span class="o">.</span><span class="n">set_fontsize</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span>
    
    <span class="n">legend</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">([</span><span class="n">line2</span><span class="p">,</span><span class="n">line1</span><span class="p">,</span><span class="n">line3</span><span class="p">],</span>
                        <span class="p">[</span><span class="s">&quot;HindIII corrected&quot;</span><span class="p">,</span><span class="s">&quot;Single correction&quot;</span><span class="p">,</span><span class="s">&quot;Iterative correction&quot;</span><span class="p">],</span><span class="n">prop</span><span class="o">=</span><span class="p">{</span><span class="s">&quot;size&quot;</span><span class="p">:</span><span class="mi">6</span><span class="p">},</span><span class="n">loc</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span><span class="n">handlelength</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">legend</span><span class="o">.</span><span class="n">draw_frame</span><span class="p">(</span><span class="bp">False</span><span class="p">)</span>
    <span class="n">removeAxes</span><span class="p">(</span><span class="n">shift</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span> <span class="n">ax2</span><span class="p">)</span>
    <span class="n">plotting</span><span class="o">.</span><span class="n">niceShow</span><span class="p">()</span> 
    
<span class="c">#plotFigure2c()</span>


<span class="k">def</span> <span class="nf">doSupplementaryCoveragePlot</span><span class="p">():</span>
    <span class="n">TR</span> <span class="o">=</span> <span class="n">HiCdataset</span><span class="p">()</span>
    <span class="n">TR</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&quot;GM-all.refined&quot;</span><span class="p">)</span>
    <span class="n">s1</span> <span class="o">=</span> <span class="n">TR</span><span class="o">.</span><span class="n">fragmentSum</span><span class="p">(</span><span class="n">strands</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">TR</span><span class="o">.</span><span class="n">saveFragments</span><span class="p">()</span>
    <span class="n">TR</span><span class="o">.</span><span class="n">maskFilter</span><span class="p">(</span><span class="n">TR</span><span class="o">.</span><span class="n">dists1</span> <span class="o">&gt;</span> <span class="mi">500</span><span class="p">)</span>
    <span class="n">TR</span><span class="o">.</span><span class="n">originalFragments</span><span class="p">()</span>
    <span class="n">s2</span> <span class="o">=</span> <span class="n">TR</span><span class="o">.</span><span class="n">fragmentSum</span><span class="p">(</span><span class="n">strands</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">resolution</span> <span class="o">=</span> <span class="mi">1000000</span>
    <span class="k">def</span> <span class="nf">coverage</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span><span class="n">s2</span><span class="p">,</span><span class="n">TR</span><span class="p">):</span>
        <span class="n">genome</span> <span class="o">=</span> <span class="n">Genome</span><span class="p">()</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">createMapping</span><span class="p">(</span><span class="n">resolution</span><span class="p">)</span>            
        <span class="n">label</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">[</span><span class="n">TR</span><span class="o">.</span><span class="n">ufragments</span> <span class="o">/</span> <span class="n">TR</span><span class="o">.</span><span class="n">fragIDmult</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="p">(</span><span class="n">TR</span><span class="o">.</span><span class="n">ufragments</span> <span class="o">%</span> <span class="n">TR</span><span class="o">.</span><span class="n">fragIDmult</span> <span class="p">)</span> <span class="o">/</span> <span class="n">resolution</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">weights</span> <span class="o">=</span> <span class="n">s1</span><span class="p">)</span>
        <span class="n">counts2</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">label</span><span class="p">,</span><span class="n">weights</span> <span class="o">=</span> <span class="n">s2</span><span class="p">)</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">cPickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s">&quot;GC1M&quot;</span><span class="p">,</span><span class="s">&#39;rb&#39;</span><span class="p">))</span>
        <span class="n">eigenvector</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">genome</span><span class="o">.</span><span class="n">chromosomeEnds</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span><span class="nb">float</span><span class="p">)</span>
        <span class="n">inds</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">inds</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="mf">0.02</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">inds</span><span class="p">)):]</span>                
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">24</span><span class="p">):</span>
            <span class="n">eigenvector</span><span class="p">[</span><span class="n">genome</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">[</span><span class="n">chrom</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span><span class="n">genome</span><span class="o">.</span><span class="n">chromosomeStarts</span><span class="p">[</span><span class="n">chrom</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">chrom</span><span class="o">-</span><span class="mi">1</span><span class="p">])]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">chrom</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">eigenvector</span><span class="p">[</span><span class="n">eigenvector</span> <span class="o">&lt;</span> <span class="mi">35</span><span class="p">]</span> <span class="o">=</span> <span class="mi">35</span>        
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">counts</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span><span class="n">counts2</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span> <span class="n">c</span> <span class="o">=</span> <span class="n">eigenvector</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span><span class="n">s</span> <span class="o">=</span> <span class="mi">6</span><span class="p">,</span><span class="n">linewidth</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
        <span class="k">print</span> <span class="n">stats</span><span class="o">.</span><span class="n">spearmanr</span><span class="p">(</span><span class="n">counts</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span><span class="n">counts2</span><span class="p">[</span><span class="n">mask</span><span class="p">])</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s">&quot;Coverage from all reads&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">5000</span><span class="p">,</span><span class="mi">10000</span><span class="p">,</span><span class="mi">15000</span><span class="p">])</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s">&quot;Coverage from RBs&quot;</span><span class="p">)</span>
        <span class="n">b</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">()</span>
        <span class="n">b</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s">&quot;GC content&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">121</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s">&quot;HinIII&quot;</span><span class="p">)</span>
    <span class="n">coverage</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span><span class="n">s2</span><span class="p">,</span><span class="n">TR</span><span class="p">)</span>
    
    <span class="n">TR</span> <span class="o">=</span> <span class="n">HiCdataset</span><span class="p">()</span>
    <span class="n">TR</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&quot;GM-NcoI.refined&quot;</span><span class="p">)</span>
    <span class="n">s1</span> <span class="o">=</span> <span class="n">TR</span><span class="o">.</span><span class="n">fragmentSum</span><span class="p">(</span><span class="n">strands</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">TR</span><span class="o">.</span><span class="n">saveFragments</span><span class="p">()</span>
    <span class="n">TR</span><span class="o">.</span><span class="n">maskFilter</span><span class="p">(</span><span class="n">TR</span><span class="o">.</span><span class="n">dists1</span> <span class="o">&gt;</span> <span class="mi">500</span><span class="p">)</span>
    <span class="n">TR</span><span class="o">.</span><span class="n">originalFragments</span><span class="p">()</span>
    <span class="n">s2</span> <span class="o">=</span> <span class="n">TR</span><span class="o">.</span><span class="n">fragmentSum</span><span class="p">(</span><span class="n">strands</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">resolution</span> <span class="o">=</span> <span class="mi">1000000</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">122</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s">&quot;NcoI&quot;</span><span class="p">)</span>
    <span class="n">coverage</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span><span class="n">s2</span><span class="p">,</span><span class="n">TR</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span> 

 





  
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" size="18" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../../np-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../../index.html">mirnylab v0.1 documentation</a> &raquo;</li>
          <li><a href="../index.html" >Module code</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2012, the Leonids.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.1.
    </div>
  </body>
</html>